
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../resources/wrapapi/">
      
      
        <link rel="next" href="constants/">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.5.9">
    
    
      
        <title>sequence - symdesign Documentation</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.f2e4d321.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../css/code_select.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="deep-purple" data-md-color-accent="green">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#sequence" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="symdesign Documentation" class="md-header__button md-logo" aria-label="symdesign Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            symdesign Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              sequence
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="deep-purple" data-md-color-accent="green"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="deep-purple" data-md-color-accent="green"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m14.3 16-.7-2h-3.2l-.7 2H7.8L11 7h2l3.2 9h-1.9M20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69m-9.15 3.96h2.3L12 9l-1.15 3.65Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="deep-purple" data-md-color-accent="green"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/kylemeador/symdesign" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../.." class="md-tabs__link">
        
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../user_guide/" class="md-tabs__link">
        
  
    
  
  User Guide

      </a>
    </li>
  

      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../" class="md-tabs__link">
          
  
  API reference

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="symdesign Documentation" class="md-nav__button md-logo" aria-label="symdesign Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    symdesign Documentation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/kylemeador/symdesign" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../user_guide/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    User Guide
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="">
            
  
  <span class="md-ellipsis">
    API reference
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            API reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    symdesign
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../flags/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    flags
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_3" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../metrics/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    metrics
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3_3" id="__nav_3_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_3">
            <span class="md-nav__icon md-icon"></span>
            metrics
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../metrics/pose/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    pose
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_4" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../protocols/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    protocols
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3_4" id="__nav_3_4_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_4">
            <span class="md-nav__icon md-icon"></span>
            protocols
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../protocols/align/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    align
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../protocols/cluster/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    cluster
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../protocols/config/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    config
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../protocols/fragdock/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    fragdock
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../protocols/pose/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    pose
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../protocols/select/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    select
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../protocols/utils/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    utils
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_5" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../resources/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    resources
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3_5" id="__nav_3_5_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_5">
            <span class="md-nav__icon md-icon"></span>
            resources
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../resources/config/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    config
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../resources/database/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    database
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../resources/distribute/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    distribute
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../resources/job/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    job
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../resources/ml/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ml
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../resources/monomer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    monomer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../resources/multimer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    multimer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_5_8" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../resources/query/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    query
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3_5_8" id="__nav_3_5_8_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_3_5_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_5_8">
            <span class="md-nav__icon md-icon"></span>
            query
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../resources/query/pdb/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    pdb
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../resources/query/uniprot/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    uniprot
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../resources/sql/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    sql
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../resources/structure_db/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    structure_db
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../resources/wrapapi/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    wrapapi
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_6" checked>
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="./" class="md-nav__link md-nav__link--active">
              
  
  <span class="md-ellipsis">
    sequence
  </span>
  

            </a>
            
              
              <label class="md-nav__link md-nav__link--active" for="__nav_3_6" id="__nav_3_6_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_6_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3_6">
            <span class="md-nav__icon md-icon"></span>
            sequence
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="constants/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    constants
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="expression/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    expression
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_7" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../structure/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    structure
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3_7" id="__nav_3_7_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_7">
            <span class="md-nav__icon md-icon"></span>
            structure
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../structure/base/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    base
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../structure/coordinates/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    coordinates
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_7_3" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../structure/fragment/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    fragment
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3_7_3" id="__nav_3_7_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_3_7_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_7_3">
            <span class="md-nav__icon md-icon"></span>
            fragment
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../structure/fragment/db/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    db
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../structure/fragment/info/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    info
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../structure/fragment/metrics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    metrics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../structure/fragment/visuals/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    visuals
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../structure/model/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    model
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../structure/sequence/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    sequence
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../structure/utils/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    utils
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_8" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../tools/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    tools
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3_8" id="__nav_3_8_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_8">
            <span class="md-nav__icon md-icon"></span>
            tools
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tools/BenchmarkCollapseHydrophobicityScale/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    BenchmarkCollapseHydrophobicityScale
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tools/ConcatenatePDBFiles/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ConcatenatePDBFiles
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tools/ProfileScripts/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ProfileScripts
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tools/SlurmControl/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    SlurmControl
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tools/distribute/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    distribute
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tools/fetch_commands_from_slurm_output/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    fetch_commands_from_slurm_output
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tools/format_commands/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    format_commands
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tools/format_pose_ids/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    format_pose_ids
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tools/generate_reference_docs/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    generate_reference_docs
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tools/get_array_numbers_from_identifiers/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    get_array_numbers_from_identifiers
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tools/list_files_in_directory/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    list_files_in_directory
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tools/list_overlap/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    list_overlap
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tools/models_to_multimodel/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    models_to_multimodel
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tools/profile_gpu_nodes/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    profile_gpu_nodes
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tools/retrieve_oligomers/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    retrieve_oligomers
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tools/retrieve_pdbs_by_advanced_query/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    retrieve_pdbs_by_advanced_query
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tools/write_indices_for_C1_docking/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    write_indices_for_C1_docking
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_9" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../utils/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    utils
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3_9" id="__nav_3_9_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_9">
            <span class="md-nav__icon md-icon"></span>
            utils
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../utils/SymEntry/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    SymEntry
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../utils/cluster/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    cluster
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../utils/guide/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    guide
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../utils/nanohedra_parsing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    nanohedra_parsing
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../utils/path/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    path
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../utils/query/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    query
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../utils/rosetta/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    rosetta
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../utils/symmetry/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    symmetry
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../utils/types/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    types
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_10" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../visualization/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    visualization
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3_10" id="__nav_3_10_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_10_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_10">
            <span class="md-nav__icon md-icon"></span>
            visualization
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../visualization/SymmetryMates/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    SymmetryMates
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../visualization/VisualizeUtils/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    VisualizeUtils
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../visualization/shapes/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    shapes
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#sequence" class="md-nav__link">
    <span class="md-ellipsis">
      sequence
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.protein_letters_alph1" class="md-nav__link">
    <span class="md-ellipsis">
      protein_letters_alph1
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.protein_letters_alph1_extended" class="md-nav__link">
    <span class="md-ellipsis">
      protein_letters_alph1_extended
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.protein_letters_extended_and_gap" class="md-nav__link">
    <span class="md-ellipsis">
      protein_letters_extended_and_gap
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.ProfileDict" class="md-nav__link">
    <span class="md-ellipsis">
      ProfileDict
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.MutationEntry" class="md-nav__link">
    <span class="md-ellipsis">
      MutationEntry
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.mutation_dictionary" class="md-nav__link">
    <span class="md-ellipsis">
      mutation_dictionary
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.sequence_dictionary" class="md-nav__link">
    <span class="md-ellipsis">
      sequence_dictionary
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.MultipleSequenceAlignment" class="md-nav__link">
    <span class="md-ellipsis">
      MultipleSequenceAlignment
    </span>
  </a>
  
    <nav class="md-nav" aria-label="MultipleSequenceAlignment">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#sequence.MultipleSequenceAlignment.query" class="md-nav__link">
    <span class="md-ellipsis">
      query
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence.MultipleSequenceAlignment.sequence_weights" class="md-nav__link">
    <span class="md-ellipsis">
      sequence_weights
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence.MultipleSequenceAlignment.alphabet_length" class="md-nav__link">
    <span class="md-ellipsis">
      alphabet_length
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence.MultipleSequenceAlignment.counts_by_position" class="md-nav__link">
    <span class="md-ellipsis">
      counts_by_position
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence.MultipleSequenceAlignment.gap_index" class="md-nav__link">
    <span class="md-ellipsis">
      gap_index
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence.MultipleSequenceAlignment.query_aligned" class="md-nav__link">
    <span class="md-ellipsis">
      query_aligned
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence.MultipleSequenceAlignment.query_length" class="md-nav__link">
    <span class="md-ellipsis">
      query_length
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence.MultipleSequenceAlignment.length" class="md-nav__link">
    <span class="md-ellipsis">
      length
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence.MultipleSequenceAlignment.number_of_positions" class="md-nav__link">
    <span class="md-ellipsis">
      number_of_positions
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence.MultipleSequenceAlignment.sequences" class="md-nav__link">
    <span class="md-ellipsis">
      sequences
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence.MultipleSequenceAlignment.sequence_identifiers" class="md-nav__link">
    <span class="md-ellipsis">
      sequence_identifiers
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence.MultipleSequenceAlignment.alphabet_type" class="md-nav__link">
    <span class="md-ellipsis">
      alphabet_type
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence.MultipleSequenceAlignment.query_indices" class="md-nav__link">
    <span class="md-ellipsis">
      query_indices
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence.MultipleSequenceAlignment.sequence_indices" class="md-nav__link">
    <span class="md-ellipsis">
      sequence_indices
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence.MultipleSequenceAlignment.numerical_alignment" class="md-nav__link">
    <span class="md-ellipsis">
      numerical_alignment
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence.MultipleSequenceAlignment.array" class="md-nav__link">
    <span class="md-ellipsis">
      array
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence.MultipleSequenceAlignment.deletion_matrix" class="md-nav__link">
    <span class="md-ellipsis">
      deletion_matrix
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence.MultipleSequenceAlignment.frequencies" class="md-nav__link">
    <span class="md-ellipsis">
      frequencies
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence.MultipleSequenceAlignment.observations_by_position" class="md-nav__link">
    <span class="md-ellipsis">
      observations_by_position
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence.MultipleSequenceAlignment.gaps_per_position" class="md-nav__link">
    <span class="md-ellipsis">
      gaps_per_position
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence.MultipleSequenceAlignment.from_dictionary" class="md-nav__link">
    <span class="md-ellipsis">
      from_dictionary
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence.MultipleSequenceAlignment.from_seq_records" class="md-nav__link">
    <span class="md-ellipsis">
      from_seq_records
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence.MultipleSequenceAlignment.weight_alignment_by_sequence" class="md-nav__link">
    <span class="md-ellipsis">
      weight_alignment_by_sequence
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence.MultipleSequenceAlignment.update_counts_by_position_with_sequence_weights" class="md-nav__link">
    <span class="md-ellipsis">
      update_counts_by_position_with_sequence_weights
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence.MultipleSequenceAlignment.get_probabilities_from_profile" class="md-nav__link">
    <span class="md-ellipsis">
      get_probabilities_from_profile
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence.MultipleSequenceAlignment.insert" class="md-nav__link">
    <span class="md-ellipsis">
      insert
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence.MultipleSequenceAlignment.reset_state" class="md-nav__link">
    <span class="md-ellipsis">
      reset_state
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence.MultipleSequenceAlignment.pad_alignment" class="md-nav__link">
    <span class="md-ellipsis">
      pad_alignment
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.create_numeric_translation_table" class="md-nav__link">
    <span class="md-ellipsis">
      create_numeric_translation_table
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.get_numeric_translation_table" class="md-nav__link">
    <span class="md-ellipsis">
      get_numeric_translation_table
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.generate_alignment" class="md-nav__link">
    <span class="md-ellipsis">
      generate_alignment
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.read_fasta_file" class="md-nav__link">
    <span class="md-ellipsis">
      read_fasta_file
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.read_sequence_file" class="md-nav__link">
    <span class="md-ellipsis">
      read_sequence_file
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.read_alignment" class="md-nav__link">
    <span class="md-ellipsis">
      read_alignment
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.write_fasta" class="md-nav__link">
    <span class="md-ellipsis">
      write_fasta
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.write_sequence_to_fasta" class="md-nav__link">
    <span class="md-ellipsis">
      write_sequence_to_fasta
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.concatenate_fasta_files" class="md-nav__link">
    <span class="md-ellipsis">
      concatenate_fasta_files
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.write_sequences" class="md-nav__link">
    <span class="md-ellipsis">
      write_sequences
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.hhblits" class="md-nav__link">
    <span class="md-ellipsis">
      hhblits
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.optimize_protein_sequence" class="md-nav__link">
    <span class="md-ellipsis">
      optimize_protein_sequence
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.make_mutations" class="md-nav__link">
    <span class="md-ellipsis">
      make_mutations
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.find_orf_offset" class="md-nav__link">
    <span class="md-ellipsis">
      find_orf_offset
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.generate_mutations" class="md-nav__link">
    <span class="md-ellipsis">
      generate_mutations
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.pdb_to_pose_offset" class="md-nav__link">
    <span class="md-ellipsis">
      pdb_to_pose_offset
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.generate_multiple_mutations" class="md-nav__link">
    <span class="md-ellipsis">
      generate_multiple_mutations
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.generate_mutations_from_reference" class="md-nav__link">
    <span class="md-ellipsis">
      generate_mutations_from_reference
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.make_sequences_from_mutations" class="md-nav__link">
    <span class="md-ellipsis">
      make_sequences_from_mutations
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.generate_sequences" class="md-nav__link">
    <span class="md-ellipsis">
      generate_sequences
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.numeric_to_sequence" class="md-nav__link">
    <span class="md-ellipsis">
      numeric_to_sequence
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.get_equivalent_indices" class="md-nav__link">
    <span class="md-ellipsis">
      get_equivalent_indices
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.get_lod" class="md-nav__link">
    <span class="md-ellipsis">
      get_lod
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.parse_pssm" class="md-nav__link">
    <span class="md-ellipsis">
      parse_pssm
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence.parse_hhblits_pssm" class="md-nav__link">
    <span class="md-ellipsis">
      parse_hhblits_pssm
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


  <h1>sequence</h1>

<div class="doc doc-object doc-module">



<a id="sequence"></a>
  <div class="doc doc-contents first">

  

  <div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h2 id="sequence.protein_letters_alph1" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">protein_letters_alph1</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">protein_letters_alph1</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="n"><span title="str.join">join</span></span><span class="p">(</span><span class="n"><span title="typing.get_args">get_args</span></span><span class="p">(</span><span class="n"><span title="sequence.protein_letters_alph1_literal">protein_letters_alph1_literal</span></span><span class="p">))</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>ACDEFGHIKLMNPQRSTVWY</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h2 id="sequence.protein_letters_alph1_extended" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">protein_letters_alph1_extended</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">protein_letters_alph1_extended</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="n"><span title="str.join">join</span></span><span class="p">(</span><span class="n"><span title="typing.get_args">get_args</span></span><span class="p">(</span><span class="n"><span title="sequence.protein_letters_alph1_extended_literal">protein_letters_alph1_extended_literal</span></span><span class="p">))</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>ACDEFGHIKLMNPQRSTVWYBXZJUO</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h2 id="sequence.protein_letters_extended_and_gap" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">protein_letters_extended_and_gap</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">protein_letters_extended_and_gap</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="n"><span title="str.join">join</span></span><span class="p">(</span><span class="n"><span title="typing.get_args">get_args</span></span><span class="p">(</span><span class="n"><span title="sequence.protein_letters_alph1_extended_and_gap_literal">protein_letters_alph1_extended_and_gap_literal</span></span><span class="p">))</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>ACDEFGHIKLMNPQRSTVWYBXZJUO-</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h2 id="sequence.ProfileDict" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">ProfileDict</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">ProfileDict</span> <span class="o">=</span> <span class="n">dict</span><span class="p">[</span><span class="n">int</span><span class="p">,</span> <span class="n"><span title="sequence.ProfileEntry">ProfileEntry</span></span><span class="p">]</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>{1: {'A': 0.04, 'C': 0.12, ..., 'lod': {'A': -5, 'C': -9, ...},
'type': 'W', 'info': 0.00, 'weight': 0.00}, {...}}</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h2 id="sequence.MutationEntry" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">MutationEntry</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">MutationEntry</span> <span class="o">=</span> <span class="n"><span title="typing.TypedDict">TypedDict</span></span><span class="p">(</span><span class="s1">&#39;MutationEntry&#39;</span><span class="p">,</span> <span class="p">{</span><span class="s1">&#39;to&#39;</span><span class="p">:</span> <span class="n"><span title="sequence.protein_letters_literal">protein_letters_literal</span></span><span class="p">,</span> <span class="s1">&#39;from&#39;</span><span class="p">:</span> <span class="n"><span title="sequence.protein_letters_literal">protein_letters_literal</span></span><span class="p">})</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Mapping of a reference sequence amino acid type, 'to', and the resulting sequence amino acid type, 'from'.</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h2 id="sequence.mutation_dictionary" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">mutation_dictionary</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">mutation_dictionary</span> <span class="o">=</span> <span class="n">dict</span><span class="p">[</span><span class="n">int</span><span class="p">,</span> <span class="n"><a class="autorefs autorefs-internal" title="sequence.MutationEntry" href="#sequence.MutationEntry">MutationEntry</a></span><span class="p">]</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>The mapping of a residue number to a mutation entry containing the reference, 'to', and sequence, 'from', amino acid 
type.</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h2 id="sequence.sequence_dictionary" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">sequence_dictionary</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">sequence_dictionary</span> <span class="o">=</span> <span class="n">dict</span><span class="p">[</span><span class="n">int</span><span class="p">,</span> <span class="n"><span title="sequence.protein_letters_literal">protein_letters_literal</span></span><span class="p">]</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>The mapping of a residue number to the corresponding amino acid type.</p>
  </div>

</div>


<div class="doc doc-object doc-class">



<h2 id="sequence.MultipleSequenceAlignment" class="doc doc-heading">
          <span class="doc doc-object-name doc-class-name">MultipleSequenceAlignment</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">MultipleSequenceAlignment</span><span class="p">(</span><span class="n">alignment</span><span class="p">:</span> <span class="n"><span title="Bio.Align.MultipleSeqAlignment">MultipleSeqAlignment</span></span><span class="p">,</span> <span class="n">aligned_sequence</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">alphabet</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="n">protein_letters_alph1_gaped</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</code></pre></div>

  <div class="doc doc-contents ">

  



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>alignment</code></b>
            (<code><span title="Bio.Align.MultipleSeqAlignment">MultipleSeqAlignment</span></code>)
        
        <div class="doc-md-description">
          <p>A MultipleSeqAlignment object which contains an array-like object of SeqRecords</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>aligned_sequence</code></b>
            (<code>str</code>, default:
                <code>None</code>
)
        
        <div class="doc-md-description">
          <p>Provide the sequence on which the alignment is based, otherwise the first
sequence will be used from the alignment</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>alphabet</code></b>
            (<code>str</code>, default:
                <code><span title="sequence.protein_letters_alph1_gaped">protein_letters_alph1_gaped</span></code>
)
        
        <div class="doc-md-description">
          <p>'ACDEFGHIKLMNPQRSTVWY-'</p>
        </div>
      </li>
  </ul>

<details class="sets" open>
  <summary>Sets</summary>
  <p>self.alignment - (Bio.Align.MultipleSeqAlignment)
self.number_of_sequences - 214
self.query - 'MGSTHLVLK...' from aligned_sequence argument OR alignment argument, index 0
self.query_with_gaps - 'MGS--THLVLK...'
self.counts - {1: {'A': 13, 'C': 1, 'D': 23, ...}, 2: {}, ...}
self.frequencies - {1: {'A': 0.05, 'C': 0.001, 'D': 0.1, ...}, 2: {}, ...}
self.observations - {1: 210, 2:211, ...}</p>
</details>
                <details class="quote">
                  <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
                  <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1367</span>
<span class="normal">1368</span>
<span class="normal">1369</span>
<span class="normal">1370</span>
<span class="normal">1371</span>
<span class="normal">1372</span>
<span class="normal">1373</span>
<span class="normal">1374</span>
<span class="normal">1375</span>
<span class="normal">1376</span>
<span class="normal">1377</span>
<span class="normal">1378</span>
<span class="normal">1379</span>
<span class="normal">1380</span>
<span class="normal">1381</span>
<span class="normal">1382</span>
<span class="normal">1383</span>
<span class="normal">1384</span>
<span class="normal">1385</span>
<span class="normal">1386</span>
<span class="normal">1387</span>
<span class="normal">1388</span>
<span class="normal">1389</span>
<span class="normal">1390</span>
<span class="normal">1391</span>
<span class="normal">1392</span>
<span class="normal">1393</span>
<span class="normal">1394</span>
<span class="normal">1395</span>
<span class="normal">1396</span>
<span class="normal">1397</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">alignment</span><span class="p">:</span> <span class="n">MultipleSeqAlignment</span><span class="p">,</span> <span class="n">aligned_sequence</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
             <span class="n">alphabet</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">protein_letters_alph1_gaped</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Take a Biopython MultipleSeqAlignment object and process for residue specific information. One-indexed</span>

<span class="sd">    Args:</span>
<span class="sd">        alignment: A MultipleSeqAlignment object which contains an array-like object of SeqRecords</span>
<span class="sd">        aligned_sequence: Provide the sequence on which the alignment is based, otherwise the first</span>
<span class="sd">            sequence will be used from the alignment</span>
<span class="sd">        alphabet: &#39;ACDEFGHIKLMNPQRSTVWY-&#39;</span>

<span class="sd">    Sets:</span>
<span class="sd">        self.alignment - (Bio.Align.MultipleSeqAlignment)</span>
<span class="sd">        self.number_of_sequences - 214</span>
<span class="sd">        self.query - &#39;MGSTHLVLK...&#39; from aligned_sequence argument OR alignment argument, index 0</span>
<span class="sd">        self.query_with_gaps - &#39;MGS--THLVLK...&#39;</span>
<span class="sd">        self.counts - {1: {&#39;A&#39;: 13, &#39;C&#39;: 1, &#39;D&#39;: 23, ...}, 2: {}, ...}</span>
<span class="sd">        self.frequencies - {1: {&#39;A&#39;: 0.05, &#39;C&#39;: 0.001, &#39;D&#39;: 0.1, ...}, 2: {}, ...}</span>
<span class="sd">        self.observations - {1: 210, 2:211, ...}</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># count_gaps: bool = False</span>
    <span class="c1"># count_gaps: Whether gaps (-) should be counted in column weights</span>
    <span class="c1"># if alignment is None:</span>
    <span class="c1">#     raise NotImplementedError(</span>
    <span class="c1">#         f&quot;Can&#39;t create a {MultipleSequenceAlignment.__name__} with alignment=None&quot;)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">alignment</span> <span class="o">=</span> <span class="n">alignment</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">alphabet</span> <span class="o">=</span> <span class="n">alphabet</span>
    <span class="k">if</span> <span class="n">aligned_sequence</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">query_aligned</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">alignment</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">query_aligned</span> <span class="o">=</span> <span class="n">aligned_sequence</span>
</code></pre></div></td></tr></table></div>
                </details>

  

  <div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h3 id="sequence.MultipleSequenceAlignment.query" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">query</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-instance-attribute"><code>instance-attribute</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">query</span><span class="p">:</span> <span class="n">str</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>The sequence used to perform the MultipleSequenceAlignment search</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="sequence.MultipleSequenceAlignment.sequence_weights" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">sequence_weights</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
      <small class="doc doc-label doc-label-writable"><code>writable</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">sequence_weights</span><span class="p">:</span> <span class="n">list</span><span class="p">[</span><span class="n">float</span><span class="p">]</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Weights for each sequence in the alignment. Default is based on the sequence "surprise", however provided
weights can also be set</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="sequence.MultipleSequenceAlignment.alphabet_length" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">alphabet_length</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">alphabet_length</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>The number of sequence characters in the character alphabet</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="sequence.MultipleSequenceAlignment.counts_by_position" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">counts_by_position</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">counts_by_position</span><span class="p">:</span> <span class="n"><span title="numpy.ndarray">ndarray</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>The counts of each alphabet character for each residue position in the alignment with shape
(number of residues, alphabet size)</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="sequence.MultipleSequenceAlignment.gap_index" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">gap_index</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">gap_index</span><span class="p">:</span> <span class="n">int</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>The index in the alphabet where the gap character resides</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="sequence.MultipleSequenceAlignment.query_aligned" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">query_aligned</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
      <small class="doc doc-label doc-label-writable"><code>writable</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">query_aligned</span><span class="p">:</span> <span class="n">str</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>The sequence used to create the MultipleSequenceAlignment potentially containing gaps from alignment</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="sequence.MultipleSequenceAlignment.query_length" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">query_length</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">query_length</span><span class="p">:</span> <span class="n">int</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>The number of residues in the MultipleSequenceAlignment query</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="sequence.MultipleSequenceAlignment.length" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">length</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">length</span><span class="p">:</span> <span class="n">int</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>The number of sequences in the MultipleSequenceAlignment</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="sequence.MultipleSequenceAlignment.number_of_positions" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">number_of_positions</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">number_of_positions</span><span class="p">:</span> <span class="n">int</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>The number of residues plus gaps found in each sequence of the MultipleSequenceAlignment</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="sequence.MultipleSequenceAlignment.sequences" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">sequences</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">sequences</span><span class="p">:</span> <span class="n"><span title="collections.abc.Iterator">Iterator</span></span><span class="p">[</span><span class="n">str</span><span class="p">]</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Iterate over the sequences present in the alignment</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="sequence.MultipleSequenceAlignment.sequence_identifiers" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">sequence_identifiers</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">sequence_identifiers</span><span class="p">:</span> <span class="n">list</span><span class="p">[</span><span class="n">str</span><span class="p">]</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Return the identifiers associated with each sequence in the alignment</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="sequence.MultipleSequenceAlignment.alphabet_type" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">alphabet_type</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
      <small class="doc doc-label doc-label-writable"><code>writable</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">alphabet_type</span><span class="p">:</span> <span class="n"><span title="sequence.alphabet_types_literal">alphabet_types_literal</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>The type of alphabet that the alignment is mapped to numerically</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="sequence.MultipleSequenceAlignment.query_indices" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">query_indices</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">query_indices</span><span class="p">:</span> <span class="n"><span title="numpy.ndarray">ndarray</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>View the query as a boolean array (1, sequence_length) where gap positions, "-", are False</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="sequence.MultipleSequenceAlignment.sequence_indices" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">sequence_indices</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
      <small class="doc doc-label doc-label-writable"><code>writable</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">sequence_indices</span><span class="p">:</span> <span class="n"><span title="numpy.ndarray">ndarray</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>View the alignment as a boolean array (length, number_of_positions) where gap positions, "-", are
False</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="sequence.MultipleSequenceAlignment.numerical_alignment" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">numerical_alignment</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">numerical_alignment</span><span class="p">:</span> <span class="n"><span title="numpy.ndarray">ndarray</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Return the alignment as an integer array (length, number_of_positions) of the amino acid characters</p>
<p>Maps the instance .alphabet characters to their resulting sequence index</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="sequence.MultipleSequenceAlignment.array" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">array</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">array</span><span class="p">:</span> <span class="n"><span title="numpy.ndarray">ndarray</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Return the alignment as a character array (length, number_of_positions) with numpy.string_ dtype</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="sequence.MultipleSequenceAlignment.deletion_matrix" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">deletion_matrix</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">deletion_matrix</span><span class="p">:</span> <span class="n"><span title="numpy.ndarray">ndarray</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Return the number of deletions at every query aligned sequence index for each sequence in the alignment</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="sequence.MultipleSequenceAlignment.frequencies" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">frequencies</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">frequencies</span><span class="p">:</span> <span class="n"><span title="numpy.ndarray">ndarray</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Access the per-residue, alphabet frequencies with shape (number of residues, alphabet characters). Bounded
between 0 and 1</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="sequence.MultipleSequenceAlignment.observations_by_position" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">observations_by_position</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">observations_by_position</span><span class="p">:</span> <span class="n"><span title="numpy.ndarray">ndarray</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>The number of sequences with observations at each residue position in the alignment</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="sequence.MultipleSequenceAlignment.gaps_per_position" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">gaps_per_position</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">gaps_per_position</span><span class="p">:</span> <span class="n"><span title="numpy.ndarray">ndarray</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>The number of gaped letters at each position in the sequence with shape (number of residues,)</p>
  </div>

</div>




<div class="doc doc-object doc-function">



<h3 id="sequence.MultipleSequenceAlignment.from_dictionary" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">from_dictionary</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">from_dictionary</span><span class="p">(</span><span class="n">named_sequences</span><span class="p">:</span> <span class="n">dict</span><span class="p">[</span><span class="n">str</span><span class="p">,</span> <span class="n">str</span><span class="p">],</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n"><a class="autorefs autorefs-internal" title="sequence.MultipleSequenceAlignment" href="#sequence.MultipleSequenceAlignment">MultipleSequenceAlignment</a></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Create a MultipleSequenceAlignment from a dictionary of named sequences</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>named_sequences</code></b>
            (<code>dict[str, str]</code>)
        
        <div class="doc-md-description">
          <p>Where name and sequence must be a string, i.e. {'1': 'MNTEELQVAAFEI...', ...}</p>
        </div>
      </li>
  </ul>
      <p>Returns:
    The MultipleSequenceAlignment object for the provided sequences</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1345</span>
<span class="normal">1346</span>
<span class="normal">1347</span>
<span class="normal">1348</span>
<span class="normal">1349</span>
<span class="normal">1350</span>
<span class="normal">1351</span>
<span class="normal">1352</span>
<span class="normal">1353</span>
<span class="normal">1354</span>
<span class="normal">1355</span>
<span class="normal">1356</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span> <span class="nf">from_dictionary</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">named_sequences</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">MultipleSequenceAlignment</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create a MultipleSequenceAlignment from a dictionary of named sequences</span>

<span class="sd">    Args:</span>
<span class="sd">        named_sequences: Where name and sequence must be a string, i.e. {&#39;1&#39;: &#39;MNTEELQVAAFEI...&#39;, ...}</span>
<span class="sd">    Returns:</span>
<span class="sd">        The MultipleSequenceAlignment object for the provided sequences</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">cls</span><span class="p">(</span><span class="n">alignment</span><span class="o">=</span><span class="n">MultipleSeqAlignment</span><span class="p">(</span>
        <span class="p">[</span><span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="n">sequence</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="n">name</span><span class="p">)</span>  <span class="c1"># annotations={&#39;molecule_type&#39;: &#39;Protein&#39;},</span>
         <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">named_sequences</span><span class="o">.</span><span class="n">items</span><span class="p">()]),</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="sequence.MultipleSequenceAlignment.from_seq_records" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">from_seq_records</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">from_seq_records</span><span class="p">(</span><span class="n">seq_records</span><span class="p">:</span> <span class="n"><span title="collections.abc.Iterable">Iterable</span></span><span class="p">[</span><span class="n"><span title="Bio.SeqRecord.SeqRecord">SeqRecord</span></span><span class="p">],</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n"><a class="autorefs autorefs-internal" title="sequence.MultipleSequenceAlignment" href="#sequence.MultipleSequenceAlignment">MultipleSequenceAlignment</a></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Create a MultipleSequenceAlignment from a SeqRecord Iterable</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>seq_records</code></b>
            (<code><span title="collections.abc.Iterable">Iterable</span>[<span title="Bio.SeqRecord.SeqRecord">SeqRecord</span>]</code>)
        
        <div class="doc-md-description">
          <p>{name: sequence, ...} ex: {'clean_asu': 'MNTEELQVAAFEI...', ...}</p>
        </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1358</span>
<span class="normal">1359</span>
<span class="normal">1360</span>
<span class="normal">1361</span>
<span class="normal">1362</span>
<span class="normal">1363</span>
<span class="normal">1364</span>
<span class="normal">1365</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span> <span class="nf">from_seq_records</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">seq_records</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="n">SeqRecord</span><span class="p">],</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">MultipleSequenceAlignment</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create a MultipleSequenceAlignment from a SeqRecord Iterable</span>

<span class="sd">    Args:</span>
<span class="sd">        seq_records: {name: sequence, ...} ex: {&#39;clean_asu&#39;: &#39;MNTEELQVAAFEI...&#39;, ...}</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">cls</span><span class="p">(</span><span class="n">alignment</span><span class="o">=</span><span class="n">MultipleSeqAlignment</span><span class="p">(</span><span class="n">seq_records</span><span class="p">),</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="sequence.MultipleSequenceAlignment.weight_alignment_by_sequence" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">weight_alignment_by_sequence</span>


</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">weight_alignment_by_sequence</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">list</span><span class="p">[</span><span class="n">float</span><span class="p">]</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Measure diversity/surprise when comparing a single alignment entry to the rest of the alignment</p>
<p>Default means for weighting sequences. Important for creating representative sequence populations in the MSA as
was described by Heinkoff and Heinkoff, 1994 (PMID: 7966282)</p>
<p>Operation is: SUM(1 / (column_j_aa_representation * aa_ij_count))</p>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
            <code>list[float]</code>
        
        <div class="doc-md-description">
          <p>Weight of each sequence in the MSA - [2.390, 2.90, 5.33, 1.123, ...]</p>
      </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1427</span>
<span class="normal">1428</span>
<span class="normal">1429</span>
<span class="normal">1430</span>
<span class="normal">1431</span>
<span class="normal">1432</span>
<span class="normal">1433</span>
<span class="normal">1434</span>
<span class="normal">1435</span>
<span class="normal">1436</span>
<span class="normal">1437</span>
<span class="normal">1438</span>
<span class="normal">1439</span>
<span class="normal">1440</span>
<span class="normal">1441</span>
<span class="normal">1442</span>
<span class="normal">1443</span>
<span class="normal">1444</span>
<span class="normal">1445</span>
<span class="normal">1446</span>
<span class="normal">1447</span>
<span class="normal">1448</span>
<span class="normal">1449</span>
<span class="normal">1450</span>
<span class="normal">1451</span>
<span class="normal">1452</span>
<span class="normal">1453</span>
<span class="normal">1454</span>
<span class="normal">1455</span>
<span class="normal">1456</span>
<span class="normal">1457</span>
<span class="normal">1458</span>
<span class="normal">1459</span>
<span class="normal">1460</span>
<span class="normal">1461</span>
<span class="normal">1462</span>
<span class="normal">1463</span>
<span class="normal">1464</span>
<span class="normal">1465</span>
<span class="normal">1466</span>
<span class="normal">1467</span>
<span class="normal">1468</span>
<span class="normal">1469</span>
<span class="normal">1470</span>
<span class="normal">1471</span>
<span class="normal">1472</span>
<span class="normal">1473</span>
<span class="normal">1474</span>
<span class="normal">1475</span>
<span class="normal">1476</span>
<span class="normal">1477</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">weight_alignment_by_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">float</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Measure diversity/surprise when comparing a single alignment entry to the rest of the alignment</span>

<span class="sd">    Default means for weighting sequences. Important for creating representative sequence populations in the MSA as</span>
<span class="sd">    was described by Heinkoff and Heinkoff, 1994 (PMID: 7966282)</span>

<span class="sd">    Operation is: SUM(1 / (column_j_aa_representation * aa_ij_count))</span>

<span class="sd">    Returns:</span>
<span class="sd">        Weight of each sequence in the MSA - [2.390, 2.90, 5.33, 1.123, ...]</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># create a 1/ obs * counts = positional_weights</span>
    <span class="c1">#  alignment.number_of_positions - 0   1   2  ...</span>
    <span class="c1">#      / obs 0 [[32]   count seq 0 &#39;-&#39; -  2   0   0  ...   [[ 64   0   0 ...]  \</span>
    <span class="c1"># 1 / |  obs 1  [33] * count seq 1 &#39;A&#39; - 10  10   0  ... =  [330 330   0 ...]   |</span>
    <span class="c1">#      \ obs 2  [33]   count seq 2 &#39;C&#39; -  8   8   1  ...    [270 270  33 ...]] /</span>
    <span class="c1">#   ...   ...]               ...  ... ... ...</span>
    <span class="n">position_weights</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">observations_by_position</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="p">:]</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">counts_by_position</span><span class="p">)</span>
    <span class="c1"># take_along_axis from this with the transposed numerical_alignment (na) where each successive na idx</span>
    <span class="c1"># is the sequence position at the na and therefore is grabbing the position_weights by that index</span>
    <span class="c1"># finally sum along each sequence</span>
    <span class="c1"># The position_weights count seq idx must be taken by a sequence index. This happens to be on NA axis 1</span>
    <span class="c1"># at the moment so specified with .T and take using axis=0. Keeping both as axis=0 doen&#39;t index</span>
    <span class="c1"># correctly. Maybe this is a case where &#39;F&#39; array ordering is needed?</span>
    <span class="n">sequence_weights</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">take_along_axis</span><span class="p">(</span><span class="n">position_weights</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">numerical_alignment</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">critical</span><span class="p">(</span><span class="s1">&#39;New sequence_weights_&#39;</span><span class="p">,</span> <span class="n">sequence_weights</span><span class="p">)</span>

    <span class="c1"># # Old calculation</span>
    <span class="c1"># counts_ = [[0 for _ in self.alphabet] for _ in range(self.number_of_positions)]  # list[list]</span>
    <span class="c1"># for sequence in self.sequences:</span>
    <span class="c1">#     for _count, aa in zip(counts_, sequence):</span>
    <span class="c1">#         _count[numerical_translation_alph1_gaped[aa]] += 1</span>
    <span class="c1">#         # self.counts[i][aa] += 1</span>
    <span class="c1">#</span>
    <span class="c1"># self._counts = counts_</span>
    <span class="c1"># logger.critical(&#39;OLD self._counts&#39;, self._counts)</span>
    <span class="c1"># self._observations = [sum(_aa_counts[:self.gap_index]) for _aa_counts in self._counts]  # list[list]</span>

    <span class="n">observations_by_position</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">observations_by_position</span>
    <span class="n">counts_by_position</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">counts_by_position</span>
    <span class="n">numerical_alignment</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">numerical_alignment</span>
    <span class="c1"># sequence_weights_ = weight_sequences(self._counts, self.alignment, self.observations_by_position)</span>
    <span class="c1"># sequence_weights_ = weight_sequences(counts_by_position, self.alignment, observations_by_position)</span>
    <span class="n">sequence_weights_</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">sequence_idx</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">):</span>
        <span class="n">sequence_weights_</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
            <span class="p">(</span><span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="n">observations_by_position</span> <span class="o">*</span> <span class="n">counts_by_position</span><span class="p">[</span><span class="n">numerical_alignment</span><span class="p">[</span><span class="n">sequence_idx</span><span class="p">]]))</span><span class="o">.</span><span class="n">sum</span><span class="p">())</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">critical</span><span class="p">(</span><span class="s1">&#39;OLD sequence_weights_&#39;</span><span class="p">,</span> <span class="n">sequence_weights_</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">sequence_weights</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="sequence.MultipleSequenceAlignment.update_counts_by_position_with_sequence_weights" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">update_counts_by_position_with_sequence_weights</span>


</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">update_counts_by_position_with_sequence_weights</span><span class="p">()</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Overwrite the current counts with weighted counts</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1479</span>
<span class="normal">1480</span>
<span class="normal">1481</span>
<span class="normal">1482</span>
<span class="normal">1483</span>
<span class="normal">1484</span>
<span class="normal">1485</span>
<span class="normal">1486</span>
<span class="normal">1487</span>
<span class="normal">1488</span>
<span class="normal">1489</span>
<span class="normal">1490</span>
<span class="normal">1491</span>
<span class="normal">1492</span>
<span class="normal">1493</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">update_counts_by_position_with_sequence_weights</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>  <span class="c1"># UNUSED</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Overwrite the current counts with weighted counts&quot;&quot;&quot;</span>
    <span class="c1"># Add each sequence weight to the indices indicated by the numerical_alignment</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_counts_by_position</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">number_of_positions</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">alphabet_length</span><span class="p">))</span>
    <span class="n">numerical_alignment</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">numerical_alignment</span>
    <span class="n">sequence_weights</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_weights</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">sequence_idx</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_counts_by_position</span><span class="p">[:,</span> <span class="n">numerical_alignment</span><span class="p">[</span><span class="n">sequence_idx</span><span class="p">]]</span> <span class="o">+=</span> <span class="n">sequence_weights</span><span class="p">[</span><span class="n">sequence_idx</span><span class="p">]</span>
    <span class="k">except</span> <span class="ne">IndexError</span><span class="p">:</span>  <span class="c1"># sequence_weights is the wrong length</span>
        <span class="k">raise</span> <span class="ne">IndexError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Couldn&#39;t index the provided &#39;sequence_weights&#39; with length </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">sequence_weights</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;sequence_weight self.counts&#39;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">counts_by_position</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;May need to refactor weight sequences() to MultipleSequenceAlignment. Take particular care in &#39;</span>
                <span class="s1">&#39;putting the alignment back together after .insert()/.delete() &lt;- if written&#39;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="sequence.MultipleSequenceAlignment.get_probabilities_from_profile" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">get_probabilities_from_profile</span>


</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">get_probabilities_from_profile</span><span class="p">(</span><span class="n">profile</span><span class="p">:</span> <span class="n"><span title="sequence.numerical_profile">numerical_profile</span></span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n"><span title="numpy.ndarray">ndarray</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>For each sequence in the alignment, extract the values from a profile corresponding to the amino acid type
of each residue in each sequence</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>profile</code></b>
            (<code><span title="sequence.numerical_profile">numerical_profile</span></code>)
        
        <div class="doc-md-description">
          <p>A profile of values with shape (length, alphabet_length) where length is the number_of_positions</p>
        </div>
      </li>
  </ul>
      <p>Returns:
    The array with shape (length, number_of_positions) with the value for each amino acid index in profile</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1804</span>
<span class="normal">1805</span>
<span class="normal">1806</span>
<span class="normal">1807</span>
<span class="normal">1808</span>
<span class="normal">1809</span>
<span class="normal">1810</span>
<span class="normal">1811</span>
<span class="normal">1812</span>
<span class="normal">1813</span>
<span class="normal">1814</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_probabilities_from_profile</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">profile</span><span class="p">:</span> <span class="n">numerical_profile</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;For each sequence in the alignment, extract the values from a profile corresponding to the amino acid type</span>
<span class="sd">    of each residue in each sequence</span>

<span class="sd">    Args:</span>
<span class="sd">        profile: A profile of values with shape (length, alphabet_length) where length is the number_of_positions</span>
<span class="sd">    Returns:</span>
<span class="sd">        The array with shape (length, number_of_positions) with the value for each amino acid index in profile</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># transposed_alignment = self.numerical_alignment.T</span>
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">take_along_axis</span><span class="p">(</span><span class="n">profile</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">numerical_alignment</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="sequence.MultipleSequenceAlignment.insert" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">insert</span>


</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">insert</span><span class="p">(</span><span class="n">at</span><span class="p">:</span> <span class="n">int</span><span class="p">,</span> <span class="n">sequence</span><span class="p">:</span> <span class="n">str</span><span class="p">,</span> <span class="n">msa_index</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Insert new sequence in the MultipleSequenceAlignment where the added sequence is added to all columns</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>at</code></b>
            (<code>int</code>)
        
        <div class="doc-md-description">
          <p>The index to insert the sequence at. By default, the index is in reference to where self.query_indices
are True, i.e. the query sequence</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>sequence</code></b>
            (<code>str</code>)
        
        <div class="doc-md-description">
          <p>The sequence to insert. Will be inserted for every sequence of the alignment</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>msa_index</code></b>
            (<code>bool</code>, default:
                <code>False</code>
)
        
        <div class="doc-md-description">
          <p>Whether the insertion index is in the frame of the entire multiple sequence alignment.
Default, False, indicates the index is in the frame of the query sequence index, i.e. no gaps</p>
        </div>
      </li>
  </ul>

<details class="sets" open>
  <summary>Sets</summary>
  <p>self.alignment: The existing alignment updated with the new sequence in alignment form</p>
</details>
          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1820</span>
<span class="normal">1821</span>
<span class="normal">1822</span>
<span class="normal">1823</span>
<span class="normal">1824</span>
<span class="normal">1825</span>
<span class="normal">1826</span>
<span class="normal">1827</span>
<span class="normal">1828</span>
<span class="normal">1829</span>
<span class="normal">1830</span>
<span class="normal">1831</span>
<span class="normal">1832</span>
<span class="normal">1833</span>
<span class="normal">1834</span>
<span class="normal">1835</span>
<span class="normal">1836</span>
<span class="normal">1837</span>
<span class="normal">1838</span>
<span class="normal">1839</span>
<span class="normal">1840</span>
<span class="normal">1841</span>
<span class="normal">1842</span>
<span class="normal">1843</span>
<span class="normal">1844</span>
<span class="normal">1845</span>
<span class="normal">1846</span>
<span class="normal">1847</span>
<span class="normal">1848</span>
<span class="normal">1849</span>
<span class="normal">1850</span>
<span class="normal">1851</span>
<span class="normal">1852</span>
<span class="normal">1853</span>
<span class="normal">1854</span>
<span class="normal">1855</span>
<span class="normal">1856</span>
<span class="normal">1857</span>
<span class="normal">1858</span>
<span class="normal">1859</span>
<span class="normal">1860</span>
<span class="normal">1861</span>
<span class="normal">1862</span>
<span class="normal">1863</span>
<span class="normal">1864</span>
<span class="normal">1865</span>
<span class="normal">1866</span>
<span class="normal">1867</span>
<span class="normal">1868</span>
<span class="normal">1869</span>
<span class="normal">1870</span>
<span class="normal">1871</span>
<span class="normal">1872</span>
<span class="normal">1873</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">insert</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">at</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">sequence</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">msa_index</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Insert new sequence in the MultipleSequenceAlignment where the added sequence is added to all columns</span>

<span class="sd">    Args:</span>
<span class="sd">        at: The index to insert the sequence at. By default, the index is in reference to where self.query_indices</span>
<span class="sd">            are True, i.e. the query sequence</span>
<span class="sd">        sequence: The sequence to insert. Will be inserted for every sequence of the alignment</span>
<span class="sd">        msa_index: Whether the insertion index is in the frame of the entire multiple sequence alignment.</span>
<span class="sd">            Default, False, indicates the index is in the frame of the query sequence index, i.e. no gaps</span>

<span class="sd">    Sets:</span>
<span class="sd">        self.alignment: The existing alignment updated with the new sequence in alignment form</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">msa_index</span><span class="p">:</span>
        <span class="n">at</span> <span class="o">=</span> <span class="n">at</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>  <span class="c1"># To get the index &#39;at&#39; for those indices that are present in the query</span>
            <span class="n">at</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">flatnonzero</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">query_indices</span><span class="p">)[</span><span class="n">at</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">IndexError</span><span class="p">:</span>  <span class="c1"># This index is outside of query</span>
            <span class="k">if</span> <span class="n">at</span> <span class="o">&gt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">query_length</span><span class="p">:</span>
                <span class="c1"># Treat as append</span>
                <span class="n">at</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">number_of_positions</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Couldn&#39;t index with a negative index...&quot;</span><span class="p">)</span>
    <span class="n">begin_slice</span> <span class="o">=</span> <span class="nb">slice</span><span class="p">(</span><span class="n">at</span><span class="p">)</span>
    <span class="n">end_slice</span> <span class="o">=</span> <span class="nb">slice</span><span class="p">(</span><span class="n">at</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Insertion is occurring at </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s1"> index </span><span class="si">{</span><span class="n">at</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="n">new_sequence</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
    <span class="n">new_alignment</span> <span class="o">=</span> <span class="n">MultipleSeqAlignment</span><span class="p">(</span>
        <span class="p">[</span><span class="n">SeqRecord</span><span class="p">(</span><span class="n">new_sequence</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">id_</span><span class="p">)</span>  <span class="c1"># annotations={&#39;molecule_type&#39;: &#39;Protein&#39;},</span>
         <span class="k">for</span> <span class="n">id_</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_identifiers</span><span class="p">])</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;number of sequences in new_alignment: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">new_alignment</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">start_alignment_slice</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignment</span><span class="p">[:,</span> <span class="n">begin_slice</span><span class="p">]</span>
    <span class="n">start_alignment_len</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">start_alignment_slice</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">start_alignment_len</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;number of sequences in start_alignment_slice: </span><span class="si">{</span><span class="n">start_alignment_len</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="n">new_alignment</span> <span class="o">=</span> <span class="n">start_alignment_slice</span> <span class="o">+</span> <span class="n">new_alignment</span>

    <span class="n">end_alignment_slice</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignment</span><span class="p">[:,</span> <span class="n">end_slice</span><span class="p">]</span>
    <span class="n">end_alignment_len</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">end_alignment_slice</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">end_alignment_len</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;number of sequences in end_alignment_slice: </span><span class="si">{</span><span class="n">end_alignment_len</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="n">new_alignment</span> <span class="o">=</span> <span class="n">new_alignment</span> <span class="o">+</span> <span class="n">end_alignment_slice</span>

    <span class="c1"># Set the alignment</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">alignment</span> <span class="o">=</span> <span class="n">new_alignment</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">query_aligned</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">new_alignment</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>

    <span class="c1"># Update alignment dependent features</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">reset_state</span><span class="p">()</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Inserted alignment has shape (</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="si">}</span><span class="s1">, </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">number_of_positions</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="sequence.MultipleSequenceAlignment.reset_state" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">reset_state</span>


</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">reset_state</span><span class="p">()</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Remove any state attributes</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1875</span>
<span class="normal">1876</span>
<span class="normal">1877</span>
<span class="normal">1878</span>
<span class="normal">1879</span>
<span class="normal">1880</span>
<span class="normal">1881</span>
<span class="normal">1882</span>
<span class="normal">1883</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">reset_state</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Remove any state attributes&quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">attr</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;_array&#39;</span><span class="p">,</span> <span class="s1">&#39;_deletion_matrix&#39;</span><span class="p">,</span> <span class="s1">&#39;_numerical_alignment&#39;</span><span class="p">,</span> <span class="s1">&#39;_sequence_indices&#39;</span><span class="p">,</span>
                 <span class="s1">&#39;_sequence_identifiers&#39;</span><span class="p">,</span> <span class="s1">&#39;_observations_by_position&#39;</span><span class="p">,</span> <span class="s1">&#39;_counts_by_position&#39;</span><span class="p">,</span> <span class="s1">&#39;_gaps_per_position&#39;</span><span class="p">,</span>
                 <span class="s1">&#39;_frequencies&#39;</span><span class="p">]:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="fm">__delattr__</span><span class="p">(</span><span class="n">attr</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
            <span class="k">continue</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="sequence.MultipleSequenceAlignment.pad_alignment" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">pad_alignment</span>


</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">pad_alignment</span><span class="p">(</span><span class="n">length</span><span class="p">,</span> <span class="n">axis</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Extend the alignment by a set length</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>length</code></b>
        
        <div class="doc-md-description">
          <p>The length to pad the alignment</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>axis</code></b>
            (<code>int</code>, default:
                <code>0</code>
)
        
        <div class="doc-md-description">
          <p>The axis to pad. 0 pads the sequences, 1 pads the residues</p>
        </div>
      </li>
  </ul>

<details class="sets" open>
  <summary>Sets</summary>
  <p>self.alignment with the specified padding</p>
</details>
          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1885</span>
<span class="normal">1886</span>
<span class="normal">1887</span>
<span class="normal">1888</span>
<span class="normal">1889</span>
<span class="normal">1890</span>
<span class="normal">1891</span>
<span class="normal">1892</span>
<span class="normal">1893</span>
<span class="normal">1894</span>
<span class="normal">1895</span>
<span class="normal">1896</span>
<span class="normal">1897</span>
<span class="normal">1898</span>
<span class="normal">1899</span>
<span class="normal">1900</span>
<span class="normal">1901</span>
<span class="normal">1902</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">pad_alignment</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">length</span><span class="p">,</span> <span class="n">axis</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Extend the alignment by a set length</span>

<span class="sd">    Args:</span>
<span class="sd">        length: The length to pad the alignment</span>
<span class="sd">        axis: The axis to pad. 0 pads the sequences, 1 pads the residues</span>

<span class="sd">    Sets:</span>
<span class="sd">        self.alignment with the specified padding</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">axis</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">dummy_record</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s1">&#39;-&#39;</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">number_of_positions</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s1">&#39;dummy&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">alignment</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">dummy_record</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">length</span><span class="p">)])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reset_state</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>  <span class="c1"># axis == 1</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">number_of_positions</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span> <span class="o">*</span> <span class="n">length</span><span class="p">,</span> <span class="n">msa_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;padded alignment has shape (</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="si">}</span><span class="s1">, </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">number_of_positions</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>


</div>



<div class="doc doc-object doc-function">



<h2 id="sequence.create_numeric_translation_table" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">create_numeric_translation_table</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">create_numeric_translation_table</span><span class="p">(</span><span class="n">alphabet</span><span class="p">:</span> <span class="n"><span title="collections.abc.Sequence">Sequence</span></span><span class="p">[</span><span class="n">str</span><span class="p">],</span> <span class="n">bytes_</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">dict</span><span class="p">[</span><span class="n">bytes</span> <span class="o">|</span> <span class="n">str</span><span class="p">,</span> <span class="n">int</span><span class="p">]</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Return the numeric translation from an alphabet to the integer position in that alphabet</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>alphabet</code></b>
            (<code><span title="collections.abc.Sequence">Sequence</span>[str]</code>)
        
        <div class="doc-md-description">
          <p>The alphabet to use. Example 'ARNDCQEGHILKMFPSTWYVBZX*'</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>bytes_</code></b>
            (<code>bool</code>, default:
                <code>True</code>
)
        
        <div class="doc-md-description">
          <p>Whether to map from byte characters</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
            <code>dict[bytes | str, int]</code>
        
        <div class="doc-md-description">
          <p>The mapping from the character to the positional integer</p>
      </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">create_numeric_translation_table</span><span class="p">(</span><span class="n">alphabet</span><span class="p">:</span> <span class="n">Sequence</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">bytes_</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">bytes</span> <span class="o">|</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Return the numeric translation from an alphabet to the integer position in that alphabet</span>

<span class="sd">    Args:</span>
<span class="sd">        alphabet: The alphabet to use. Example &#39;ARNDCQEGHILKMFPSTWYVBZX*&#39;</span>
<span class="sd">        bytes_: Whether to map from byte characters</span>

<span class="sd">    Returns:</span>
<span class="sd">        The mapping from the character to the positional integer</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">bytes_</span><span class="p">:</span>
        <span class="n">alphabet</span> <span class="o">=</span> <span class="p">(</span><span class="n">char</span><span class="o">.</span><span class="n">encode</span><span class="p">()</span> <span class="k">for</span> <span class="n">char</span> <span class="ow">in</span> <span class="n">alphabet</span><span class="p">)</span>

    <span class="k">return</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">alphabet</span><span class="p">,</span> <span class="n">count</span><span class="p">()))</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="sequence.get_numeric_translation_table" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">get_numeric_translation_table</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">get_numeric_translation_table</span><span class="p">(</span><span class="n">alphabet_type</span><span class="p">:</span> <span class="n"><span title="sequence.alphabet_types_literal">alphabet_types_literal</span></span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n"><span title="collections.defaultdict">defaultdict</span></span><span class="p">[</span><span class="n">str</span><span class="p">,</span> <span class="n">int</span><span class="p">]</span> <span class="o">|</span> <span class="n">dict</span><span class="p">[</span><span class="n">str</span><span class="p">,</span> <span class="n">int</span><span class="p">]</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Given an amino acid alphabet type, return the corresponding numerical translation table.
If a table is passed, just return it</p>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
            <code><span title="collections.defaultdict">defaultdict</span>[str, int] | dict[str, int]</code>
        
        <div class="doc-md-description">
          <p>The integer mapping to the sequence of the requested alphabet</p>
      </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_numeric_translation_table</span><span class="p">(</span><span class="n">alphabet_type</span><span class="p">:</span> <span class="n">alphabet_types_literal</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">defaultdict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">|</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Given an amino acid alphabet type, return the corresponding numerical translation table.</span>
<span class="sd">    If a table is passed, just return it</span>

<span class="sd">    Returns:</span>
<span class="sd">        The integer mapping to the sequence of the requested alphabet</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># try:</span>
    <span class="c1">#     match alphabet_type:</span>
    <span class="c1">#         case &#39;protein_letters_alph1&#39;:</span>
    <span class="c1">#             numeric_translation_table = numerical_translation_alph1_bytes</span>
    <span class="c1">#         case &#39;protein_letters_alph3&#39;:</span>
    <span class="c1">#             numeric_translation_table = numerical_translation_alph3_bytes</span>
    <span class="c1">#         case &#39;protein_letters_alph1_gaped&#39;:</span>
    <span class="c1">#             numeric_translation_table = numerical_translation_alph1_gaped_bytes</span>
    <span class="c1">#         case &#39;protein_letters_alph3_gaped&#39;:</span>
    <span class="c1">#             numeric_translation_table = numerical_translation_alph3_gaped_bytes</span>
    <span class="c1">#         case &#39;protein_letters_alph1_unknown&#39;:</span>
    <span class="c1">#             numeric_translation_table = numerical_translation_alph1_unknown_bytes</span>
    <span class="c1">#         case &#39;protein_letters_alph3_unknown&#39;:</span>
    <span class="c1">#             numeric_translation_table = numerical_translation_alph3_unknown_bytes</span>
    <span class="c1">#         case &#39;protein_letters_alph1_unknown_gaped&#39;:</span>
    <span class="c1">#             numeric_translation_table = numerical_translation_alph1_unknown_gaped_bytes</span>
    <span class="c1">#         case &#39;protein_letters_alph3_unknown_gaped&#39;:</span>
    <span class="c1">#             numeric_translation_table = numerical_translation_alph3_unknown_gaped_bytes</span>
    <span class="c1">#         case _:</span>
    <span class="c1">#             try:  # To see if we already have the alphabet, and just return defaultdict</span>
    <span class="c1">#                 numeric_translation_table = alphabet_to_type[alphabet_type]</span>
    <span class="c1">#             except KeyError:</span>
    <span class="c1">#                 # raise ValueError(wrong_alphabet_type)</span>
    <span class="c1">#                 logger.warning(wrong_alphabet_type)</span>
    <span class="c1">#                 numeric_translation_table = create_numeric_translation_table(alphabet_type)</span>
    <span class="c1"># except SyntaxError:  # python version not 3.10</span>
    <span class="k">if</span> <span class="n">alphabet_type</span> <span class="o">==</span> <span class="s1">&#39;protein_letters_alph1&#39;</span><span class="p">:</span>
        <span class="n">numeric_translation_table</span> <span class="o">=</span> <span class="n">numerical_translation_alph1_bytes</span>
    <span class="k">elif</span> <span class="n">alphabet_type</span> <span class="o">==</span> <span class="s1">&#39;protein_letters_alph3&#39;</span><span class="p">:</span>
        <span class="n">numeric_translation_table</span> <span class="o">=</span> <span class="n">numerical_translation_alph3_bytes</span>
    <span class="k">elif</span> <span class="n">alphabet_type</span> <span class="o">==</span> <span class="s1">&#39;protein_letters_alph1_gaped&#39;</span><span class="p">:</span>
        <span class="n">numeric_translation_table</span> <span class="o">=</span> <span class="n">numerical_translation_alph1_gaped_bytes</span>
    <span class="k">elif</span> <span class="n">alphabet_type</span> <span class="o">==</span> <span class="s1">&#39;protein_letters_alph3_gaped&#39;</span><span class="p">:</span>
        <span class="n">numeric_translation_table</span> <span class="o">=</span> <span class="n">numerical_translation_alph3_gaped_bytes</span>
    <span class="k">elif</span> <span class="n">alphabet_type</span> <span class="o">==</span> <span class="s1">&#39;protein_letters_alph1_unknown&#39;</span><span class="p">:</span>
        <span class="n">numeric_translation_table</span> <span class="o">=</span> <span class="n">numerical_translation_alph1_unknown_bytes</span>
    <span class="k">elif</span> <span class="n">alphabet_type</span> <span class="o">==</span> <span class="s1">&#39;protein_letters_alph3_unknown&#39;</span><span class="p">:</span>
        <span class="n">numeric_translation_table</span> <span class="o">=</span> <span class="n">numerical_translation_alph3_unknown_bytes</span>
    <span class="k">elif</span> <span class="n">alphabet_type</span> <span class="o">==</span> <span class="s1">&#39;protein_letters_alph1_unknown_gaped&#39;</span><span class="p">:</span>
        <span class="n">numeric_translation_table</span> <span class="o">=</span> <span class="n">numerical_translation_alph1_unknown_gaped_bytes</span>
    <span class="k">elif</span> <span class="n">alphabet_type</span> <span class="o">==</span> <span class="s1">&#39;protein_letters_alph3_unknown_gaped&#39;</span><span class="p">:</span>
        <span class="n">numeric_translation_table</span> <span class="o">=</span> <span class="n">numerical_translation_alph3_unknown_gaped_bytes</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>  <span class="c1"># To see if we already have the alphabet, and return the defaultdict</span>
            <span class="n">_type</span> <span class="o">=</span> <span class="n">alphabet_to_alphabet_type</span><span class="p">[</span><span class="n">alphabet_type</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;The alphabet &#39;</span><span class="si">{</span><span class="n">alphabet_type</span><span class="si">}</span><span class="s2">&#39; isn&#39;t an allowed alphabet_type.&quot;</span>
                <span class="sa">f</span><span class="s2">&quot; See </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">alphabet_types</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="c1"># raise ValueError(wrong_alphabet_type)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">get_numeric_translation_table</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">: The alphabet_type &#39;</span><span class="si">{</span><span class="n">alphabet_type</span><span class="si">}</span><span class="s2">&#39; &quot;</span>
                       <span class="s2">&quot;isn&#39;t viable. Attempting to create it&quot;</span><span class="p">)</span>
        <span class="n">numeric_translation_table</span> <span class="o">=</span> <span class="n">create_numeric_translation_table</span><span class="p">(</span><span class="n">alphabet_type</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">numeric_translation_table</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="sequence.generate_alignment" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">generate_alignment</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">generate_alignment</span><span class="p">(</span><span class="n">seq1</span><span class="p">:</span> <span class="n"><span title="collections.abc.Sequence">Sequence</span></span><span class="p">[</span><span class="n">str</span><span class="p">],</span> <span class="n">seq2</span><span class="p">:</span> <span class="n"><span title="collections.abc.Sequence">Sequence</span></span><span class="p">[</span><span class="n">str</span><span class="p">],</span> <span class="n">matrix</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="n">default_substitution_matrix_name</span><span class="p">,</span> <span class="n">local</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">top_alignment</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">alignment_kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n"><span title="Bio.Align.Alignment">Alignment</span></span> <span class="o">|</span> <span class="n"><span title="Bio.Align.PairwiseAlignments">PairwiseAlignments</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Use Biopython's pairwise2 to generate a sequence alignment</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>seq1</code></b>
            (<code><span title="collections.abc.Sequence">Sequence</span>[str]</code>)
        
        <div class="doc-md-description">
          <p>The first sequence to align</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>seq2</code></b>
            (<code><span title="collections.abc.Sequence">Sequence</span>[str]</code>)
        
        <div class="doc-md-description">
          <p>The second sequence to align</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>matrix</code></b>
            (<code>str</code>, default:
                <code><span title="sequence.default_substitution_matrix_name">default_substitution_matrix_name</span></code>
)
        
        <div class="doc-md-description">
          <p>The matrix used to compare character similarities</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>local</code></b>
            (<code>bool</code>, default:
                <code>False</code>
)
        
        <div class="doc-md-description">
          <p>Whether to run a local alignment. Only use for generally similar sequences!</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>top_alignment</code></b>
            (<code>bool</code>, default:
                <code>True</code>
)
        
        <div class="doc-md-description">
          <p>Only include the highest scoring alignment</p>
        </div>
      </li>
  </ul>



  <p>Other Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>query_left_open_gap_score</code></b>
        
        <div class="doc-md-description">
          <p>int = 0 - The score used for opening a gap in the alignment procedure</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>query_left_extend_gap_score</code></b>
        
        <div class="doc-md-description">
          <p>int = 0 - The score used for extending a gap in the alignment procedure</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>target_left_open_gap_score</code></b>
        
        <div class="doc-md-description">
          <p>int = 0 - The score used for opening a gap in the alignment procedure</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>target_left_extend_gap_score</code></b>
        
        <div class="doc-md-description">
          <p>int = 0 - The score used for extending a gap in the alignment procedure</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>query_internal_open_gap_score</code></b>
        
        <div class="doc-md-description">
          <p>int = -12 - The score used for opening a gap in the alignment procedure</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>query_internal_extend_gap_score</code></b>
        
        <div class="doc-md-description">
          <p>int = -1 - The score used for extending a gap in the alignment procedure</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>target_internal_open_gap_score</code></b>
        
        <div class="doc-md-description">
          <p>int = -12 - The score used for opening a gap in the alignment procedure</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>target_internal_extend_gap_score</code></b>
        
        <div class="doc-md-description">
          <p>int = -1 - The score used for extending a gap in the alignment procedure</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>query_right_open_gap_score</code></b>
        
        <div class="doc-md-description">
          <p>int = 0 - The score used for opening a gap in the alignment procedure</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>query_right_extend_gap_score</code></b>
        
        <div class="doc-md-description">
          <p>int = 0 - The score used for extending a gap in the alignment procedure</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>target_right_open_gap_score</code></b>
        
        <div class="doc-md-description">
          <p>int = 0 - The score used for opening a gap in the alignment procedure</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>target_right_extend_gap_score</code></b>
        
        <div class="doc-md-description">
          <p>int = 0 - The score used for extending a gap in the alignment procedure</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
            <code><span title="Bio.Align.Alignment">Alignment</span> | <span title="Bio.Align.PairwiseAlignments">PairwiseAlignments</span></code>
        
        <div class="doc-md-description">
          <p>The resulting alignment(s). Will be an Alignment object if top_alignment is True else PairwiseAlignments object</p>
      </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">generate_alignment</span><span class="p">(</span><span class="n">seq1</span><span class="p">:</span> <span class="n">Sequence</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">seq2</span><span class="p">:</span> <span class="n">Sequence</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">matrix</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">default_substitution_matrix_name</span><span class="p">,</span>
                       <span class="n">local</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">top_alignment</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">alignment_kwargs</span><span class="p">)</span> \
        <span class="o">-&gt;</span> <span class="n">Alignment</span> <span class="o">|</span> <span class="n">PairwiseAlignments</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Use Biopython&#39;s pairwise2 to generate a sequence alignment</span>

<span class="sd">    Args:</span>
<span class="sd">        seq1: The first sequence to align</span>
<span class="sd">        seq2: The second sequence to align</span>
<span class="sd">        matrix: The matrix used to compare character similarities</span>
<span class="sd">        local: Whether to run a local alignment. Only use for generally similar sequences!</span>
<span class="sd">        top_alignment: Only include the highest scoring alignment</span>

<span class="sd">    Keyword Args:</span>
<span class="sd">        query_left_open_gap_score: int = 0 - The score used for opening a gap in the alignment procedure</span>
<span class="sd">        query_left_extend_gap_score: int = 0 - The score used for extending a gap in the alignment procedure</span>
<span class="sd">        target_left_open_gap_score: int = 0 - The score used for opening a gap in the alignment procedure</span>
<span class="sd">        target_left_extend_gap_score: int = 0 - The score used for extending a gap in the alignment procedure</span>
<span class="sd">        query_internal_open_gap_score: int = -12 - The score used for opening a gap in the alignment procedure</span>
<span class="sd">        query_internal_extend_gap_score: int = -1 - The score used for extending a gap in the alignment procedure</span>
<span class="sd">        target_internal_open_gap_score: int = -12 - The score used for opening a gap in the alignment procedure</span>
<span class="sd">        target_internal_extend_gap_score: int = -1 - The score used for extending a gap in the alignment procedure</span>
<span class="sd">        query_right_open_gap_score: int = 0 - The score used for opening a gap in the alignment procedure</span>
<span class="sd">        query_right_extend_gap_score: int = 0 - The score used for extending a gap in the alignment procedure</span>
<span class="sd">        target_right_open_gap_score: int = 0 - The score used for opening a gap in the alignment procedure</span>
<span class="sd">        target_right_extend_gap_score: int = 0 - The score used for extending a gap in the alignment procedure</span>

<span class="sd">    Returns:</span>
<span class="sd">        The resulting alignment(s). Will be an Alignment object if top_alignment is True else PairwiseAlignments object</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">matrix_</span> <span class="o">=</span> <span class="n">_substitution_matrices_cache</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">matrix</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">matrix_</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>  <span class="c1"># To get the new matrix and store for future ops</span>
            <span class="n">matrix_</span> <span class="o">=</span> <span class="n">_substitution_matrices_cache</span><span class="p">[</span><span class="n">matrix</span><span class="p">]</span> <span class="o">=</span> <span class="n">substitution_matrices</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">matrix</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">FileNotFoundError</span><span class="p">:</span>  <span class="c1"># Missing this</span>
            <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Couldn&#39;t find the substitution matrix &#39;</span><span class="si">{</span><span class="n">matrix</span><span class="si">}</span><span class="s2">&#39; &quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">local</span><span class="p">:</span>
        <span class="n">mode</span> <span class="o">=</span> <span class="s1">&#39;local&#39;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">mode</span> <span class="o">=</span> <span class="s1">&#39;global&#39;</span>

    <span class="k">if</span> <span class="n">alignment_kwargs</span><span class="p">:</span>
        <span class="n">protein_alignment_variables_</span> <span class="o">=</span> <span class="n">protein_alignment_variables</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">protein_alignment_variables_</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="o">**</span><span class="n">alignment_kwargs</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">protein_alignment_variables_</span> <span class="o">=</span> <span class="n">protein_alignment_variables</span>
    <span class="c1"># logger.debug(f&#39;Generating sequence alignment between:\n{seq1}\n\tAND:\n{seq2}&#39;)</span>
    <span class="c1"># Create sequence alignment</span>
    <span class="n">aligner</span> <span class="o">=</span> <span class="n">PairwiseAligner</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="n">mode</span><span class="p">,</span> <span class="n">substitution_matrix</span><span class="o">=</span><span class="n">matrix_</span><span class="p">,</span> <span class="o">**</span><span class="n">protein_alignment_variables_</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>  <span class="c1"># sequence contains letters not in the alphabet</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Sequence1: </span><span class="si">{</span><span class="n">seq1</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Sequence2: </span><span class="si">{</span><span class="n">seq2</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">raise</span>
    <span class="n">first_alignment</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Found alignment with score: </span><span class="si">{</span><span class="n">alignments</span><span class="o">.</span><span class="n">score</span><span class="si">}</span><span class="se">\n</span><span class="si">{</span><span class="n">first_alignment</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="c1"># print(&quot;Number of alignments: %d&quot; % len(alignments))</span>
    <span class="c1">#   Number of alignments: 1</span>
    <span class="c1"># alignment = alignments[0]</span>
    <span class="c1"># print(&quot;Score = %.1f&quot; % alignment.score)</span>
    <span class="c1">#   Score = 13.0</span>
    <span class="c1"># print(alignment)</span>
    <span class="c1">#   target            0 KEVLA 5</span>
    <span class="c1">#                     0 -|||- 5</span>
    <span class="c1">#   query             0 -EVL- 3</span>
    <span class="c1"># print(alignment.target)</span>
    <span class="c1"># print(alignment.query)</span>
    <span class="c1"># print(alignment.indices)  # returns array([[ 0,  1,  2,  3,  4], [ -1,  0,  1,  2, -1]]) where -1 are gaps</span>
    <span class="c1"># This would be useful in checking for gaps during generate_mutations()</span>
    <span class="c1"># print(alignment.inverse_indices)  # returns [array([ 0,  1,  2,  3,  4], [ 1,  2,  3]])</span>
    <span class="c1"># where -1 are outside array and each index is the position in the alignment. This would be useful for instance with</span>
    <span class="c1"># get_equivalent_indices() which is precalculated and now does this routine twice during Entity.__init__()</span>
    <span class="c1"># logger.debug(f&#39;Generated alignment:\n{pairwise2.format_alignment(*align[0])}&#39;)</span>

    <span class="c1"># return align[0] if top_alignment else align</span>
    <span class="k">return</span> <span class="n">first_alignment</span> <span class="k">if</span> <span class="n">top_alignment</span> <span class="k">else</span> <span class="n">alignments</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="sequence.read_fasta_file" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">read_fasta_file</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">read_fasta_file</span><span class="p">(</span><span class="n">file_name</span><span class="p">:</span> <span class="n"><span title="typing.AnyStr">AnyStr</span></span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n"><span title="collections.abc.Iterable">Iterable</span></span><span class="p">[</span><span class="n"><span title="Bio.SeqRecord.SeqRecord">SeqRecord</span></span><span class="p">]</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Opens a fasta file and return a parser object to load the sequences to SeqRecords</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>file_name</code></b>
            (<code><span title="typing.AnyStr">AnyStr</span></code>)
        
        <div class="doc-md-description">
          <p>The location of the file on disk</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
            <code><span title="collections.abc.Iterable">Iterable</span>[<span title="Bio.SeqRecord.SeqRecord">SeqRecord</span>]</code>
        
        <div class="doc-md-description">
          <p>An iterator of the sequences in the file [record1, record2, ...]</p>
      </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">read_fasta_file</span><span class="p">(</span><span class="n">file_name</span><span class="p">:</span> <span class="n">AnyStr</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Iterable</span><span class="p">[</span><span class="n">SeqRecord</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Opens a fasta file and return a parser object to load the sequences to SeqRecords</span>

<span class="sd">    Args:</span>
<span class="sd">        file_name: The location of the file on disk</span>

<span class="sd">    Returns:</span>
<span class="sd">        An iterator of the sequences in the file [record1, record2, ...]</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="s1">&#39;fasta&#39;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="sequence.read_sequence_file" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">read_sequence_file</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">read_sequence_file</span><span class="p">(</span><span class="n">file_name</span><span class="p">:</span> <span class="n"><span title="typing.AnyStr">AnyStr</span></span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n"><span title="collections.abc.Iterable">Iterable</span></span><span class="p">[</span><span class="n"><span title="Bio.SeqRecord.SeqRecord">SeqRecord</span></span><span class="p">]</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Opens a fasta file and return a parser object to load the sequences to SeqRecords</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>file_name</code></b>
            (<code><span title="typing.AnyStr">AnyStr</span></code>)
        
        <div class="doc-md-description">
          <p>The location of the file on disk</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
            <code><span title="collections.abc.Iterable">Iterable</span>[<span title="Bio.SeqRecord.SeqRecord">SeqRecord</span>]</code>
        
        <div class="doc-md-description">
          <p>An iterator of the sequences in the file [record1, record2, ...]</p>
      </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">read_sequence_file</span><span class="p">(</span><span class="n">file_name</span><span class="p">:</span> <span class="n">AnyStr</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Iterable</span><span class="p">[</span><span class="n">SeqRecord</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Opens a fasta file and return a parser object to load the sequences to SeqRecords</span>

<span class="sd">    Args:</span>
<span class="sd">        file_name: The location of the file on disk</span>

<span class="sd">    Returns:</span>
<span class="sd">        An iterator of the sequences in the file [record1, record2, ...]</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="s1">&#39;csv&#39;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="sequence.read_alignment" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">read_alignment</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">read_alignment</span><span class="p">(</span><span class="n">file_name</span><span class="p">:</span> <span class="n"><span title="typing.AnyStr">AnyStr</span></span><span class="p">,</span> <span class="n">alignment_type</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="s1">&#39;fasta&#39;</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n"><span title="Bio.Align.MultipleSeqAlignment">MultipleSeqAlignment</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Open an alignment file and parse the alignment to a Biopython MultipleSeqAlignment</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>file_name</code></b>
            (<code><span title="typing.AnyStr">AnyStr</span></code>)
        
        <div class="doc-md-description">
          <p>The location of the file on disk</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>alignment_type</code></b>
            (<code>str</code>, default:
                <code>&#39;fasta&#39;</code>
)
        
        <div class="doc-md-description">
          <p>The type of file that the alignment is stored in. Used for parsing</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
            <code><span title="Bio.Align.MultipleSeqAlignment">MultipleSeqAlignment</span></code>
        
        <div class="doc-md-description">
          <p>The parsed alignment</p>
      </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@utils</span><span class="o">.</span><span class="n">handle_errors</span><span class="p">(</span><span class="n">errors</span><span class="o">=</span><span class="p">(</span><span class="ne">FileNotFoundError</span><span class="p">,))</span>
<span class="k">def</span> <span class="nf">read_alignment</span><span class="p">(</span><span class="n">file_name</span><span class="p">:</span> <span class="n">AnyStr</span><span class="p">,</span> <span class="n">alignment_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;fasta&#39;</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">MultipleSeqAlignment</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Open an alignment file and parse the alignment to a Biopython MultipleSeqAlignment</span>

<span class="sd">    Args:</span>
<span class="sd">        file_name: The location of the file on disk</span>
<span class="sd">        alignment_type: The type of file that the alignment is stored in. Used for parsing</span>

<span class="sd">    Returns:</span>
<span class="sd">        The parsed alignment</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="n">alignment_type</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="sequence.write_fasta" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">write_fasta</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">write_fasta</span><span class="p">(</span><span class="n">sequence_records</span><span class="p">:</span> <span class="n"><span title="collections.abc.Iterable">Iterable</span></span><span class="p">[</span><span class="n"><span title="Bio.SeqRecord.SeqRecord">SeqRecord</span></span><span class="p">],</span> <span class="n">file_name</span><span class="p">:</span> <span class="n"><span title="typing.AnyStr">AnyStr</span></span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">out_dir</span><span class="p">:</span> <span class="n"><span title="typing.AnyStr">AnyStr</span></span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">())</span> <span class="o">-&gt;</span> <span class="n"><span title="typing.AnyStr">AnyStr</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Write an iterator of SeqRecords to a .fasta file with fasta format. '.fasta' is appended if not specified in name</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>sequence_records</code></b>
            (<code><span title="collections.abc.Iterable">Iterable</span>[<span title="Bio.SeqRecord.SeqRecord">SeqRecord</span>]</code>)
        
        <div class="doc-md-description">
          <p>The sequences to write. Should be Biopython SeqRecord format</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>file_name</code></b>
            (<code><span title="typing.AnyStr">AnyStr</span></code>, default:
                <code>None</code>
)
        
        <div class="doc-md-description">
          <p>The explicit name of the file</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>name</code></b>
            (<code>str</code>, default:
                <code>None</code>
)
        
        <div class="doc-md-description">
          <p>The name of the file to output</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>out_dir</code></b>
            (<code><span title="typing.AnyStr">AnyStr</span></code>, default:
                <code><span title="os.getcwd">getcwd</span>()</code>
)
        
        <div class="doc-md-description">
          <p>The location on disk to output file</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
            <code><span title="typing.AnyStr">AnyStr</span></code>
        
        <div class="doc-md-description">
          <p>The name of the output file</p>
      </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">write_fasta</span><span class="p">(</span><span class="n">sequence_records</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="n">SeqRecord</span><span class="p">],</span> <span class="n">file_name</span><span class="p">:</span> <span class="n">AnyStr</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                <span class="n">out_dir</span><span class="p">:</span> <span class="n">AnyStr</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">())</span> <span class="o">-&gt;</span> <span class="n">AnyStr</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Write an iterator of SeqRecords to a .fasta file with fasta format. &#39;.fasta&#39; is appended if not specified in name</span>

<span class="sd">    Args:</span>
<span class="sd">        sequence_records: The sequences to write. Should be Biopython SeqRecord format</span>
<span class="sd">        file_name: The explicit name of the file</span>
<span class="sd">        name: The name of the file to output</span>
<span class="sd">        out_dir: The location on disk to output file</span>

<span class="sd">    Returns:</span>
<span class="sd">        The name of the output file</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">file_name</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">file_name</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">out_dir</span><span class="p">,</span> <span class="n">name</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">file_name</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.fasta&#39;</span><span class="p">):</span>
            <span class="n">file_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">file_name</span><span class="si">}</span><span class="s1">.fasta&#39;</span>

    <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">sequence_records</span><span class="p">,</span> <span class="n">file_name</span><span class="p">,</span> <span class="s1">&#39;fasta&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">file_name</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="sequence.write_sequence_to_fasta" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">write_sequence_to_fasta</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">write_sequence_to_fasta</span><span class="p">(</span><span class="n">sequence</span><span class="p">:</span> <span class="n">str</span><span class="p">,</span> <span class="n">file_name</span><span class="p">:</span> <span class="n"><span title="typing.AnyStr">AnyStr</span></span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="n">str</span><span class="p">,</span> <span class="n">out_dir</span><span class="p">:</span> <span class="n"><span title="typing.AnyStr">AnyStr</span></span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">())</span> <span class="o">-&gt;</span> <span class="n"><span title="typing.AnyStr">AnyStr</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Write an iterator of SeqRecords to a .fasta file with fasta format. '.fasta' is appended if not specified in name</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>sequence</code></b>
            (<code>str</code>)
        
        <div class="doc-md-description">
          <p>The sequence to write</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>name</code></b>
            (<code>str</code>)
        
        <div class="doc-md-description">
          <p>The name of the sequence. Will be used as the default file_name base name if file_name not provided</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>file_name</code></b>
            (<code><span title="typing.AnyStr">AnyStr</span></code>)
        
        <div class="doc-md-description">
          <p>The explicit name of the file</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>out_dir</code></b>
            (<code><span title="typing.AnyStr">AnyStr</span></code>, default:
                <code><span title="os.getcwd">getcwd</span>()</code>
)
        
        <div class="doc-md-description">
          <p>The location on disk to output the file. Only used if file_name not explicitly provided</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
            <code><span title="typing.AnyStr">AnyStr</span></code>
        
        <div class="doc-md-description">
          <p>The path to the output file</p>
      </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">write_sequence_to_fasta</span><span class="p">(</span><span class="n">sequence</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">file_name</span><span class="p">:</span> <span class="n">AnyStr</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">out_dir</span><span class="p">:</span> <span class="n">AnyStr</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">())</span> <span class="o">-&gt;</span> <span class="n">AnyStr</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Write an iterator of SeqRecords to a .fasta file with fasta format. &#39;.fasta&#39; is appended if not specified in name</span>

<span class="sd">    Args:</span>
<span class="sd">        sequence: The sequence to write</span>
<span class="sd">        name: The name of the sequence. Will be used as the default file_name base name if file_name not provided</span>
<span class="sd">        file_name: The explicit name of the file</span>
<span class="sd">        out_dir: The location on disk to output the file. Only used if file_name not explicitly provided</span>

<span class="sd">    Returns:</span>
<span class="sd">        The path to the output file</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">file_name</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">file_name</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">out_dir</span><span class="p">,</span> <span class="n">name</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">file_name</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.fasta&#39;</span><span class="p">):</span>
            <span class="n">file_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">file_name</span><span class="si">}</span><span class="s1">.fasta&#39;</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
        <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;&gt;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="se">\n</span><span class="si">{</span><span class="n">sequence</span><span class="si">}</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">file_name</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="sequence.concatenate_fasta_files" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">concatenate_fasta_files</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">concatenate_fasta_files</span><span class="p">(</span><span class="n">file_names</span><span class="p">:</span> <span class="n"><span title="collections.abc.Iterable">Iterable</span></span><span class="p">[</span><span class="n"><span title="typing.AnyStr">AnyStr</span></span><span class="p">],</span> <span class="n">out_path</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="s1">&#39;concatenated_fasta&#39;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n"><span title="typing.AnyStr">AnyStr</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Take multiple fasta files and concatenate into a single file</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>file_names</code></b>
            (<code><span title="collections.abc.Iterable">Iterable</span>[<span title="typing.AnyStr">AnyStr</span>]</code>)
        
        <div class="doc-md-description">
          <p>The name of the files to concatenate</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>out_path</code></b>
            (<code>str</code>, default:
                <code>&#39;concatenated_fasta&#39;</code>
)
        
        <div class="doc-md-description">
          <p>The location on disk to output file</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
            <code><span title="typing.AnyStr">AnyStr</span></code>
        
        <div class="doc-md-description">
          <p>The name of the output file</p>
      </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">concatenate_fasta_files</span><span class="p">(</span><span class="n">file_names</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="n">AnyStr</span><span class="p">],</span> <span class="n">out_path</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;concatenated_fasta&#39;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">AnyStr</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Take multiple fasta files and concatenate into a single file</span>

<span class="sd">    Args:</span>
<span class="sd">        file_names: The name of the files to concatenate</span>
<span class="sd">        out_path: The location on disk to output file</span>

<span class="sd">    Returns:</span>
<span class="sd">        The name of the output file</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">seq_records</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">file_names</span><span class="p">:</span>
        <span class="n">seq_records</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">read_fasta_file</span><span class="p">(</span><span class="n">file</span><span class="p">)))</span>
    <span class="k">return</span> <span class="n">write_fasta</span><span class="p">(</span><span class="n">seq_records</span><span class="p">,</span> <span class="n">out_dir</span><span class="o">=</span><span class="n">out_path</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="sequence.write_sequences" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">write_sequences</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">write_sequences</span><span class="p">(</span><span class="n">sequences</span><span class="p">:</span> <span class="n"><span title="collections.abc.Sequence">Sequence</span></span> <span class="o">|</span> <span class="n">dict</span><span class="p">[</span><span class="n">str</span><span class="p">,</span> <span class="n"><span title="collections.abc.Sequence">Sequence</span></span><span class="p">],</span> <span class="n">names</span><span class="p">:</span> <span class="n"><span title="collections.abc.Sequence">Sequence</span></span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">out_path</span><span class="p">:</span> <span class="n"><span title="typing.AnyStr">AnyStr</span></span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">(),</span> <span class="n">file_name</span><span class="p">:</span> <span class="n"><span title="typing.AnyStr">AnyStr</span></span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">csv</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n"><span title="typing.AnyStr">AnyStr</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Write a fasta file from sequence(s). If a single sequence is provided, pass as a string</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>sequences</code></b>
            (<code><span title="collections.abc.Sequence">Sequence</span> | dict[str, <span title="collections.abc.Sequence">Sequence</span>]</code>)
        
        <div class="doc-md-description">
          <p>If a list, can be list of tuples(name, sequence), or list[sequence] where names contain the
corresponding sequence names. If dict, uses key as name, value as sequence. If str, treats as the sequence</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>names</code></b>
            (<code><span title="collections.abc.Sequence">Sequence</span></code>, default:
                <code>None</code>
)
        
        <div class="doc-md-description">
          <p>The name or names of the sequence record(s). If a single name, will be used as the default file_name
base name if file_name not provided. Otherwise, will be used iteratively</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>out_path</code></b>
            (<code><span title="typing.AnyStr">AnyStr</span></code>, default:
                <code><span title="os.getcwd">getcwd</span>()</code>
)
        
        <div class="doc-md-description">
          <p>The location on disk to output file</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>file_name</code></b>
            (<code><span title="typing.AnyStr">AnyStr</span></code>, default:
                <code>None</code>
)
        
        <div class="doc-md-description">
          <p>The explicit name of the file</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>csv</code></b>
            (<code>bool</code>, default:
                <code>False</code>
)
        
        <div class="doc-md-description">
          <p>Whether the file should be written as a .csv. Default is .fasta</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
            <code><span title="typing.AnyStr">AnyStr</span></code>
        
        <div class="doc-md-description">
          <p>The name of the output file</p>
      </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">write_sequences</span><span class="p">(</span><span class="n">sequences</span><span class="p">:</span> <span class="n">Sequence</span> <span class="o">|</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Sequence</span><span class="p">],</span> <span class="n">names</span><span class="p">:</span> <span class="n">Sequence</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                    <span class="n">out_path</span><span class="p">:</span> <span class="n">AnyStr</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">(),</span> <span class="n">file_name</span><span class="p">:</span> <span class="n">AnyStr</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">csv</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">AnyStr</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Write a fasta file from sequence(s). If a single sequence is provided, pass as a string</span>

<span class="sd">    Args:</span>
<span class="sd">        sequences: If a list, can be list of tuples(name, sequence), or list[sequence] where names contain the</span>
<span class="sd">            corresponding sequence names. If dict, uses key as name, value as sequence. If str, treats as the sequence</span>
<span class="sd">        names: The name or names of the sequence record(s). If a single name, will be used as the default file_name</span>
<span class="sd">            base name if file_name not provided. Otherwise, will be used iteratively</span>
<span class="sd">        out_path: The location on disk to output file</span>
<span class="sd">        file_name: The explicit name of the file</span>
<span class="sd">        csv: Whether the file should be written as a .csv. Default is .fasta</span>

<span class="sd">    Returns:</span>
<span class="sd">        The name of the output file</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">file_name</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">names</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>  <span class="c1"># Not an iterable</span>
            <span class="n">file_name</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">out_path</span><span class="p">,</span> <span class="n">names</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Must provide argument &#39;file_name&#39; or &#39;names&#39; as a str to </span><span class="si">{</span><span class="n">write_sequences</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">()&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">csv</span><span class="p">:</span>
        <span class="n">start</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;,&#39;</span>
        <span class="n">extension</span> <span class="o">=</span> <span class="s1">&#39;.csv&#39;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">start</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;&gt;&#39;</span><span class="p">,</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span>
        <span class="n">extension</span> <span class="o">=</span> <span class="s1">&#39;.fasta&#39;</span>

    <span class="n">provided_extension</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">file_name</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">file_name</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="n">extension</span><span class="p">)</span> <span class="ow">or</span> <span class="n">provided_extension</span> <span class="o">!=</span> <span class="n">extension</span><span class="p">:</span>
        <span class="n">file_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">file_name</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="si">}{</span><span class="n">extension</span><span class="si">}</span><span class="s1">&#39;</span>

    <span class="k">def</span> <span class="nf">data_dump</span><span class="p">():</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">write_sequences</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2"> can&#39;t parse data to make fasta</span><span class="se">\n</span><span class="s2">&quot;</span> \
               <span class="sa">f</span><span class="s1">&#39;names=</span><span class="si">{</span><span class="n">names</span><span class="si">}</span><span class="s1">, sequences=</span><span class="si">{</span><span class="n">sequences</span><span class="si">}</span><span class="s1">, extension=</span><span class="si">{</span><span class="n">extension</span><span class="si">}</span><span class="s1">, out_path=</span><span class="si">{</span><span class="n">out_path</span><span class="si">}</span><span class="s1">, &#39;</span> \
               <span class="sa">f</span><span class="s1">&#39;file_name=</span><span class="si">{</span><span class="n">file_name</span><span class="si">}</span><span class="s1">&#39;</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">sequences</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">):</span>
            <span class="n">sequences</span> <span class="o">=</span> <span class="n">sequences</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">sequences</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No sequences provided, couldn&#39;t write anything&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">sequences</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">tuple</span><span class="p">):</span>  <span class="c1"># Where seq[0] is name, seq[1] is seq</span>
                <span class="n">formatted_sequence_gen</span> <span class="o">=</span> <span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">start</span><span class="si">}{</span><span class="n">name</span><span class="si">}{</span><span class="n">sep</span><span class="si">}{</span><span class="n">seq</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">seq</span><span class="p">,</span> <span class="o">*</span><span class="n">_</span> <span class="ow">in</span> <span class="n">sequences</span><span class="p">)</span>
            <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">names</span><span class="p">,</span> <span class="n">Sequence</span><span class="p">):</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">str</span><span class="p">):</span>
                    <span class="n">formatted_sequence_gen</span> <span class="o">=</span> <span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">start</span><span class="si">}{</span><span class="n">name</span><span class="si">}{</span><span class="n">sep</span><span class="si">}{</span><span class="n">seq</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">names</span><span class="p">,</span> <span class="n">sequences</span><span class="p">))</span>
                <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">Sequence</span><span class="p">):</span>
                    <span class="n">formatted_sequence_gen</span> <span class="o">=</span> \
                        <span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">start</span><span class="si">}{</span><span class="n">name</span><span class="si">}{</span><span class="n">sep</span><span class="si">}{</span><span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">names</span><span class="p">,</span> <span class="n">sequences</span><span class="p">))</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="n">data_dump</span><span class="p">())</span>
            <span class="c1"># elif isinstance(sequences[0], list):  # Where interior list is alphabet (AA or DNA)</span>
            <span class="c1">#     for idx, seq in enumerate(sequences):</span>
            <span class="c1">#         outfile.write(f&#39;&gt;{name}_{idx}\n&#39;)  # Write header</span>
            <span class="c1">#         # Check if alphabet is 3 letter protein</span>
            <span class="c1">#         outfile.write(f&#39;{&quot; &quot;.join(seq)}\n&#39; if len(seq[0]) == 3 else f&#39;{&quot;&quot;.join(seq)}\n&#39;)</span>
            <span class="c1"># elif isinstance(sequences[0], str):  # Likely 3 aa format...</span>
            <span class="c1">#     outfile.write(f&#39;&gt;{name}\n{&quot; &quot;.join(sequences)}\n&#39;)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="n">data_dump</span><span class="p">())</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">sequences</span><span class="p">,</span> <span class="nb">dict</span><span class="p">):</span>
            <span class="n">formatted_sequence_gen</span> <span class="o">=</span> <span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">start</span><span class="si">}{</span><span class="n">name</span><span class="si">}{</span><span class="n">sep</span><span class="si">}{</span><span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">sequences</span><span class="o">.</span><span class="n">items</span><span class="p">())</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">sequences</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">):</span>  <span class="c1"># Where seq[0] is name, seq[1] is seq</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">name</span><span class="p">,</span> <span class="n">seq</span><span class="p">,</span> <span class="o">*</span><span class="n">_</span> <span class="o">=</span> <span class="n">sequences</span>
            <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;When using a tuple, expected that the tuple contain (name, sequence) pairs&quot;</span><span class="p">)</span>
            <span class="n">formatted_sequence_gen</span> <span class="o">=</span> <span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">start</span><span class="si">}{</span><span class="n">name</span><span class="si">}{</span><span class="n">sep</span><span class="si">}{</span><span class="n">seq</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,)</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">names</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>  <span class="c1"># Assume sequences is a str or tuple</span>
            <span class="n">formatted_sequence_gen</span> <span class="o">=</span> <span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">start</span><span class="si">}{</span><span class="n">names</span><span class="si">}{</span><span class="n">sep</span><span class="si">}{</span><span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sequences</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="n">data_dump</span><span class="p">())</span>
        <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">%s</span><span class="se">\n</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">formatted_sequence_gen</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">file_name</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="sequence.hhblits" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">hhblits</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">hhblits</span><span class="p">(</span><span class="n">name</span><span class="p">:</span> <span class="n">str</span><span class="p">,</span> <span class="n">sequence_file</span><span class="p">:</span> <span class="n"><span title="collections.abc.Sequence">Sequence</span></span><span class="p">[</span><span class="n">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">sequence</span><span class="p">:</span> <span class="n"><span title="collections.abc.Sequence">Sequence</span></span><span class="p">[</span><span class="n">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">out_dir</span><span class="p">:</span> <span class="n"><span title="typing.AnyStr">AnyStr</span></span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">(),</span> <span class="n">threads</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="n">hhblits_threads</span><span class="p">,</span> <span class="n">return_command</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">list</span><span class="p">[</span><span class="n">str</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Generate a position specific scoring matrix from HHblits using Hidden Markov Models</p>
<p>When the command is run, it is possible to create six files in out_dir with the pattern '/outdir/name.<em>'
where the '</em>' extensions are:
 hhblits profile - .hmm</p>
<p>hhblits resulting cluster description - .hrr</p>
<p>sequence alignment in a3m format - .a3m</p>
<p>sequence file - .seq (if sequence_file)</p>
<p>sequence alignment in stockholm format - .sto (if return_command is False)</p>
<p>sequence alignment in fasta format - .fasta (if return_command is False)</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>name</code></b>
            (<code>str</code>)
        
        <div class="doc-md-description">
          <p>The name associated with the sequence</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>sequence_file</code></b>
            (<code><span title="collections.abc.Sequence">Sequence</span>[str]</code>, default:
                <code>None</code>
)
        
        <div class="doc-md-description">
          <p>The file containing the sequence to use</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>sequence</code></b>
            (<code><span title="collections.abc.Sequence">Sequence</span>[str]</code>, default:
                <code>None</code>
)
        
        <div class="doc-md-description">
          <p>The sequence to use. Used in place of sequence_file</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>out_dir</code></b>
            (<code><span title="typing.AnyStr">AnyStr</span></code>, default:
                <code><span title="os.getcwd">getcwd</span>()</code>
)
        
        <div class="doc-md-description">
          <p>Disk location where generated files should be written</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>threads</code></b>
            (<code>int</code>, default:
                <code><span title="sequence.hhblits_threads">hhblits_threads</span></code>
)
        
        <div class="doc-md-description">
          <p>Number of cpu's to use for the process</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>return_command</code></b>
            (<code>bool</code>, default:
                <code>False</code>
)
        
        <div class="doc-md-description">
          <p>Whether to simply return the hhblits command</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
            <code>list[str] | None</code>
        
        <div class="doc-md-description">
          <p>The command if return_command is True, otherwise None</p>
      </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">hhblits</span><span class="p">(</span><span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">sequence_file</span><span class="p">:</span> <span class="n">Sequence</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">sequence</span><span class="p">:</span> <span class="n">Sequence</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">out_dir</span><span class="p">:</span> <span class="n">AnyStr</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">(),</span> <span class="n">threads</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="n">hhblits_threads</span><span class="p">,</span>
            <span class="n">return_command</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Generate a position specific scoring matrix from HHblits using Hidden Markov Models</span>

<span class="sd">    When the command is run, it is possible to create six files in out_dir with the pattern &#39;/outdir/name.*&#39;</span>
<span class="sd">    where the &#39;*&#39; extensions are:</span>
<span class="sd">     hhblits profile - .hmm</span>

<span class="sd">     hhblits resulting cluster description - .hrr</span>

<span class="sd">     sequence alignment in a3m format - .a3m</span>

<span class="sd">     sequence file - .seq (if sequence_file)</span>

<span class="sd">     sequence alignment in stockholm format - .sto (if return_command is False)</span>

<span class="sd">     sequence alignment in fasta format - .fasta (if return_command is False)</span>

<span class="sd">    Args:</span>
<span class="sd">        name: The name associated with the sequence</span>
<span class="sd">        sequence_file: The file containing the sequence to use</span>
<span class="sd">        sequence: The sequence to use. Used in place of sequence_file</span>
<span class="sd">        out_dir: Disk location where generated files should be written</span>
<span class="sd">        threads: Number of cpu&#39;s to use for the process</span>
<span class="sd">        return_command: Whether to simply return the hhblits command</span>

<span class="sd">    Raises:</span>
<span class="sd">        RuntimeError if hhblits command is run and returns a non-zero exit code</span>

<span class="sd">    Returns:</span>
<span class="sd">        The command if return_command is True, otherwise None</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">sequence_file</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">sequence</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">hhblits</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">: Can&#39;t proceed without argument &#39;sequence_file&#39; or &#39;sequence&#39;&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">sequence_file</span> <span class="o">=</span> <span class="n">write_sequences</span><span class="p">((</span><span class="n">name</span><span class="p">,</span> <span class="n">sequence</span><span class="p">),</span> <span class="n">file_name</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">out_dir</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s1">.seq&#39;</span><span class="p">))</span>
    <span class="n">pssm_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">out_dir</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s1">.hmm&#39;</span><span class="p">)</span>
    <span class="n">a3m_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">out_dir</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s1">.a3m&#39;</span><span class="p">)</span>
    <span class="c1"># Todo for higher performance set up https://www.howtoforge.com/storing-files-directories-in-memory-with-tmpfs</span>
    <span class="c1">#  Create a ramdisk to store a database chunk to make hhblits/Jackhmmer run fast.</span>
    <span class="c1">#  sudo mkdir -m 777 --parents /tmp/ramdisk</span>
    <span class="c1">#  sudo mount -t tmpfs -o size=9G ramdisk /tmp/ramdisk</span>
    <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span><span class="n">putils</span><span class="o">.</span><span class="n">hhblits_exe</span><span class="p">,</span> <span class="s1">&#39;-d&#39;</span><span class="p">,</span> <span class="n">putils</span><span class="o">.</span><span class="n">uniclust_db</span><span class="p">,</span> <span class="s1">&#39;-i&#39;</span><span class="p">,</span> <span class="n">sequence_file</span><span class="p">,</span> <span class="s1">&#39;-ohhm&#39;</span><span class="p">,</span> <span class="n">pssm_file</span><span class="p">,</span> <span class="s1">&#39;-oa3m&#39;</span><span class="p">,</span> <span class="n">a3m_file</span><span class="p">,</span>
           <span class="s1">&#39;-hide_cons&#39;</span><span class="p">,</span> <span class="s1">&#39;-hide_pred&#39;</span><span class="p">,</span> <span class="s1">&#39;-hide_dssp&#39;</span><span class="p">,</span> <span class="s1">&#39;-E&#39;</span><span class="p">,</span> <span class="s1">&#39;1E-06&#39;</span><span class="p">,</span> <span class="s1">&#39;-v&#39;</span><span class="p">,</span> <span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="s1">&#39;-cpu&#39;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">threads</span><span class="p">)]</span>

    <span class="k">if</span> <span class="n">return_command</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">cmd</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s1"> Profile Command: </span><span class="si">{</span><span class="n">subprocess</span><span class="o">.</span><span class="n">list2cmdline</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">p</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
    <span class="n">p</span><span class="o">.</span><span class="n">communicate</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">p</span><span class="o">.</span><span class="n">returncode</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1"># temp_file = os.path.join(out_path, f&#39;{self.name}.hold&#39;)</span>
        <span class="n">temp_file</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">out_dir</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s1">.hold&#39;</span><span class="p">)</span>
        <span class="n">temp_file</span><span class="o">.</span><span class="n">unlink</span><span class="p">(</span><span class="n">missing_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="c1"># if os.path.exists(temp_file):  # remove hold file blocking progress</span>
        <span class="c1">#     os.remove(temp_file)</span>
        <span class="k">raise</span> <span class="n">utils</span><span class="o">.</span><span class="n">SymDesignException</span><span class="p">(</span>
            <span class="sa">f</span><span class="s1">&#39;Profile generation for </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s1"> got stuck. Found return code </span><span class="si">{</span><span class="n">p</span><span class="o">.</span><span class="n">returncode</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>  <span class="c1">#</span>

    <span class="c1"># Preferred alignment type</span>
    <span class="n">msa_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">out_dir</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s1">.sto&#39;</span><span class="p">)</span>
    <span class="n">p</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">([</span><span class="n">putils</span><span class="o">.</span><span class="n">reformat_msa_exe_path</span><span class="p">,</span> <span class="n">a3m_file</span><span class="p">,</span> <span class="n">msa_file</span><span class="p">,</span> <span class="s1">&#39;-num&#39;</span><span class="p">,</span> <span class="s1">&#39;-uc&#39;</span><span class="p">])</span>
    <span class="n">p</span><span class="o">.</span><span class="n">communicate</span><span class="p">()</span>
    <span class="n">fasta_msa</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">out_dir</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s1">.fasta&#39;</span><span class="p">)</span>
    <span class="n">p</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">([</span><span class="n">putils</span><span class="o">.</span><span class="n">reformat_msa_exe_path</span><span class="p">,</span> <span class="n">a3m_file</span><span class="p">,</span> <span class="n">fasta_msa</span><span class="p">,</span> <span class="s1">&#39;-M&#39;</span><span class="p">,</span> <span class="s1">&#39;first&#39;</span><span class="p">,</span> <span class="s1">&#39;-r&#39;</span><span class="p">])</span>
    <span class="n">p</span><span class="o">.</span><span class="n">communicate</span><span class="p">()</span>

    <span class="k">return</span> <span class="kc">None</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="sequence.optimize_protein_sequence" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">optimize_protein_sequence</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">optimize_protein_sequence</span><span class="p">(</span><span class="n">sequence</span><span class="p">:</span> <span class="n">str</span><span class="p">,</span> <span class="n">species</span><span class="p">:</span> <span class="n"><span title="sequence.optimization_species_literal">optimization_species_literal</span></span> <span class="o">=</span> <span class="s1">&#39;e_coli&#39;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">str</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Optimize a sequence for expression in a desired organism</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>sequence</code></b>
            (<code>str</code>)
        
        <div class="doc-md-description">
          <p>The sequence of interest</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>species</code></b>
            (<code><span title="sequence.optimization_species_literal">optimization_species_literal</span></code>, default:
                <code>&#39;e_coli&#39;</code>
)
        
        <div class="doc-md-description">
          <p>The species context to optimize nucleotide sequence usage</p>
        </div>
      </li>
  </ul>
      <p>Returns:
    The input sequence optimized to nucleotides for expression considerations</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">optimize_protein_sequence</span><span class="p">(</span><span class="n">sequence</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">species</span><span class="p">:</span> <span class="n">optimization_species_literal</span> <span class="o">=</span> <span class="s1">&#39;e_coli&#39;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Optimize a sequence for expression in a desired organism</span>

<span class="sd">    Args:</span>
<span class="sd">        sequence: The sequence of interest</span>
<span class="sd">        species: The species context to optimize nucleotide sequence usage</span>
<span class="sd">    Returns:</span>
<span class="sd">        The input sequence optimized to nucleotides for expression considerations</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">seq_length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
    <span class="n">species</span> <span class="o">=</span> <span class="n">species</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="kn">from</span> <span class="nn">dnachisel</span> <span class="kn">import</span> <span class="n">reverse_translate</span><span class="p">,</span> <span class="n">AvoidHairpins</span><span class="p">,</span> <span class="n">AvoidPattern</span><span class="p">,</span> <span class="n">AvoidRareCodons</span><span class="p">,</span> <span class="n">CodonOptimize</span><span class="p">,</span> \
        <span class="n">DnaOptimizationProblem</span><span class="p">,</span> <span class="n">EnforceGCContent</span><span class="p">,</span> <span class="n">EnforceTranslation</span><span class="p">,</span>  <span class="n">UniquifyAllKmers</span>
    <span class="k">except</span> <span class="ne">ModuleNotFoundError</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Can&#39;t </span><span class="si">{</span><span class="n">optimize_protein_sequence</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2"> as the dependency `DnaChisel` isn`t available&quot;</span><span class="p">)</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">dna_sequence</span> <span class="o">=</span> <span class="n">reverse_translate</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">error</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s1">&#39;Warning an invalid character was found in the protein sequence: </span><span class="si">{</span><span class="n">error</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="n">problem</span> <span class="o">=</span> <span class="n">DnaOptimizationProblem</span><span class="p">(</span>
        <span class="n">sequence</span><span class="o">=</span><span class="n">dna_sequence</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>  <span class="c1"># max_random_iters=20000,</span>
        <span class="n">objectives</span><span class="o">=</span><span class="p">[</span><span class="n">CodonOptimize</span><span class="p">(</span><span class="n">species</span><span class="o">=</span><span class="n">species</span><span class="p">)],</span>
        <span class="c1"># method=&#39;harmonize_rca&#39;)] &lt;- Useful for folding speed when original organism known</span>
        <span class="n">constraints</span><span class="o">=</span><span class="p">[</span><span class="n">EnforceGCContent</span><span class="p">(</span><span class="n">mini</span><span class="o">=</span><span class="mf">0.25</span><span class="p">,</span> <span class="n">maxi</span><span class="o">=</span><span class="mf">0.65</span><span class="p">),</span>  <span class="c1"># Twist required</span>
                     <span class="n">EnforceGCContent</span><span class="p">(</span><span class="n">mini</span><span class="o">=</span><span class="mf">0.35</span><span class="p">,</span> <span class="n">maxi</span><span class="o">=</span><span class="mf">0.65</span><span class="p">,</span> <span class="n">window</span><span class="o">=</span><span class="mi">50</span><span class="p">),</span>  <span class="c1"># Twist required</span>
                     <span class="n">AvoidHairpins</span><span class="p">(</span><span class="n">stem_size</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">hairpin_window</span><span class="o">=</span><span class="mi">48</span><span class="p">),</span>  <span class="c1"># Efficient translate</span>
                     <span class="n">AvoidPattern</span><span class="p">(</span><span class="s1">&#39;GGAGG&#39;</span><span class="p">,</span> <span class="n">location</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">seq_length</span><span class="p">,</span> <span class="mi">1</span><span class="p">)),</span>  <span class="c1"># Ribosome bind</span>
                     <span class="n">AvoidPattern</span><span class="p">(</span><span class="s1">&#39;TAAGGAG&#39;</span><span class="p">,</span> <span class="n">location</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">seq_length</span><span class="p">,</span> <span class="mi">1</span><span class="p">)),</span>  <span class="c1"># Ribosome bind</span>
                     <span class="n">AvoidPattern</span><span class="p">(</span><span class="s1">&#39;AAAAA&#39;</span><span class="p">,</span> <span class="n">location</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">seq_length</span><span class="p">,</span> <span class="mi">0</span><span class="p">)),</span>  <span class="c1"># Terminator</span>
                     <span class="c1"># AvoidPattern(&#39;TTTTT&#39;, location=(1, seq_length, 1)),  # Terminator</span>
                     <span class="n">AvoidPattern</span><span class="p">(</span><span class="s1">&#39;GGGGGGGGGG&#39;</span><span class="p">,</span> <span class="n">location</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">seq_length</span><span class="p">,</span> <span class="mi">0</span><span class="p">)),</span>  <span class="c1"># Homopoly</span>
                     <span class="c1"># AvoidPattern(&#39;CCCCCCCCCC&#39;, location=(1, seq_length)),  # Homopoly</span>
                     <span class="n">UniquifyAllKmers</span><span class="p">(</span><span class="mi">20</span><span class="p">),</span>  <span class="c1"># Twist required</span>
                     <span class="n">AvoidRareCodons</span><span class="p">(</span><span class="mf">0.08</span><span class="p">,</span> <span class="n">species</span><span class="o">=</span><span class="n">species</span><span class="p">),</span>
                     <span class="n">EnforceTranslation</span><span class="p">(),</span>
                     <span class="c1"># EnforceMeltingTemperature(mini=10,maxi=62,location=(1, seq_length)),</span>
                     <span class="p">])</span>

    <span class="c1"># Solve constraints and solve in regard to the objective</span>
    <span class="n">problem</span><span class="o">.</span><span class="n">max_random_iters</span> <span class="o">=</span> <span class="mi">20000</span>
    <span class="n">problem</span><span class="o">.</span><span class="n">resolve_constraints</span><span class="p">()</span>
    <span class="n">problem</span><span class="o">.</span><span class="n">optimize</span><span class="p">()</span>

    <span class="c1"># Display summaries of constraints that pass</span>
    <span class="c1"># print(problem.constraints_text_summary())</span>
    <span class="c1"># print(problem.objectives_text_summary())</span>

    <span class="c1"># Get final sequence as string</span>
    <span class="n">final_sequence</span> <span class="o">=</span> <span class="n">problem</span><span class="o">.</span><span class="n">sequence</span>
    <span class="c1"># Get final sequene as BioPython record</span>
    <span class="c1"># final_record = problem.to_record(with_sequence_edits=True)</span>

    <span class="k">return</span> <span class="n">final_sequence</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="sequence.make_mutations" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">make_mutations</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">make_mutations</span><span class="p">(</span><span class="n">sequence</span><span class="p">:</span> <span class="n"><span title="collections.abc.Sequence">Sequence</span></span><span class="p">,</span> <span class="n">mutations</span><span class="p">:</span> <span class="n">dict</span><span class="p">[</span><span class="n">int</span><span class="p">,</span> <span class="n">dict</span><span class="p">[</span><span class="n">str</span><span class="p">,</span> <span class="n">str</span><span class="p">]],</span> <span class="n">find_orf</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">str</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Modify a sequence to contain mutations specified by a mutation dictionary</p>
<p>Assumes a zero-index sequence and zero-index mutations
Args:
    sequence: 'Wild-type' sequence to mutate
    mutations: {mutation_index: {'from': AA, 'to': AA}, ...}
    find_orf: Whether to find the correct ORF for the mutations and the seq
Returns:
    seq: The mutated sequence</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">make_mutations</span><span class="p">(</span><span class="n">sequence</span><span class="p">:</span> <span class="n">Sequence</span><span class="p">,</span> <span class="n">mutations</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]],</span> <span class="n">find_orf</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Modify a sequence to contain mutations specified by a mutation dictionary</span>

<span class="sd">    Assumes a zero-index sequence and zero-index mutations</span>
<span class="sd">    Args:</span>
<span class="sd">        sequence: &#39;Wild-type&#39; sequence to mutate</span>
<span class="sd">        mutations: {mutation_index: {&#39;from&#39;: AA, &#39;to&#39;: AA}, ...}</span>
<span class="sd">        find_orf: Whether to find the correct ORF for the mutations and the seq</span>
<span class="sd">    Returns:</span>
<span class="sd">        seq: The mutated sequence</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Seq can be either list or string</span>
    <span class="k">if</span> <span class="n">find_orf</span><span class="p">:</span>
        <span class="n">offset</span> <span class="o">=</span> <span class="n">find_orf_offset</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">mutations</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Found ORF. Offset = </span><span class="si">{</span><span class="n">offset</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">offset</span> <span class="o">=</span> <span class="mi">0</span>  <span class="c1"># ZERO_OFFSET</span>

    <span class="n">seq</span> <span class="o">=</span> <span class="n">sequence</span>
    <span class="n">index_errors</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">mutations</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">index</span> <span class="o">=</span> <span class="n">key</span> <span class="o">+</span> <span class="n">offset</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">seq</span><span class="p">[</span><span class="n">index</span><span class="p">]</span> <span class="o">==</span> <span class="n">mutation</span><span class="p">[</span><span class="s1">&#39;from&#39;</span><span class="p">]:</span>  <span class="c1"># Adjust key for zero index slicing</span>
                <span class="n">seq</span> <span class="o">=</span> <span class="n">seq</span><span class="p">[:</span><span class="n">index</span><span class="p">]</span> <span class="o">+</span> <span class="n">mutation</span><span class="p">[</span><span class="s1">&#39;to&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">seq</span><span class="p">[</span><span class="n">index</span> <span class="o">+</span> <span class="mi">1</span><span class="p">:]</span>
            <span class="k">else</span><span class="p">:</span>  <span class="c1"># Find correct offset, or mark mutation source as doomed</span>
                <span class="n">index_errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">key</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">IndexError</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">key</span> <span class="o">+</span> <span class="n">offset</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">index_errors</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">make_mutations</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s1"> index errors: </span><span class="si">{</span><span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span><span class="w"> </span><span class="n">index_errors</span><span class="p">))</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">seq</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="sequence.find_orf_offset" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">find_orf_offset</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">find_orf_offset</span><span class="p">(</span><span class="n">sequence</span><span class="p">:</span> <span class="n"><span title="collections.abc.Sequence">Sequence</span></span><span class="p">,</span> <span class="n">mutations</span><span class="p">:</span> <span class="n"><a class="autorefs autorefs-internal" title="sequence.mutation_dictionary" href="#sequence.mutation_dictionary">mutation_dictionary</a></span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">int</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Using a sequence and mutation data, find the open reading frame that matches mutations closest</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>sequence</code></b>
            (<code><span title="collections.abc.Sequence">Sequence</span></code>)
        
        <div class="doc-md-description">
          <p>Sequence to search for ORF in 1 letter format</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>mutations</code></b>
            (<code><a class="autorefs autorefs-internal" title="sequence.mutation_dictionary" href="#sequence.mutation_dictionary">mutation_dictionary</a></code>)
        
        <div class="doc-md-description">
          <p>{mutation_index: {'from': AA, 'to': AA}, ...} zero-indexed sequence dictionary</p>
        </div>
      </li>
  </ul>
      <p>Returns:
    The zero-indexed integer to offset the provided sequence to best match the provided mutations</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span>
<span class="normal">739</span>
<span class="normal">740</span>
<span class="normal">741</span>
<span class="normal">742</span>
<span class="normal">743</span>
<span class="normal">744</span>
<span class="normal">745</span>
<span class="normal">746</span>
<span class="normal">747</span>
<span class="normal">748</span>
<span class="normal">749</span>
<span class="normal">750</span>
<span class="normal">751</span>
<span class="normal">752</span>
<span class="normal">753</span>
<span class="normal">754</span>
<span class="normal">755</span>
<span class="normal">756</span>
<span class="normal">757</span>
<span class="normal">758</span>
<span class="normal">759</span>
<span class="normal">760</span>
<span class="normal">761</span>
<span class="normal">762</span>
<span class="normal">763</span>
<span class="normal">764</span>
<span class="normal">765</span>
<span class="normal">766</span>
<span class="normal">767</span>
<span class="normal">768</span>
<span class="normal">769</span>
<span class="normal">770</span>
<span class="normal">771</span>
<span class="normal">772</span>
<span class="normal">773</span>
<span class="normal">774</span>
<span class="normal">775</span>
<span class="normal">776</span>
<span class="normal">777</span>
<span class="normal">778</span>
<span class="normal">779</span>
<span class="normal">780</span>
<span class="normal">781</span>
<span class="normal">782</span>
<span class="normal">783</span>
<span class="normal">784</span>
<span class="normal">785</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">find_orf_offset</span><span class="p">(</span><span class="n">sequence</span><span class="p">:</span> <span class="n">Sequence</span><span class="p">,</span> <span class="n">mutations</span><span class="p">:</span> <span class="n">mutation_dictionary</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Using a sequence and mutation data, find the open reading frame that matches mutations closest</span>

<span class="sd">    Args:</span>
<span class="sd">        sequence: Sequence to search for ORF in 1 letter format</span>
<span class="sd">        mutations: {mutation_index: {&#39;from&#39;: AA, &#39;to&#39;: AA}, ...} zero-indexed sequence dictionary</span>
<span class="sd">    Returns:</span>
<span class="sd">        The zero-indexed integer to offset the provided sequence to best match the provided mutations</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">gen_offset_repr</span><span class="p">():</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s1">&#39;Found the orf_offsets: </span><span class="si">{</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s2">=</span><span class="si">{</span><span class="n">v</span><span class="si">}</span><span class="s2">&quot;</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">k</span><span class="p">,</span><span class="w"> </span><span class="n">v</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">orf_offsets</span><span class="o">.</span><span class="n">items</span><span class="p">())</span><span class="si">}</span><span class="s1">&#39;</span>

    <span class="n">unsolvable</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="n">orf_start_idx</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">orf_offsets</span> <span class="o">=</span> <span class="p">{</span><span class="n">idx</span><span class="p">:</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">aa</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span> <span class="k">if</span> <span class="n">aa</span> <span class="o">==</span> <span class="s1">&#39;M&#39;</span><span class="p">}</span>
    <span class="n">methionine_positions</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">orf_offsets</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
    <span class="k">while</span> <span class="kc">True</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">orf_offsets</span><span class="p">:</span>  <span class="c1"># MET is missing for the sequnce/we haven&#39;t found the ORF start and need to scan a range</span>
            <span class="n">orf_offsets</span> <span class="o">=</span> <span class="p">{</span><span class="n">start_idx</span><span class="p">:</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">start_idx</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="o">-</span><span class="mi">30</span><span class="p">,</span> <span class="mi">50</span><span class="p">)}</span>

        <span class="c1"># Weight potential MET offsets by finding the one which gives the highest number correct mutation sites</span>
        <span class="k">for</span> <span class="n">test_orf_index</span> <span class="ow">in</span> <span class="n">orf_offsets</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">mutation_index</span><span class="p">,</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">mutations</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">sequence</span><span class="p">[</span><span class="n">test_orf_index</span> <span class="o">+</span> <span class="n">mutation_index</span><span class="p">]</span> <span class="o">==</span> <span class="n">mutation</span><span class="p">[</span><span class="s1">&#39;from&#39;</span><span class="p">]:</span>
                        <span class="n">orf_offsets</span><span class="p">[</span><span class="n">test_orf_index</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">except</span> <span class="ne">IndexError</span><span class="p">:</span>  <span class="c1"># We have reached the end of the sequence</span>
                    <span class="k">break</span>

        <span class="c1"># logger.debug(gen_offset_repr())</span>
        <span class="n">max_count</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">orf_offsets</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>
        <span class="c1"># Check if likely ORF has been identified (count &lt; number mutations/2). If not, MET is missing/not the ORF start</span>
        <span class="k">if</span> <span class="n">max_count</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">mutations</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">unsolvable</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Couldn&#39;t find a orf_offset max_count </span><span class="si">{</span><span class="n">max_count</span><span class="si">}</span><span class="s2"> &lt; </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">mutations</span><span class="p">)</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="mi">2</span><span class="si">}</span><span class="s2"> (half the &quot;</span>
                                   <span class="sa">f</span><span class="s2">&quot;mutations). The orf_start_idx=</span><span class="si">{</span><span class="n">orf_start_idx</span><span class="si">}</span><span class="s2"> still</span><span class="se">\n\t</span><span class="si">{</span><span class="n">gen_offset_repr</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">orf_offsets</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">unsolvable</span> <span class="o">=</span> <span class="kc">True</span>  <span class="c1"># if we reach this spot again, the problem is deemed unsolvable</span>
        <span class="k">else</span><span class="p">:</span>  <span class="c1"># Find the index of the max_count</span>
            <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">count_</span> <span class="ow">in</span> <span class="n">orf_offsets</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">max_count</span> <span class="o">==</span> <span class="n">count_</span><span class="p">:</span>  <span class="c1"># orf_offsets[offset]:</span>
                    <span class="n">orf_start_idx</span> <span class="o">=</span> <span class="n">idx</span>  <span class="c1"># Select the first occurrence of the max count</span>
                    <span class="k">break</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Couldn&#39;t find a orf_offset count == </span><span class="si">{</span><span class="n">max_count</span><span class="si">}</span><span class="s2"> (the max_count). &quot;</span>
                                   <span class="sa">f</span><span class="s2">&quot;The orf_start_idx=</span><span class="si">{</span><span class="n">orf_start_idx</span><span class="si">}</span><span class="s2"> still</span><span class="se">\n\t</span><span class="si">{</span><span class="n">gen_offset_repr</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="c1"># For cases where the orf doesn&#39;t begin on Met, try to find a prior Met. Otherwise, selects the id&#39;d Met</span>
            <span class="n">closest_met</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="k">for</span> <span class="n">met_index</span> <span class="ow">in</span> <span class="n">methionine_positions</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">met_index</span> <span class="o">&lt;=</span> <span class="n">orf_start_idx</span><span class="p">:</span>
                    <span class="n">closest_met</span> <span class="o">=</span> <span class="n">met_index</span>
                <span class="k">else</span><span class="p">:</span>  <span class="c1"># We have passed the identified orf_start_idx</span>
                    <span class="k">if</span> <span class="n">closest_met</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                        <span class="n">orf_start_idx</span> <span class="o">=</span> <span class="n">closest_met</span>
                    <span class="k">break</span>
            <span class="k">break</span>

    <span class="k">return</span> <span class="n">orf_start_idx</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="sequence.generate_mutations" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">generate_mutations</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">generate_mutations</span><span class="p">(</span><span class="n">reference</span><span class="p">:</span> <span class="n"><span title="collections.abc.Sequence">Sequence</span></span><span class="p">,</span> <span class="n">query</span><span class="p">:</span> <span class="n"><span title="collections.abc.Sequence">Sequence</span></span><span class="p">,</span> <span class="n">offset</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">keep_gaps</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">remove_termini</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">remove_query_gaps</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">only_gaps</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">zero_index</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">return_all</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">return_to</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">return_from</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n"><a class="autorefs autorefs-internal" title="sequence.mutation_dictionary" href="#sequence.mutation_dictionary">mutation_dictionary</a></span> <span class="o">|</span> <span class="n"><a class="autorefs autorefs-internal" title="sequence.sequence_dictionary" href="#sequence.sequence_dictionary">sequence_dictionary</a></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Create mutation data in a typical A5K format. Integer indexed dictionary keys with the index matching reference
sequence. Sequence mutations accessed by "from" and "to" keys. By default, only mutated positions are returned and
all gaped sequences are excluded</p>
<p>For PDB comparison, reference should be expression sequence (SEQRES), query should be atomic sequence (ATOM)</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>reference</code></b>
            (<code><span title="collections.abc.Sequence">Sequence</span></code>)
        
        <div class="doc-md-description">
          <p>Reference sequence to align mutations against. Character values are returned to the "from" key</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>query</code></b>
            (<code><span title="collections.abc.Sequence">Sequence</span></code>)
        
        <div class="doc-md-description">
          <p>Query sequence. Character values are returned to the "to" key</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>offset</code></b>
            (<code>bool</code>, default:
                <code>True</code>
)
        
        <div class="doc-md-description">
          <p>Whether sequences are different lengths. Will create an alignment of the two sequences</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>keep_gaps</code></b>
            (<code>bool</code>, default:
                <code>False</code>
)
        
        <div class="doc-md-description">
          <p>Return gaped indices, i.e. outside the aligned sequences or missing internal characters</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>remove_termini</code></b>
            (<code>bool</code>, default:
                <code>True</code>
)
        
        <div class="doc-md-description">
          <p>Remove indices that are outside the reference sequence boundaries</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>remove_query_gaps</code></b>
            (<code>bool</code>, default:
                <code>True</code>
)
        
        <div class="doc-md-description">
          <p>Remove indices where there are gaps present in the query sequence</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>only_gaps</code></b>
            (<code>bool</code>, default:
                <code>False</code>
)
        
        <div class="doc-md-description">
          <p>Only include reference indices that are missing query residues. All "to" values will be a gap "-"</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>zero_index</code></b>
            (<code>bool</code>, default:
                <code>False</code>
)
        
        <div class="doc-md-description">
          <p>Whether to return the indices zero-indexed (like python) or one-indexed</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>return_all</code></b>
            (<code>bool</code>, default:
                <code>False</code>
)
        
        <div class="doc-md-description">
          <p>Whether to return all the indices and there corresponding mutational data</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>return_to</code></b>
            (<code>bool</code>, default:
                <code>False</code>
)
        
        <div class="doc-md-description">
          <p>Whether to return only the 'to' amino acid type</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>return_from</code></b>
            (<code>bool</code>, default:
                <code>False</code>
)
        
        <div class="doc-md-description">
          <p>Whether to return only the 'from' amino acid type</p>
        </div>
      </li>
  </ul>
      <p>Returns:
    Mutation index to mutations with format
        {1: {'from': 'A', 'to': 'K'}, ...}
        unless return_to or return_from is True, then
        {1: 'K', ...} or {1: 'A', ...}, respectively</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">799</span>
<span class="normal">800</span>
<span class="normal">801</span>
<span class="normal">802</span>
<span class="normal">803</span>
<span class="normal">804</span>
<span class="normal">805</span>
<span class="normal">806</span>
<span class="normal">807</span>
<span class="normal">808</span>
<span class="normal">809</span>
<span class="normal">810</span>
<span class="normal">811</span>
<span class="normal">812</span>
<span class="normal">813</span>
<span class="normal">814</span>
<span class="normal">815</span>
<span class="normal">816</span>
<span class="normal">817</span>
<span class="normal">818</span>
<span class="normal">819</span>
<span class="normal">820</span>
<span class="normal">821</span>
<span class="normal">822</span>
<span class="normal">823</span>
<span class="normal">824</span>
<span class="normal">825</span>
<span class="normal">826</span>
<span class="normal">827</span>
<span class="normal">828</span>
<span class="normal">829</span>
<span class="normal">830</span>
<span class="normal">831</span>
<span class="normal">832</span>
<span class="normal">833</span>
<span class="normal">834</span>
<span class="normal">835</span>
<span class="normal">836</span>
<span class="normal">837</span>
<span class="normal">838</span>
<span class="normal">839</span>
<span class="normal">840</span>
<span class="normal">841</span>
<span class="normal">842</span>
<span class="normal">843</span>
<span class="normal">844</span>
<span class="normal">845</span>
<span class="normal">846</span>
<span class="normal">847</span>
<span class="normal">848</span>
<span class="normal">849</span>
<span class="normal">850</span>
<span class="normal">851</span>
<span class="normal">852</span>
<span class="normal">853</span>
<span class="normal">854</span>
<span class="normal">855</span>
<span class="normal">856</span>
<span class="normal">857</span>
<span class="normal">858</span>
<span class="normal">859</span>
<span class="normal">860</span>
<span class="normal">861</span>
<span class="normal">862</span>
<span class="normal">863</span>
<span class="normal">864</span>
<span class="normal">865</span>
<span class="normal">866</span>
<span class="normal">867</span>
<span class="normal">868</span>
<span class="normal">869</span>
<span class="normal">870</span>
<span class="normal">871</span>
<span class="normal">872</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">generate_mutations</span><span class="p">(</span><span class="n">reference</span><span class="p">:</span> <span class="n">Sequence</span><span class="p">,</span> <span class="n">query</span><span class="p">:</span> <span class="n">Sequence</span><span class="p">,</span> <span class="n">offset</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">keep_gaps</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
                       <span class="n">remove_termini</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">remove_query_gaps</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">only_gaps</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
                       <span class="n">zero_index</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
                       <span class="n">return_all</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">return_to</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">return_from</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> \
        <span class="o">-&gt;</span> <span class="n">mutation_dictionary</span> <span class="o">|</span> <span class="n">sequence_dictionary</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create mutation data in a typical A5K format. Integer indexed dictionary keys with the index matching reference</span>
<span class="sd">    sequence. Sequence mutations accessed by &quot;from&quot; and &quot;to&quot; keys. By default, only mutated positions are returned and</span>
<span class="sd">    all gaped sequences are excluded</span>

<span class="sd">    For PDB comparison, reference should be expression sequence (SEQRES), query should be atomic sequence (ATOM)</span>

<span class="sd">    Args:</span>
<span class="sd">        reference: Reference sequence to align mutations against. Character values are returned to the &quot;from&quot; key</span>
<span class="sd">        query: Query sequence. Character values are returned to the &quot;to&quot; key</span>
<span class="sd">        offset: Whether sequences are different lengths. Will create an alignment of the two sequences</span>
<span class="sd">        keep_gaps: Return gaped indices, i.e. outside the aligned sequences or missing internal characters</span>
<span class="sd">        remove_termini: Remove indices that are outside the reference sequence boundaries</span>
<span class="sd">        remove_query_gaps: Remove indices where there are gaps present in the query sequence</span>
<span class="sd">        only_gaps: Only include reference indices that are missing query residues. All &quot;to&quot; values will be a gap &quot;-&quot;</span>
<span class="sd">        zero_index: Whether to return the indices zero-indexed (like python) or one-indexed</span>
<span class="sd">        return_all: Whether to return all the indices and there corresponding mutational data</span>
<span class="sd">        return_to: Whether to return only the &#39;to&#39; amino acid type</span>
<span class="sd">        return_from: Whether to return only the &#39;from&#39; amino acid type</span>
<span class="sd">    Returns:</span>
<span class="sd">        Mutation index to mutations with format</span>
<span class="sd">            {1: {&#39;from&#39;: &#39;A&#39;, &#39;to&#39;: &#39;K&#39;}, ...}</span>
<span class="sd">            unless return_to or return_from is True, then</span>
<span class="sd">            {1: &#39;K&#39;, ...} or {1: &#39;A&#39;, ...}, respectively</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">offset</span><span class="p">:</span>
        <span class="n">alignment</span> <span class="o">=</span> <span class="n">generate_alignment</span><span class="p">(</span><span class="n">reference</span><span class="p">,</span> <span class="n">query</span><span class="p">)</span>
        <span class="n">align_seq_1</span><span class="p">,</span> <span class="n">align_seq_2</span> <span class="o">=</span> <span class="n">alignment</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">align_seq_1</span><span class="p">,</span> <span class="n">align_seq_2</span> <span class="o">=</span> <span class="n">reference</span><span class="p">,</span> <span class="n">query</span>

    <span class="n">idx_offset</span> <span class="o">=</span> <span class="mi">0</span> <span class="k">if</span> <span class="n">zero_index</span> <span class="k">else</span> <span class="n">ZERO_OFFSET</span>

    <span class="c1"># Get the first matching index of the reference sequence</span>
    <span class="n">starting_idx_of_seq1</span> <span class="o">=</span> <span class="n">align_seq_1</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="n">reference</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="c1"># Ensure iteration sequence1/reference starts at idx 0       v</span>
    <span class="n">sequence_iterator</span> <span class="o">=</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">align_seq_1</span><span class="p">,</span> <span class="n">align_seq_2</span><span class="p">),</span> <span class="o">-</span><span class="n">starting_idx_of_seq1</span> <span class="o">+</span> <span class="n">idx_offset</span><span class="p">)</span>
    <span class="c1"># Extract differences from the alignment</span>
    <span class="k">if</span> <span class="n">return_all</span><span class="p">:</span>
        <span class="n">mutations</span> <span class="o">=</span> <span class="p">{</span><span class="n">idx</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;from&#39;</span><span class="p">:</span> <span class="n">char1</span><span class="p">,</span> <span class="s1">&#39;to&#39;</span><span class="p">:</span> <span class="n">char2</span><span class="p">}</span> <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="p">(</span><span class="n">char1</span><span class="p">,</span> <span class="n">char2</span><span class="p">)</span> <span class="ow">in</span> <span class="n">sequence_iterator</span><span class="p">}</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">mutations</span> <span class="o">=</span> <span class="p">{</span><span class="n">idx</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;from&#39;</span><span class="p">:</span> <span class="n">char1</span><span class="p">,</span> <span class="s1">&#39;to&#39;</span><span class="p">:</span> <span class="n">char2</span><span class="p">}</span> <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="p">(</span><span class="n">char1</span><span class="p">,</span> <span class="n">char2</span><span class="p">)</span> <span class="ow">in</span> <span class="n">sequence_iterator</span> <span class="k">if</span> <span class="n">char1</span> <span class="o">!=</span> <span class="n">char2</span><span class="p">}</span>

    <span class="c1"># Find last index of reference (including internal gaps)</span>
    <span class="n">starting_key_of_seq1</span> <span class="o">=</span> <span class="n">idx_offset</span>
    <span class="n">ending_key_of_seq1</span> <span class="o">=</span> <span class="n">starting_key_of_seq1</span> <span class="o">+</span> <span class="n">align_seq_1</span><span class="o">.</span><span class="n">rfind</span><span class="p">(</span><span class="n">reference</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
    <span class="n">remove_mutation_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="n">only_gaps</span><span class="p">:</span>  <span class="c1"># Remove the actual mutations, keep internal and external gap indices and the reference sequence</span>
        <span class="n">keep_gaps</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">remove_mutation_list</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">entry</span> <span class="k">for</span> <span class="n">entry</span><span class="p">,</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">mutations</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
                                     <span class="k">if</span> <span class="n">mutation</span><span class="p">[</span><span class="s1">&#39;from&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;-&#39;</span> <span class="ow">and</span> <span class="n">mutation</span><span class="p">[</span><span class="s1">&#39;to&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;-&#39;</span><span class="p">])</span>
    <span class="k">if</span> <span class="n">keep_gaps</span><span class="p">:</span>  <span class="c1"># Leave all types of keep_gaps, otherwise check for each requested type</span>
        <span class="n">remove_termini</span> <span class="o">=</span> <span class="n">remove_query_gaps</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="k">if</span> <span class="n">remove_termini</span><span class="p">:</span>  <span class="c1"># Remove indices outside of sequence 1</span>
        <span class="n">remove_mutation_list</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">entry</span> <span class="k">for</span> <span class="n">entry</span> <span class="ow">in</span> <span class="n">mutations</span>
                                     <span class="k">if</span> <span class="n">entry</span> <span class="o">&lt;</span> <span class="n">starting_key_of_seq1</span> <span class="ow">or</span> <span class="n">entry</span> <span class="o">&gt;</span> <span class="n">ending_key_of_seq1</span><span class="p">])</span>

    <span class="k">if</span> <span class="n">remove_query_gaps</span><span class="p">:</span>  <span class="c1"># Remove indices where sequence 2 is gaped</span>
        <span class="n">remove_mutation_list</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">entry</span> <span class="k">for</span> <span class="n">entry</span><span class="p">,</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">mutations</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
                                     <span class="k">if</span> <span class="n">starting_key_of_seq1</span> <span class="o">&lt;</span> <span class="n">entry</span> <span class="o">&lt;=</span> <span class="n">ending_key_of_seq1</span> <span class="ow">and</span> <span class="n">mutation</span><span class="p">[</span><span class="s1">&#39;to&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;-&#39;</span><span class="p">])</span>
    <span class="k">for</span> <span class="n">entry</span> <span class="ow">in</span> <span class="n">remove_mutation_list</span><span class="p">:</span>
        <span class="n">mutations</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">entry</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">return_to</span><span class="p">:</span>
        <span class="n">mutations</span> <span class="o">=</span> <span class="p">{</span><span class="n">idx</span><span class="p">:</span> <span class="n">_mutation_dictionary</span><span class="p">[</span><span class="s1">&#39;to&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">_mutation_dictionary</span> <span class="ow">in</span> <span class="n">mutations</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
    <span class="k">elif</span> <span class="n">return_from</span><span class="p">:</span>
        <span class="n">mutations</span> <span class="o">=</span> <span class="p">{</span><span class="n">idx</span><span class="p">:</span> <span class="n">_mutation_dictionary</span><span class="p">[</span><span class="s1">&#39;from&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">_mutation_dictionary</span> <span class="ow">in</span> <span class="n">mutations</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>

    <span class="k">return</span> <span class="n">mutations</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="sequence.pdb_to_pose_offset" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">pdb_to_pose_offset</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">pdb_to_pose_offset</span><span class="p">(</span><span class="n">reference_sequence</span><span class="p">:</span> <span class="n">dict</span><span class="p">[</span><span class="n"><span title="typing.Any">Any</span></span><span class="p">,</span> <span class="n"><span title="collections.abc.Sequence">Sequence</span></span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">dict</span><span class="p">[</span><span class="n"><span title="typing.Any">Any</span></span><span class="p">,</span> <span class="n">int</span><span class="p">]</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Take a dictionary with chain name as keys and return the length of Pose numbering offset</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>reference_sequence</code></b>
            (<code>dict[<span title="typing.Any">Any</span>, <span title="collections.abc.Sequence">Sequence</span>]</code>)
        
        <div class="doc-md-description">
          <p>{key1: 'MSGKLDA...', ...} or {key2: {1: 'A', 2: 'S', ...}, ...}</p>
        </div>
      </li>
  </ul>
      <p>Returns:
    {key1: 0, key2: 123, ...}</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">875</span>
<span class="normal">876</span>
<span class="normal">877</span>
<span class="normal">878</span>
<span class="normal">879</span>
<span class="normal">880</span>
<span class="normal">881</span>
<span class="normal">882</span>
<span class="normal">883</span>
<span class="normal">884</span>
<span class="normal">885</span>
<span class="normal">886</span>
<span class="normal">887</span>
<span class="normal">888</span>
<span class="normal">889</span>
<span class="normal">890</span>
<span class="normal">891</span>
<span class="normal">892</span>
<span class="normal">893</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">pdb_to_pose_offset</span><span class="p">(</span><span class="n">reference_sequence</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="n">Any</span><span class="p">,</span> <span class="n">Sequence</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="n">Any</span><span class="p">,</span> <span class="nb">int</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Take a dictionary with chain name as keys and return the length of Pose numbering offset</span>

<span class="sd">    Args:</span>
<span class="sd">        reference_sequence: {key1: &#39;MSGKLDA...&#39;, ...} or {key2: {1: &#39;A&#39;, 2: &#39;S&#39;, ...}, ...}</span>
<span class="sd">    Returns:</span>
<span class="sd">        {key1: 0, key2: 123, ...}</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">offset</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="c1"># prior_chain = None</span>
    <span class="n">prior_chains_len</span> <span class="o">=</span> <span class="n">prior_key</span> <span class="o">=</span> <span class="mi">0</span>  <span class="c1"># prior_key not used as 0 but to ensure initialized nonetheless</span>
    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">key</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">reference_sequence</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">idx</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">prior_chains_len</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">reference_sequence</span><span class="p">[</span><span class="n">prior_key</span><span class="p">])</span>
        <span class="n">offset</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">prior_chains_len</span>
        <span class="c1"># insert function here? Make this a decorator!?</span>
        <span class="n">prior_key</span> <span class="o">=</span> <span class="n">key</span>

    <span class="k">return</span> <span class="n">offset</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="sequence.generate_multiple_mutations" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">generate_multiple_mutations</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">generate_multiple_mutations</span><span class="p">(</span><span class="n">reference</span><span class="p">:</span> <span class="n">dict</span><span class="p">[</span><span class="n">str</span><span class="p">,</span> <span class="n">str</span><span class="p">],</span> <span class="n">sequences</span><span class="p">:</span> <span class="n">dict</span><span class="p">[</span><span class="n">str</span><span class="p">,</span> <span class="n">dict</span><span class="p">[</span><span class="n">str</span><span class="p">,</span> <span class="n">str</span><span class="p">]],</span> <span class="n">pose_num</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">dict</span><span class="p">[</span><span class="n">str</span><span class="p">,</span> <span class="n">dict</span><span class="p">[</span><span class="n">str</span><span class="p">,</span> <span class="n">dict</span><span class="p">[</span><span class="n">int</span><span class="p">]]]</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Extract mutation data from multiple sequence dictionaries with regard to a reference. Default is Pose numbering</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>reference</code></b>
            (<code>dict[str, str]</code>)
        
        <div class="doc-md-description">
          <p>{chain: sequence, ...} The reference sequence to compare sequences to</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>sequences</code></b>
            (<code>dict[str, dict[str, str]]</code>)
        
        <div class="doc-md-description">
          <p>{pdb_code: {chain: sequence, ...}, ...}</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>pose_num</code></b>
            (<code>bool</code>, default:
                <code>True</code>
)
        
        <div class="doc-md-description">
          <p>Whether to return the mutations in Pose numbering with the first Entity as 1 and the
second Entity as Entity1 last residue + 1</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
            <code>dict[str, dict[str, dict[int]]]</code>
        
        <div class="doc-md-description">
          <p>{pdb_code: {chain_id: {mutation_index: {'from': 'A', 'to': 'K'}, ...}, ...}, ...}</p>
      </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">896</span>
<span class="normal">897</span>
<span class="normal">898</span>
<span class="normal">899</span>
<span class="normal">900</span>
<span class="normal">901</span>
<span class="normal">902</span>
<span class="normal">903</span>
<span class="normal">904</span>
<span class="normal">905</span>
<span class="normal">906</span>
<span class="normal">907</span>
<span class="normal">908</span>
<span class="normal">909</span>
<span class="normal">910</span>
<span class="normal">911</span>
<span class="normal">912</span>
<span class="normal">913</span>
<span class="normal">914</span>
<span class="normal">915</span>
<span class="normal">916</span>
<span class="normal">917</span>
<span class="normal">918</span>
<span class="normal">919</span>
<span class="normal">920</span>
<span class="normal">921</span>
<span class="normal">922</span>
<span class="normal">923</span>
<span class="normal">924</span>
<span class="normal">925</span>
<span class="normal">926</span>
<span class="normal">927</span>
<span class="normal">928</span>
<span class="normal">929</span>
<span class="normal">930</span>
<span class="normal">931</span>
<span class="normal">932</span>
<span class="normal">933</span>
<span class="normal">934</span>
<span class="normal">935</span>
<span class="normal">936</span>
<span class="normal">937</span>
<span class="normal">938</span>
<span class="normal">939</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">generate_multiple_mutations</span><span class="p">(</span>
    <span class="n">reference</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span> <span class="n">sequences</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]],</span> <span class="n">pose_num</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="p">]]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Extract mutation data from multiple sequence dictionaries with regard to a reference. Default is Pose numbering</span>

<span class="sd">    Args:</span>
<span class="sd">        reference: {chain: sequence, ...} The reference sequence to compare sequences to</span>
<span class="sd">        sequences: {pdb_code: {chain: sequence, ...}, ...}</span>
<span class="sd">        pose_num: Whether to return the mutations in Pose numbering with the first Entity as 1 and the</span>
<span class="sd">            second Entity as Entity1 last residue + 1</span>

<span class="sd">    Returns:</span>
<span class="sd">        {pdb_code: {chain_id: {mutation_index: {&#39;from&#39;: &#39;A&#39;, &#39;to&#39;: &#39;K&#39;}, ...}, ...}, ...}</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Add reference sequence mutations</span>
    <span class="n">mutations</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;reference&#39;</span><span class="p">:</span> <span class="p">{</span><span class="n">chain</span><span class="p">:</span> <span class="p">{</span><span class="n">sequence_idx</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;from&#39;</span><span class="p">:</span> <span class="n">aa</span><span class="p">,</span> <span class="s1">&#39;to&#39;</span><span class="p">:</span> <span class="n">aa</span><span class="p">}</span>
                                       <span class="k">for</span> <span class="n">sequence_idx</span><span class="p">,</span> <span class="n">aa</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">ref_sequence</span><span class="p">,</span> <span class="mi">1</span><span class="p">)}</span>
                               <span class="k">for</span> <span class="n">chain</span><span class="p">,</span> <span class="n">ref_sequence</span> <span class="ow">in</span> <span class="n">reference</span><span class="o">.</span><span class="n">items</span><span class="p">()}}</span>
    <span class="c1">#                         returns {1: {&#39;from&#39;: &#39;A&#39;, &#39;to&#39;: &#39;K&#39;}, ...}</span>
    <span class="c1"># mutations = {pdb: {chain: generate_mutations(sequence, reference[chain], offset=False)</span>
    <span class="c1">#                    for chain, sequence in chain_sequences.items()}</span>
    <span class="c1">#              for pdb, chain_sequences in pdb_sequences.items()}</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">chain_sequences</span> <span class="ow">in</span> <span class="n">sequences</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">mutations</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">chain</span><span class="p">,</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">chain_sequences</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="n">mutations</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">chain</span><span class="p">]</span> <span class="o">=</span> <span class="n">generate_mutations</span><span class="p">(</span><span class="n">reference</span><span class="p">[</span><span class="n">chain</span><span class="p">],</span> <span class="n">sequence</span><span class="p">,</span> <span class="n">offset</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;The &#39;reference&#39; and &#39;sequences&#39; have different chains. Chain &#39;</span><span class="si">{</span><span class="n">chain</span><span class="si">}</span><span class="s2">&#39; isn&#39;t in the reference&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">pose_num</span><span class="p">:</span>
        <span class="n">offset_dict</span> <span class="o">=</span> <span class="n">pdb_to_pose_offset</span><span class="p">(</span><span class="n">reference</span><span class="p">)</span>
        <span class="c1"># pose_mutations = {}</span>
        <span class="c1"># for chain, offset in offset_dict.items():</span>
        <span class="c1">#     for pdb_code in mutations:</span>
        <span class="c1">#         if pdb_code not in pose_mutations:</span>
        <span class="c1">#             pose_mutations[pdb_code] = {}</span>
        <span class="c1">#         pose_mutations[pdb_code][chain] = {}</span>
        <span class="c1">#         for mutation_idx in mutations[pdb_code][chain]:</span>
        <span class="c1">#             pose_mutations[pdb_code][chain][mutation_idx + offset] = mutations[pdb_code][chain][mutation_idx]</span>
        <span class="c1"># mutations = pose_mutations</span>
        <span class="n">mutations</span> <span class="o">=</span> <span class="p">{</span><span class="n">name</span><span class="p">:</span> <span class="p">{</span><span class="n">chain</span><span class="p">:</span> <span class="p">{</span><span class="n">idx</span> <span class="o">+</span> <span class="n">offset</span><span class="p">:</span> <span class="n">mutation</span> <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">chain_mutations</span><span class="p">[</span><span class="n">chain</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
                            <span class="k">for</span> <span class="n">chain</span><span class="p">,</span> <span class="n">offset</span> <span class="ow">in</span> <span class="n">offset_dict</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span> <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">chain_mutations</span> <span class="ow">in</span> <span class="n">mutations</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
    <span class="k">return</span> <span class="n">mutations</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="sequence.generate_mutations_from_reference" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">generate_mutations_from_reference</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">generate_mutations_from_reference</span><span class="p">(</span><span class="n">reference</span><span class="p">:</span> <span class="n"><span title="collections.abc.Sequence">Sequence</span></span><span class="p">[</span><span class="n">str</span><span class="p">],</span> <span class="n">sequences</span><span class="p">:</span> <span class="n">dict</span><span class="p">[</span><span class="n">str</span><span class="p">,</span> <span class="n"><span title="collections.abc.Sequence">Sequence</span></span><span class="p">[</span><span class="n">str</span><span class="p">]],</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">dict</span><span class="p">[</span><span class="n">str</span><span class="p">,</span> <span class="n"><a class="autorefs autorefs-internal" title="sequence.mutation_dictionary" href="#sequence.mutation_dictionary">mutation_dictionary</a></span> <span class="o">|</span> <span class="n"><a class="autorefs autorefs-internal" title="sequence.sequence_dictionary" href="#sequence.sequence_dictionary">sequence_dictionary</a></span><span class="p">]</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Generate mutation data from multiple alias mapped sequence dictionaries with regard to a single reference.</p>
<p>Defaults to returning only mutations (return_all=False) and forgoes any sequence alignment (offset=False)</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>reference</code></b>
            (<code><span title="collections.abc.Sequence">Sequence</span>[str]</code>)
        
        <div class="doc-md-description">
          <p>The reference sequence to align each sequence against.
Character values are returned to the "from" key</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>sequences</code></b>
            (<code>dict[str, <span title="collections.abc.Sequence">Sequence</span>[str]]</code>)
        
        <div class="doc-md-description">
          <p>The template sequences to align, i.e. {alias: sequence, ...}.
Character values are returned to the "to" key</p>
        </div>
      </li>
  </ul>



  <p>Other Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>offset</code></b>
        
        <div class="doc-md-description">
          <p>bool = True - Whether sequences are different lengths. Will create an alignment of the two sequences</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>keep_gaps</code></b>
        
        <div class="doc-md-description">
          <p>bool = False - Return gaped indices, i.e. outside the aligned sequences or missing internal
characters</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>remove_termini</code></b>
        
        <div class="doc-md-description">
          <p>bool = True - Remove indices that are outside the reference sequence boundaries</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>remove_query_gaps</code></b>
        
        <div class="doc-md-description">
          <p>bool = True - Remove indices where there are gaps present in the query sequence</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>only_gaps</code></b>
        
        <div class="doc-md-description">
          <p>bool = False - Only include reference indices that are missing query residues.
All "to" values will be a gap "-"</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>zero_index</code></b>
        
        <div class="doc-md-description">
          <p>bool = False - Whether to return the indices zero-indexed (like python Sequence) or one-indexed</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>return_all</code></b>
        
        <div class="doc-md-description">
          <p>bool = False - Whether to return all the indices and there corresponding mutational data</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>return_to</code></b>
        
        <div class="doc-md-description">
          <p>bool = False - Whether to return only the "to" amino acid type</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>return_from</code></b>
        
        <div class="doc-md-description">
          <p>bool = False - Whether to return only the "from" amino acid type</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
            <code>dict[str, <a class="autorefs autorefs-internal" title="sequence.mutation_dictionary" href="#sequence.mutation_dictionary">mutation_dictionary</a> | <a class="autorefs autorefs-internal" title="sequence.sequence_dictionary" href="#sequence.sequence_dictionary">sequence_dictionary</a>]</code>
        
        <div class="doc-md-description">
          <p>{alias: {mutation_index: {'from': 'A', 'to': 'K'}, ...}, ...} unless return_to or return_from is True, then
{alias: {mutation_index: 'K', ...}, ...}</p>
      </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">942</span>
<span class="normal">943</span>
<span class="normal">944</span>
<span class="normal">945</span>
<span class="normal">946</span>
<span class="normal">947</span>
<span class="normal">948</span>
<span class="normal">949</span>
<span class="normal">950</span>
<span class="normal">951</span>
<span class="normal">952</span>
<span class="normal">953</span>
<span class="normal">954</span>
<span class="normal">955</span>
<span class="normal">956</span>
<span class="normal">957</span>
<span class="normal">958</span>
<span class="normal">959</span>
<span class="normal">960</span>
<span class="normal">961</span>
<span class="normal">962</span>
<span class="normal">963</span>
<span class="normal">964</span>
<span class="normal">965</span>
<span class="normal">966</span>
<span class="normal">967</span>
<span class="normal">968</span>
<span class="normal">969</span>
<span class="normal">970</span>
<span class="normal">971</span>
<span class="normal">972</span>
<span class="normal">973</span>
<span class="normal">974</span>
<span class="normal">975</span>
<span class="normal">976</span>
<span class="normal">977</span>
<span class="normal">978</span>
<span class="normal">979</span>
<span class="normal">980</span>
<span class="normal">981</span>
<span class="normal">982</span>
<span class="normal">983</span>
<span class="normal">984</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">generate_mutations_from_reference</span><span class="p">(</span><span class="n">reference</span><span class="p">:</span> <span class="n">Sequence</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">sequences</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Sequence</span><span class="p">[</span><span class="nb">str</span><span class="p">]],</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> \
        <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">mutation_dictionary</span> <span class="o">|</span> <span class="n">sequence_dictionary</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Generate mutation data from multiple alias mapped sequence dictionaries with regard to a single reference.</span>

<span class="sd">    Defaults to returning only mutations (return_all=False) and forgoes any sequence alignment (offset=False)</span>

<span class="sd">    Args:</span>
<span class="sd">        reference: The reference sequence to align each sequence against.</span>
<span class="sd">            Character values are returned to the &quot;from&quot; key</span>
<span class="sd">        sequences: The template sequences to align, i.e. {alias: sequence, ...}.</span>
<span class="sd">            Character values are returned to the &quot;to&quot; key</span>

<span class="sd">    Keyword Args:</span>
<span class="sd">        offset: bool = True - Whether sequences are different lengths. Will create an alignment of the two sequences</span>
<span class="sd">        keep_gaps: bool = False - Return gaped indices, i.e. outside the aligned sequences or missing internal</span>
<span class="sd">            characters</span>
<span class="sd">        remove_termini: bool = True - Remove indices that are outside the reference sequence boundaries</span>
<span class="sd">        remove_query_gaps: bool = True - Remove indices where there are gaps present in the query sequence</span>
<span class="sd">        only_gaps: bool = False - Only include reference indices that are missing query residues.</span>
<span class="sd">            All &quot;to&quot; values will be a gap &quot;-&quot;</span>
<span class="sd">        zero_index: bool = False - Whether to return the indices zero-indexed (like python Sequence) or one-indexed</span>
<span class="sd">        return_all: bool = False - Whether to return all the indices and there corresponding mutational data</span>
<span class="sd">        return_to: bool = False - Whether to return only the &quot;to&quot; amino acid type</span>
<span class="sd">        return_from: bool = False - Whether to return only the &quot;from&quot; amino acid type</span>

<span class="sd">    Returns:</span>
<span class="sd">        {alias: {mutation_index: {&#39;from&#39;: &#39;A&#39;, &#39;to&#39;: &#39;K&#39;}, ...}, ...} unless return_to or return_from is True, then</span>
<span class="sd">            {alias: {mutation_index: &#39;K&#39;, ...}, ...}</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># offset_value = kwargs.get(&#39;offset&#39;)</span>
    <span class="n">kwargs</span><span class="p">[</span><span class="s1">&#39;offset&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;offset&#39;</span><span class="p">)</span> <span class="ow">or</span> <span class="kc">False</span><span class="p">)</span>  <span class="c1"># Default to False if there was no argument passed</span>
    <span class="n">mutations</span> <span class="o">=</span> <span class="p">{</span><span class="n">alias</span><span class="p">:</span> <span class="n">generate_mutations</span><span class="p">(</span><span class="n">reference</span><span class="p">,</span> <span class="n">sequence</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>  <span class="c1"># offset=False,</span>
                 <span class="k">for</span> <span class="n">alias</span><span class="p">,</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">sequences</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>

    <span class="c1"># Add the reference sequence to mutation data</span>
    <span class="k">if</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;return_to&#39;</span><span class="p">)</span> <span class="ow">or</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;return_from&#39;</span><span class="p">):</span>
        <span class="n">mutations</span><span class="p">[</span><span class="n">putils</span><span class="o">.</span><span class="n">reference_name</span><span class="p">]</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">enumerate</span><span class="p">(</span><span class="n">reference</span><span class="p">,</span> <span class="mi">0</span> <span class="k">if</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;zero_index&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="mi">1</span><span class="p">))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">mutations</span><span class="p">[</span><span class="n">putils</span><span class="o">.</span><span class="n">reference_name</span><span class="p">]</span> <span class="o">=</span> \
            <span class="p">{</span><span class="n">sequence_idx</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;from&#39;</span><span class="p">:</span> <span class="n">aa</span><span class="p">,</span> <span class="s1">&#39;to&#39;</span><span class="p">:</span> <span class="n">aa</span><span class="p">}</span>
             <span class="k">for</span> <span class="n">sequence_idx</span><span class="p">,</span> <span class="n">aa</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">reference</span><span class="p">,</span> <span class="mi">0</span> <span class="k">if</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;zero_index&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="mi">1</span><span class="p">)}</span>

    <span class="k">return</span> <span class="n">mutations</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="sequence.make_sequences_from_mutations" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">make_sequences_from_mutations</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">make_sequences_from_mutations</span><span class="p">(</span><span class="n">wild_type</span><span class="p">:</span> <span class="n">str</span><span class="p">,</span> <span class="n">pdb_mutations</span><span class="p">:</span> <span class="n">dict</span><span class="p">,</span> <span class="n">aligned</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">dict</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Takes a list of sequence mutations and returns the mutated form on wildtype</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>wild_type</code></b>
            (<code>str</code>)
        
        <div class="doc-md-description">
          <p>Sequence to mutate</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>pdb_mutations</code></b>
            (<code>dict</code>)
        
        <div class="doc-md-description">
          <p>{name: {mutation_index: {'from': AA, 'to': AA}, ...}, ...}, ...}</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>aligned</code></b>
            (<code>bool</code>, default:
                <code>False</code>
)
        
        <div class="doc-md-description">
          <p>Whether the input sequences are already aligned</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
            <code>dict</code>
        
        <div class="doc-md-description">
          <p>{name: sequence, ...}</p>
      </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">987</span>
<span class="normal">988</span>
<span class="normal">989</span>
<span class="normal">990</span>
<span class="normal">991</span>
<span class="normal">992</span>
<span class="normal">993</span>
<span class="normal">994</span>
<span class="normal">995</span>
<span class="normal">996</span>
<span class="normal">997</span>
<span class="normal">998</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">make_sequences_from_mutations</span><span class="p">(</span><span class="n">wild_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">pdb_mutations</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span> <span class="n">aligned</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Takes a list of sequence mutations and returns the mutated form on wildtype</span>

<span class="sd">    Args:</span>
<span class="sd">        wild_type: Sequence to mutate</span>
<span class="sd">        pdb_mutations: {name: {mutation_index: {&#39;from&#39;: AA, &#39;to&#39;: AA}, ...}, ...}, ...}</span>
<span class="sd">        aligned: Whether the input sequences are already aligned</span>

<span class="sd">    Returns:</span>
<span class="sd">        {name: sequence, ...}</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">{</span><span class="n">pdb</span><span class="p">:</span> <span class="n">make_mutations</span><span class="p">(</span><span class="n">wild_type</span><span class="p">,</span> <span class="n">mutations</span><span class="p">,</span> <span class="n">find_orf</span><span class="o">=</span><span class="ow">not</span> <span class="n">aligned</span><span class="p">)</span> <span class="k">for</span> <span class="n">pdb</span><span class="p">,</span> <span class="n">mutations</span> <span class="ow">in</span> <span class="n">pdb_mutations</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="sequence.generate_sequences" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">generate_sequences</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">generate_sequences</span><span class="p">(</span><span class="n">wild_type_sequences</span><span class="p">:</span> <span class="n">dict</span><span class="p">,</span> <span class="n">all_design_mutations</span><span class="p">:</span> <span class="n">dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">dict</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Separate chains from mutation dictionary and generate mutated sequences</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>wild_type_sequences</code></b>
            (<code>dict</code>)
        
        <div class="doc-md-description">
          <p>{chain: sequence, ...}</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>all_design_mutations</code></b>
            (<code>dict</code>)
        
        <div class="doc-md-description">
          <p>{'name': {chain: {mutation_index: {'from': AA, 'to': AA}, ...}, ...}, ...}
Index so mutation_index starts at 1</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
            <code>dict</code>
        
        <div class="doc-md-description">
          <p>{chain: {name: sequence, ...}</p>
      </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1001</span>
<span class="normal">1002</span>
<span class="normal">1003</span>
<span class="normal">1004</span>
<span class="normal">1005</span>
<span class="normal">1006</span>
<span class="normal">1007</span>
<span class="normal">1008</span>
<span class="normal">1009</span>
<span class="normal">1010</span>
<span class="normal">1011</span>
<span class="normal">1012</span>
<span class="normal">1013</span>
<span class="normal">1014</span>
<span class="normal">1015</span>
<span class="normal">1016</span>
<span class="normal">1017</span>
<span class="normal">1018</span>
<span class="normal">1019</span>
<span class="normal">1020</span>
<span class="normal">1021</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">generate_sequences</span><span class="p">(</span><span class="n">wild_type_sequences</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span> <span class="n">all_design_mutations</span><span class="p">:</span> <span class="nb">dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Separate chains from mutation dictionary and generate mutated sequences</span>

<span class="sd">    Args:</span>
<span class="sd">        wild_type_sequences: {chain: sequence, ...}</span>
<span class="sd">        all_design_mutations: {&#39;name&#39;: {chain: {mutation_index: {&#39;from&#39;: AA, &#39;to&#39;: AA}, ...}, ...}, ...}</span>
<span class="sd">            Index so mutation_index starts at 1</span>

<span class="sd">    Returns:</span>
<span class="sd">        {chain: {name: sequence, ...}</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">mutated_sequences</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">chain</span> <span class="ow">in</span> <span class="n">wild_type_sequences</span><span class="p">:</span>
        <span class="c1"># pdb_chain_mutations = {pdb: chain_mutations.get(chain) for pdb, chain_mutations in all_design_mutations.items()}</span>
        <span class="n">pdb_chain_mutations</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">pdb</span><span class="p">,</span> <span class="n">chain_mutations</span> <span class="ow">in</span> <span class="n">all_design_mutations</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">chain</span> <span class="ow">in</span> <span class="n">chain_mutations</span><span class="p">:</span>
                <span class="n">pdb_chain_mutations</span><span class="p">[</span><span class="n">pdb</span><span class="p">]</span> <span class="o">=</span> <span class="n">all_design_mutations</span><span class="p">[</span><span class="n">pdb</span><span class="p">][</span><span class="n">chain</span><span class="p">]</span>
        <span class="n">mutated_sequences</span><span class="p">[</span><span class="n">chain</span><span class="p">]</span> <span class="o">=</span> <span class="n">make_sequences_from_mutations</span><span class="p">(</span><span class="n">wild_type_sequences</span><span class="p">[</span><span class="n">chain</span><span class="p">],</span> <span class="n">pdb_chain_mutations</span><span class="p">,</span>
                                                                 <span class="n">aligned</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">mutated_sequences</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="sequence.numeric_to_sequence" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">numeric_to_sequence</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">numeric_to_sequence</span><span class="p">(</span><span class="n">numeric_sequence</span><span class="p">:</span> <span class="n"><span title="numpy.ndarray">ndarray</span></span><span class="p">,</span> <span class="n">translation_table</span><span class="p">:</span> <span class="n">dict</span><span class="p">[</span><span class="n">str</span><span class="p">,</span> <span class="n">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">alphabet_order</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n"><span title="numpy.ndarray">ndarray</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Convert a numeric sequence array into a sequence array</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>numeric_sequence</code></b>
            (<code><span title="numpy.ndarray">ndarray</span></code>)
        
        <div class="doc-md-description">
          <p>The sequence to convert</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>translation_table</code></b>
            (<code>dict[str, int]</code>, default:
                <code>None</code>
)
        
        <div class="doc-md-description">
          <p>If a translation table is provided, it will be used. If not, use alphabet_order</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>alphabet_order</code></b>
            (<code>int</code>, default:
                <code>1</code>
)
        
        <div class="doc-md-description">
          <p>The alphabetical order of the amino acid alphabet. Can be either 1 or 3</p>
        </div>
      </li>
  </ul>
      <p>Returns:
    The alphabetically encoded sequence where each entry along axis=-1 is the one letter amino acid</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1024</span>
<span class="normal">1025</span>
<span class="normal">1026</span>
<span class="normal">1027</span>
<span class="normal">1028</span>
<span class="normal">1029</span>
<span class="normal">1030</span>
<span class="normal">1031</span>
<span class="normal">1032</span>
<span class="normal">1033</span>
<span class="normal">1034</span>
<span class="normal">1035</span>
<span class="normal">1036</span>
<span class="normal">1037</span>
<span class="normal">1038</span>
<span class="normal">1039</span>
<span class="normal">1040</span>
<span class="normal">1041</span>
<span class="normal">1042</span>
<span class="normal">1043</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">numeric_to_sequence</span><span class="p">(</span><span class="n">numeric_sequence</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">translation_table</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                        <span class="n">alphabet_order</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Convert a numeric sequence array into a sequence array</span>

<span class="sd">    Args:</span>
<span class="sd">        numeric_sequence: The sequence to convert</span>
<span class="sd">        translation_table: If a translation table is provided, it will be used. If not, use alphabet_order</span>
<span class="sd">        alphabet_order: The alphabetical order of the amino acid alphabet. Can be either 1 or 3</span>
<span class="sd">    Returns:</span>
<span class="sd">        The alphabetically encoded sequence where each entry along axis=-1 is the one letter amino acid</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">translation_table</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">vectorize</span><span class="p">(</span><span class="n">translation_table</span><span class="o">.</span><span class="fm">__getitem__</span><span class="p">)(</span><span class="n">numeric_sequence</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">alphabet_order</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">vectorize</span><span class="p">(</span><span class="n">sequence_translation_alph1</span><span class="o">.</span><span class="fm">__getitem__</span><span class="p">)(</span><span class="n">numeric_sequence</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">alphabet_order</span> <span class="o">==</span> <span class="mi">3</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">vectorize</span><span class="p">(</span><span class="n">sequence_translation_alph3</span><span class="o">.</span><span class="fm">__getitem__</span><span class="p">)(</span><span class="n">numeric_sequence</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The &#39;alphabet_order&#39; </span><span class="si">{</span><span class="n">alphabet_order</span><span class="si">}</span><span class="s2"> isn&#39;t valid. Choose from either 1 or 3&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="sequence.get_equivalent_indices" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">get_equivalent_indices</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">get_equivalent_indices</span><span class="p">(</span><span class="n">target</span><span class="p">:</span> <span class="n"><span title="collections.abc.Sequence">Sequence</span></span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">query</span><span class="p">:</span> <span class="n"><span title="collections.abc.Sequence">Sequence</span></span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">mutation_allowed</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">tuple</span><span class="p">[</span><span class="n">list</span><span class="p">[</span><span class="n">int</span><span class="p">],</span> <span class="n">list</span><span class="p">[</span><span class="n">int</span><span class="p">]]</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>From two sequences, find the indices where both sequences are equal</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>target</code></b>
            (<code><span title="collections.abc.Sequence">Sequence</span></code>, default:
                <code>None</code>
)
        
        <div class="doc-md-description">
          <p>The first sequence to compare</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>query</code></b>
            (<code><span title="collections.abc.Sequence">Sequence</span></code>, default:
                <code>None</code>
)
        
        <div class="doc-md-description">
          <p>The second sequence to compare</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>mutation_allowed</code></b>
            (<code>bool</code>, default:
                <code>False</code>
)
        
        <div class="doc-md-description">
          <p>Whether equivalent indices can exist at mutation sites</p>
        </div>
      </li>
  </ul>
      <p>Returns:
    The pair of indices where the sequences align.
        Ex: sequence1 = A B C D E F ...
            sequence2 = A B - D E F ...
        returns        [0,1,  3,4,5, ...],
                       [0,1,  2,3,4, ...]</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1046</span>
<span class="normal">1047</span>
<span class="normal">1048</span>
<span class="normal">1049</span>
<span class="normal">1050</span>
<span class="normal">1051</span>
<span class="normal">1052</span>
<span class="normal">1053</span>
<span class="normal">1054</span>
<span class="normal">1055</span>
<span class="normal">1056</span>
<span class="normal">1057</span>
<span class="normal">1058</span>
<span class="normal">1059</span>
<span class="normal">1060</span>
<span class="normal">1061</span>
<span class="normal">1062</span>
<span class="normal">1063</span>
<span class="normal">1064</span>
<span class="normal">1065</span>
<span class="normal">1066</span>
<span class="normal">1067</span>
<span class="normal">1068</span>
<span class="normal">1069</span>
<span class="normal">1070</span>
<span class="normal">1071</span>
<span class="normal">1072</span>
<span class="normal">1073</span>
<span class="normal">1074</span>
<span class="normal">1075</span>
<span class="normal">1076</span>
<span class="normal">1077</span>
<span class="normal">1078</span>
<span class="normal">1079</span>
<span class="normal">1080</span>
<span class="normal">1081</span>
<span class="normal">1082</span>
<span class="normal">1083</span>
<span class="normal">1084</span>
<span class="normal">1085</span>
<span class="normal">1086</span>
<span class="normal">1087</span>
<span class="normal">1088</span>
<span class="normal">1089</span>
<span class="normal">1090</span>
<span class="normal">1091</span>
<span class="normal">1092</span>
<span class="normal">1093</span>
<span class="normal">1094</span>
<span class="normal">1095</span>
<span class="normal">1096</span>
<span class="normal">1097</span>
<span class="normal">1098</span>
<span class="normal">1099</span>
<span class="normal">1100</span>
<span class="normal">1101</span>
<span class="normal">1102</span>
<span class="normal">1103</span>
<span class="normal">1104</span>
<span class="normal">1105</span>
<span class="normal">1106</span>
<span class="normal">1107</span>
<span class="normal">1108</span>
<span class="normal">1109</span>
<span class="normal">1110</span>
<span class="normal">1111</span>
<span class="normal">1112</span>
<span class="normal">1113</span>
<span class="normal">1114</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_equivalent_indices</span><span class="p">(</span><span class="n">target</span><span class="p">:</span> <span class="n">Sequence</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">query</span><span class="p">:</span> <span class="n">Sequence</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">mutation_allowed</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> \
        <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">list</span><span class="p">[</span><span class="nb">int</span><span class="p">],</span> <span class="nb">list</span><span class="p">[</span><span class="nb">int</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;From two sequences, find the indices where both sequences are equal</span>

<span class="sd">    Args:</span>
<span class="sd">        target: The first sequence to compare</span>
<span class="sd">        query: The second sequence to compare</span>
<span class="sd">        mutation_allowed: Whether equivalent indices can exist at mutation sites</span>
<span class="sd">    Returns:</span>
<span class="sd">        The pair of indices where the sequences align.</span>
<span class="sd">            Ex: sequence1 = A B C D E F ...</span>
<span class="sd">                sequence2 = A B - D E F ...</span>
<span class="sd">            returns        [0,1,  3,4,5, ...],</span>
<span class="sd">                           [0,1,  2,3,4, ...]</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># alignment: Sequence = None</span>
    <span class="c1">#     alignment: An existing Bio.Align.Alignment object</span>
    <span class="c1"># if alignment is None:</span>
    <span class="k">if</span> <span class="n">target</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">query</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># # Get all mutations from the alignment of sequence1 and sequence2</span>
        <span class="n">mutations</span> <span class="o">=</span> <span class="n">generate_mutations</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">query</span><span class="p">,</span> <span class="n">keep_gaps</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">return_all</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="c1"># alignment = generate_alignment(target, query)</span>
        <span class="c1"># alignment.inverse_indices</span>
        <span class="c1"># return</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Can&#39;t </span><span class="si">{</span><span class="n">get_equivalent_indices</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2"> without passing either &#39;alignment&#39; or &quot;</span>
                         <span class="sa">f</span><span class="s2">&quot;&#39;target&#39; and &#39;query&#39;&quot;</span><span class="p">)</span>
    <span class="c1"># else:  # Todo this may not be ever useful since the alignment needs to go into the generate_mutations()</span>
    <span class="c1">#     raise NotImplementedError(f&quot;Set {get_equivalent_indices.__name__} up with an Alignment object from Bio.Align&quot;)</span>

    <span class="n">target_mutations</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">mutation</span><span class="p">[</span><span class="s1">&#39;to&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">mutations</span><span class="o">.</span><span class="n">values</span><span class="p">()])</span>
    <span class="n">query_mutations</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">mutation</span><span class="p">[</span><span class="s1">&#39;from&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">mutations</span><span class="o">.</span><span class="n">values</span><span class="p">()])</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Sequence info:</span><span class="se">\n</span><span class="s2">target :</span><span class="si">{</span><span class="n">target_mutations</span><span class="si">}</span><span class="se">\n</span><span class="s2">query  :</span><span class="si">{</span><span class="n">query_mutations</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="c1"># Get only those indices where there is an aligned aa on the opposite chain</span>
    <span class="n">sequence1_indices</span><span class="p">,</span> <span class="n">sequence2_indices</span> <span class="o">=</span> <span class="p">[],</span> <span class="p">[]</span>
    <span class="c1"># to_idx = from_idx = 0</span>
    <span class="n">to_idx</span><span class="p">,</span> <span class="n">from_idx</span> <span class="o">=</span> <span class="n">count</span><span class="p">(),</span> <span class="n">count</span><span class="p">()</span>
    <span class="c1"># sequence1 &#39;from&#39; is fixed, sequence2 &#39;to&#39; is moving</span>
    <span class="k">for</span> <span class="n">mutation_idx</span><span class="p">,</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">mutations</span><span class="o">.</span><span class="n">values</span><span class="p">()):</span>
        <span class="k">if</span> <span class="n">mutation</span><span class="p">[</span><span class="s1">&#39;to&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">mutation</span><span class="p">[</span><span class="s1">&#39;from&#39;</span><span class="p">]:</span>  <span class="c1"># They are equal</span>
            <span class="n">sequence1_indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">next</span><span class="p">(</span><span class="n">from_idx</span><span class="p">))</span>  <span class="c1"># from_idx)</span>
            <span class="n">sequence2_indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">next</span><span class="p">(</span><span class="n">to_idx</span><span class="p">))</span>  <span class="c1"># to_idx)</span>
        <span class="k">elif</span> <span class="n">mutation</span><span class="p">[</span><span class="s1">&#39;from&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;-&#39;</span><span class="p">:</span>  <span class="c1"># increment to_idx/fixed_idx</span>
            <span class="c1"># to_idx += 1</span>
            <span class="nb">next</span><span class="p">(</span><span class="n">to_idx</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">mutation</span><span class="p">[</span><span class="s1">&#39;to&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;-&#39;</span><span class="p">:</span>  <span class="c1"># increment from_idx/moving_idx</span>
            <span class="c1"># from_idx += 1</span>
            <span class="nb">next</span><span class="p">(</span><span class="n">from_idx</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">mutation</span><span class="p">[</span><span class="s1">&#39;to&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="n">mutation</span><span class="p">[</span><span class="s1">&#39;from&#39;</span><span class="p">]:</span>
            <span class="k">if</span> <span class="n">mutation_allowed</span><span class="p">:</span>
                <span class="n">sequence1_indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">next</span><span class="p">(</span><span class="n">from_idx</span><span class="p">))</span>
                <span class="n">sequence2_indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">next</span><span class="p">(</span><span class="n">to_idx</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="nb">next</span><span class="p">(</span><span class="n">from_idx</span><span class="p">)</span>
                <span class="nb">next</span><span class="p">(</span><span class="n">to_idx</span><span class="p">)</span>
            <span class="c1"># to_idx += 1</span>
            <span class="c1"># from_idx += 1</span>
        <span class="k">else</span><span class="p">:</span>  <span class="c1"># What else is there</span>
            <span class="n">target_mutations</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">mutation</span><span class="p">[</span><span class="s1">&#39;to&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">mutations</span><span class="o">.</span><span class="n">values</span><span class="p">()])</span>
            <span class="n">query_mutations</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">mutation</span><span class="p">[</span><span class="s1">&#39;from&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">mutations</span><span class="o">.</span><span class="n">values</span><span class="p">()])</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;This should never be reached. Ran into error at index </span><span class="si">{</span><span class="n">mutation_idx</span><span class="si">}</span><span class="s2">:</span><span class="se">\n</span><span class="s2">&quot;</span>
                               <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">mutation</span><span class="si">}</span><span class="se">\n</span><span class="s2">Sequence info:</span><span class="se">\n</span><span class="s2">target :</span><span class="si">{</span><span class="n">target_mutations</span><span class="si">}</span><span class="se">\n</span><span class="s2">query  :</span><span class="si">{</span><span class="n">query_mutations</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="c1"># else:  # They are equal</span>
        <span class="c1">#     sequence1_indices.append(next(from_idx))  # from_idx)</span>
        <span class="c1">#     sequence2_indices.append(next(to_idx))  # to_idx)</span>
        <span class="c1">#     # to_idx += 1</span>
        <span class="c1">#     # from_idx += 1</span>

    <span class="k">return</span> <span class="n">sequence1_indices</span><span class="p">,</span> <span class="n">sequence2_indices</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="sequence.get_lod" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">get_lod</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">get_lod</span><span class="p">(</span><span class="n">frequencies</span><span class="p">:</span> <span class="n">dict</span><span class="p">[</span><span class="n"><span title="sequence.protein_letters_literal">protein_letters_literal</span></span><span class="p">,</span> <span class="n">float</span><span class="p">],</span> <span class="n">background</span><span class="p">:</span> <span class="n">dict</span><span class="p">[</span><span class="n"><span title="sequence.protein_letters_literal">protein_letters_literal</span></span><span class="p">,</span> <span class="n">float</span><span class="p">],</span> <span class="n">as_int</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">dict</span><span class="p">[</span><span class="n">str</span><span class="p">,</span> <span class="n">int</span> <span class="o">|</span> <span class="n">float</span><span class="p">]</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Get the log of the odds that an amino acid is in a frequency distribution compared to a background frequency</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>frequencies</code></b>
            (<code>dict[<span title="sequence.protein_letters_literal">protein_letters_literal</span>, float]</code>)
        
        <div class="doc-md-description">
          <p>{'A': 0.11, 'C': 0.01, 'D': 0.034, ...}</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>background</code></b>
            (<code>dict[<span title="sequence.protein_letters_literal">protein_letters_literal</span>, float]</code>)
        
        <div class="doc-md-description">
          <p>{'A': 0.10, 'C': 0.02, 'D': 0.04, ...}</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>as_int</code></b>
            (<code>bool</code>, default:
                <code>True</code>
)
        
        <div class="doc-md-description">
          <p>Whether to round the lod values to an integer</p>
        </div>
      </li>
  </ul>
      <p>Returns:
     The log of odds for each amino acid type {'A': 2, 'C': -9, 'D': -1, ...}</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1123</span>
<span class="normal">1124</span>
<span class="normal">1125</span>
<span class="normal">1126</span>
<span class="normal">1127</span>
<span class="normal">1128</span>
<span class="normal">1129</span>
<span class="normal">1130</span>
<span class="normal">1131</span>
<span class="normal">1132</span>
<span class="normal">1133</span>
<span class="normal">1134</span>
<span class="normal">1135</span>
<span class="normal">1136</span>
<span class="normal">1137</span>
<span class="normal">1138</span>
<span class="normal">1139</span>
<span class="normal">1140</span>
<span class="normal">1141</span>
<span class="normal">1142</span>
<span class="normal">1143</span>
<span class="normal">1144</span>
<span class="normal">1145</span>
<span class="normal">1146</span>
<span class="normal">1147</span>
<span class="normal">1148</span>
<span class="normal">1149</span>
<span class="normal">1150</span>
<span class="normal">1151</span>
<span class="normal">1152</span>
<span class="normal">1153</span>
<span class="normal">1154</span>
<span class="normal">1155</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_lod</span><span class="p">(</span><span class="n">frequencies</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="n">protein_letters_literal</span><span class="p">,</span> <span class="nb">float</span><span class="p">],</span>
            <span class="n">background</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="n">protein_letters_literal</span><span class="p">,</span> <span class="nb">float</span><span class="p">],</span> <span class="n">as_int</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span> <span class="o">|</span> <span class="nb">float</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get the log of the odds that an amino acid is in a frequency distribution compared to a background frequency</span>

<span class="sd">    Args:</span>
<span class="sd">        frequencies: {&#39;A&#39;: 0.11, &#39;C&#39;: 0.01, &#39;D&#39;: 0.034, ...}</span>
<span class="sd">        background: {&#39;A&#39;: 0.10, &#39;C&#39;: 0.02, &#39;D&#39;: 0.04, ...}</span>
<span class="sd">        as_int: Whether to round the lod values to an integer</span>
<span class="sd">    Returns:</span>
<span class="sd">         The log of odds for each amino acid type {&#39;A&#39;: 2, &#39;C&#39;: -9, &#39;D&#39;: -1, ...}</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">lods</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">aa</span><span class="p">,</span> <span class="n">freq</span> <span class="ow">in</span> <span class="n">frequencies</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">try</span><span class="p">:</span>  <span class="c1"># Todo why is this 2. * the log2? I believe this is a heuristic of BLOSUM62. This should be removed...</span>
            <span class="n">lods</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="mf">2.</span> <span class="o">*</span> <span class="n">log2</span><span class="p">(</span><span class="n">freq</span> <span class="o">/</span> <span class="n">background</span><span class="p">[</span><span class="n">aa</span><span class="p">]))</span>  <span class="c1"># + 0.0</span>
        <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>  <span class="c1"># math domain error</span>
            <span class="n">lods</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">9</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">protein_letters_alph1</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">aa</span><span class="si">}</span><span class="s1"> was not in the background frequencies: </span><span class="si">{</span><span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">background</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>  <span class="c1"># We shouldn&#39;t worry about a missing value if it&#39;s not an amino acid</span>
                <span class="k">continue</span>
        <span class="k">except</span> <span class="ne">ZeroDivisionError</span><span class="p">:</span>  <span class="c1"># background is 0. We may need a pseudocount...</span>
            <span class="k">raise</span> <span class="ne">ZeroDivisionError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">aa</span><span class="si">}</span><span class="s1"> has a background frequency of 0. Consider adding a pseudocount&#39;</span><span class="p">)</span>
        <span class="c1"># if lods[aa] &lt; -9:</span>
        <span class="c1">#     lods[aa] = -9</span>
        <span class="c1"># elif round_lod:</span>
        <span class="c1">#     lods[aa] = round(lods[aa])</span>

    <span class="k">if</span> <span class="n">as_int</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">{</span><span class="n">aa</span><span class="p">:</span> <span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">value</span><span class="p">)</span> <span class="k">if</span> <span class="n">value</span> <span class="o">&gt;=</span> <span class="o">-</span><span class="mi">9</span> <span class="k">else</span> <span class="o">-</span><span class="mi">9</span><span class="p">)</span> <span class="k">for</span> <span class="n">aa</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">lods</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
    <span class="k">else</span><span class="p">:</span>  <span class="c1"># ensure that -9 is the lowest value (formatting issues if 2 digits)</span>
        <span class="k">return</span> <span class="p">{</span><span class="n">aa</span><span class="p">:</span> <span class="p">(</span><span class="n">value</span> <span class="k">if</span> <span class="n">value</span> <span class="o">&gt;=</span> <span class="o">-</span><span class="mi">9</span> <span class="k">else</span> <span class="o">-</span><span class="mi">9</span><span class="p">)</span> <span class="k">for</span> <span class="n">aa</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">lods</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="sequence.parse_pssm" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">parse_pssm</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">parse_pssm</span><span class="p">(</span><span class="n">file</span><span class="p">:</span> <span class="n"><span title="typing.AnyStr">AnyStr</span></span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">dict</span><span class="p">[</span><span class="n">int</span><span class="p">,</span> <span class="n">dict</span><span class="p">[</span><span class="n">str</span><span class="p">,</span> <span class="n">str</span> <span class="o">|</span> <span class="n">float</span> <span class="o">|</span> <span class="n">int</span> <span class="o">|</span> <span class="n">dict</span><span class="p">[</span><span class="n">str</span><span class="p">,</span> <span class="n">int</span><span class="p">]]]</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Take the contents of a pssm file, parse, and input into a pose profile dictionary.</p>
<p>Resulting dictionary is indexed according to the values in the pssm file</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>file</code></b>
            (<code><span title="typing.AnyStr">AnyStr</span></code>)
        
        <div class="doc-md-description">
          <p>The location of the file on disk</p>
        </div>
      </li>
  </ul>
      <p>Returns:
    Dictionary containing residue indexed profile information
        i.e. {1: {'A': 0, 'R': 0, ..., 'lod': {'A': -5, 'R': -5, ...}, 'type': 'W', 'info': 3.20, 'weight': 0.73},
              2: {}, ...}</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1158</span>
<span class="normal">1159</span>
<span class="normal">1160</span>
<span class="normal">1161</span>
<span class="normal">1162</span>
<span class="normal">1163</span>
<span class="normal">1164</span>
<span class="normal">1165</span>
<span class="normal">1166</span>
<span class="normal">1167</span>
<span class="normal">1168</span>
<span class="normal">1169</span>
<span class="normal">1170</span>
<span class="normal">1171</span>
<span class="normal">1172</span>
<span class="normal">1173</span>
<span class="normal">1174</span>
<span class="normal">1175</span>
<span class="normal">1176</span>
<span class="normal">1177</span>
<span class="normal">1178</span>
<span class="normal">1179</span>
<span class="normal">1180</span>
<span class="normal">1181</span>
<span class="normal">1182</span>
<span class="normal">1183</span>
<span class="normal">1184</span>
<span class="normal">1185</span>
<span class="normal">1186</span>
<span class="normal">1187</span>
<span class="normal">1188</span>
<span class="normal">1189</span>
<span class="normal">1190</span>
<span class="normal">1191</span>
<span class="normal">1192</span>
<span class="normal">1193</span>
<span class="normal">1194</span>
<span class="normal">1195</span>
<span class="normal">1196</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">parse_pssm</span><span class="p">(</span><span class="n">file</span><span class="p">:</span> <span class="n">AnyStr</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">float</span> <span class="o">|</span> <span class="nb">int</span> <span class="o">|</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Take the contents of a pssm file, parse, and input into a pose profile dictionary.</span>

<span class="sd">    Resulting dictionary is indexed according to the values in the pssm file</span>

<span class="sd">    Args:</span>
<span class="sd">        file: The location of the file on disk</span>
<span class="sd">    Returns:</span>
<span class="sd">        Dictionary containing residue indexed profile information</span>
<span class="sd">            i.e. {1: {&#39;A&#39;: 0, &#39;R&#39;: 0, ..., &#39;lod&#39;: {&#39;A&#39;: -5, &#39;R&#39;: -5, ...}, &#39;type&#39;: &#39;W&#39;, &#39;info&#39;: 3.20, &#39;weight&#39;: 0.73},</span>
<span class="sd">                  2: {}, ...}</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">lines</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>

    <span class="n">pose_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">lines</span><span class="p">:</span>
        <span class="n">line_data</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">line_data</span><span class="p">)</span> <span class="o">==</span> <span class="mi">44</span><span class="p">:</span>
            <span class="n">residue_number</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">line_data</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">pose_dict</span><span class="p">[</span><span class="n">residue_number</span><span class="p">]</span> <span class="o">=</span> \
                <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">protein_letters_alph3</span><span class="p">,</span>
                         <span class="p">[</span><span class="n">x</span> <span class="o">/</span> <span class="mf">100.</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">map</span><span class="p">(</span><span class="nb">int</span><span class="p">,</span> <span class="n">line_data</span><span class="p">[</span><span class="mi">22</span><span class="p">:</span><span class="nb">len</span><span class="p">(</span>
                             <span class="n">protein_letters_alph3</span><span class="p">)</span> <span class="o">+</span> <span class="mi">22</span><span class="p">])]))</span>
            <span class="c1"># pose_dict[residue_number] = aa_counts_alph3.copy()</span>
            <span class="c1"># for i, aa in enumerate(protein_letters_alph3, 22):</span>
            <span class="c1">#     # Get normalized counts for pose_dict</span>
            <span class="c1">#     pose_dict[residue_number][aa] = int(line_data[i]) / 100.</span>

            <span class="c1"># for i, aa in enumerate(protein_letters_alph3, 2):</span>
            <span class="c1">#     pose_dict[residue_number][&#39;lod&#39;][aa] = line_data[i]</span>
            <span class="n">pose_dict</span><span class="p">[</span><span class="n">residue_number</span><span class="p">][</span><span class="s1">&#39;lod&#39;</span><span class="p">]</span> <span class="o">=</span> \
                <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">protein_letters_alph3</span><span class="p">,</span> <span class="n">line_data</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="nb">len</span><span class="p">(</span>
                    <span class="n">protein_letters_alph3</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">]))</span>
            <span class="n">pose_dict</span><span class="p">[</span><span class="n">residue_number</span><span class="p">][</span><span class="s1">&#39;type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">line_data</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">pose_dict</span><span class="p">[</span><span class="n">residue_number</span><span class="p">][</span><span class="s1">&#39;info&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">line_data</span><span class="p">[</span><span class="mi">42</span><span class="p">])</span>
            <span class="n">pose_dict</span><span class="p">[</span><span class="n">residue_number</span><span class="p">][</span><span class="s1">&#39;weight&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">line_data</span><span class="p">[</span><span class="mi">43</span><span class="p">])</span>

    <span class="k">return</span> <span class="n">pose_dict</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="sequence.parse_hhblits_pssm" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">parse_hhblits_pssm</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">parse_hhblits_pssm</span><span class="p">(</span><span class="n">file</span><span class="p">:</span> <span class="n"><span title="typing.AnyStr">AnyStr</span></span><span class="p">,</span> <span class="n">null_background</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n"><a class="autorefs autorefs-internal" title="sequence.ProfileDict" href="#sequence.ProfileDict">ProfileDict</a></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Take contents of protein.hmm, parse file and input into pose_dict. File is Single AA code alphabetical order</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>file</code></b>
            (<code><span title="typing.AnyStr">AnyStr</span></code>)
        
        <div class="doc-md-description">
          <p>The file to parse, typically with the extension '.hmm'</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>null_background</code></b>
            (<code>bool</code>, default:
                <code>True</code>
)
        
        <div class="doc-md-description">
          <p>Whether to use the null background for the specific protein</p>
        </div>
      </li>
  </ul>
      <p>Returns:
    Dictionary containing residue indexed profile information
        Ex: {1: {'A': 0.04, 'C': 0.12, ..., 'lod': {'A': -5, 'C': -9, ...}, 'type': 'W', 'info': 0.00,
                 'weight': 0.00}, {...}}</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/sequence/__init__.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1199</span>
<span class="normal">1200</span>
<span class="normal">1201</span>
<span class="normal">1202</span>
<span class="normal">1203</span>
<span class="normal">1204</span>
<span class="normal">1205</span>
<span class="normal">1206</span>
<span class="normal">1207</span>
<span class="normal">1208</span>
<span class="normal">1209</span>
<span class="normal">1210</span>
<span class="normal">1211</span>
<span class="normal">1212</span>
<span class="normal">1213</span>
<span class="normal">1214</span>
<span class="normal">1215</span>
<span class="normal">1216</span>
<span class="normal">1217</span>
<span class="normal">1218</span>
<span class="normal">1219</span>
<span class="normal">1220</span>
<span class="normal">1221</span>
<span class="normal">1222</span>
<span class="normal">1223</span>
<span class="normal">1224</span>
<span class="normal">1225</span>
<span class="normal">1226</span>
<span class="normal">1227</span>
<span class="normal">1228</span>
<span class="normal">1229</span>
<span class="normal">1230</span>
<span class="normal">1231</span>
<span class="normal">1232</span>
<span class="normal">1233</span>
<span class="normal">1234</span>
<span class="normal">1235</span>
<span class="normal">1236</span>
<span class="normal">1237</span>
<span class="normal">1238</span>
<span class="normal">1239</span>
<span class="normal">1240</span>
<span class="normal">1241</span>
<span class="normal">1242</span>
<span class="normal">1243</span>
<span class="normal">1244</span>
<span class="normal">1245</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">parse_hhblits_pssm</span><span class="p">(</span><span class="n">file</span><span class="p">:</span> <span class="n">AnyStr</span><span class="p">,</span> <span class="n">null_background</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">ProfileDict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Take contents of protein.hmm, parse file and input into pose_dict. File is Single AA code alphabetical order</span>

<span class="sd">    Args:</span>
<span class="sd">        file: The file to parse, typically with the extension &#39;.hmm&#39;</span>
<span class="sd">        null_background: Whether to use the null background for the specific protein</span>
<span class="sd">    Returns:</span>
<span class="sd">        Dictionary containing residue indexed profile information</span>
<span class="sd">            Ex: {1: {&#39;A&#39;: 0.04, &#39;C&#39;: 0.12, ..., &#39;lod&#39;: {&#39;A&#39;: -5, &#39;C&#39;: -9, ...}, &#39;type&#39;: &#39;W&#39;, &#39;info&#39;: 0.00,</span>
<span class="sd">                     &#39;weight&#39;: 0.00}, {...}}</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">null_bg</span> <span class="o">=</span> <span class="n">latest_uniclust_background_frequencies</span>

    <span class="k">def</span> <span class="nf">to_freq</span><span class="p">(</span><span class="n">value</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">value</span> <span class="o">==</span> <span class="s1">&#39;*&#39;</span><span class="p">:</span>  <span class="c1"># When frequency is zero</span>
            <span class="k">return</span> <span class="mf">0.0001</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Equation: value = -1000 * log_2(frequency)</span>
            <span class="k">return</span> <span class="mi">2</span> <span class="o">**</span> <span class="p">(</span><span class="o">-</span><span class="nb">int</span><span class="p">(</span><span class="n">value</span><span class="p">)</span> <span class="o">/</span> <span class="mi">1000</span><span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">lines</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>

    <span class="n">evolutionary_profile</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">dummy</span> <span class="o">=</span> <span class="mf">0.</span>
    <span class="n">read</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">lines</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">read</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;#&#39;</span><span class="p">:</span>
                <span class="n">read</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">4</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;NULL&#39;</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">null_background</span><span class="p">:</span>  <span class="c1"># Use the provided null background from the profile search</span>
                    <span class="n">null</span><span class="p">,</span> <span class="o">*</span><span class="n">background_values</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
                    <span class="c1"># null = &#39;NULL&#39;, background_values = list[str] [&#39;3706&#39;, &#39;5728&#39;, ...]</span>
                    <span class="n">null_bg</span> <span class="o">=</span> <span class="p">{</span><span class="n">aa</span><span class="p">:</span> <span class="n">to_freq</span><span class="p">(</span><span class="n">value</span><span class="p">)</span> <span class="k">for</span> <span class="n">value</span><span class="p">,</span> <span class="n">aa</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">background_values</span><span class="p">,</span>
                                                                       <span class="n">protein_letters_alph3</span><span class="p">)}</span>

            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">())</span> <span class="o">==</span> <span class="mi">23</span><span class="p">:</span>
                <span class="n">residue_type</span><span class="p">,</span> <span class="n">residue_number</span><span class="p">,</span> <span class="o">*</span><span class="n">position_values</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
                <span class="n">aa_freqs</span> <span class="o">=</span> <span class="p">{</span><span class="n">aa</span><span class="p">:</span> <span class="n">to_freq</span><span class="p">(</span><span class="n">value</span><span class="p">)</span> <span class="k">for</span> <span class="n">value</span><span class="p">,</span> <span class="n">aa</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">position_values</span><span class="p">,</span>
                                                                    <span class="n">protein_letters_alph1</span><span class="p">)}</span>

                <span class="n">evolutionary_profile</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">residue_number</span><span class="p">)]</span> <span class="o">=</span> \
                    <span class="nb">dict</span><span class="p">(</span><span class="n">lod</span><span class="o">=</span><span class="n">get_lod</span><span class="p">(</span><span class="n">aa_freqs</span><span class="p">,</span> <span class="n">null_bg</span><span class="p">),</span> <span class="nb">type</span><span class="o">=</span><span class="n">residue_type</span><span class="p">,</span> <span class="n">info</span><span class="o">=</span><span class="n">dummy</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="n">dummy</span><span class="p">,</span> <span class="o">**</span><span class="n">aa_freqs</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">evolutionary_profile</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>

</div>












                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var tab,labels=set.querySelector(".tabbed-labels");for(tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../resources/wrapapi/" class="md-footer__link md-footer__link--prev" aria-label="Previous: wrapapi">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                wrapapi
              </div>
            </div>
          </a>
        
        
          
          <a href="constants/" class="md-footer__link md-footer__link--next" aria-label="Next: constants">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                constants
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; 2024 Kyle Meador
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        <div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://www.linkedin.com/in/kylemeador" target="_blank" rel="noopener" title="www.linkedin.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3zM135.4 416H69V202.2h66.5V416zm-33.2-243c-21.3 0-38.5-17.3-38.5-38.5S80.9 96 102.2 96c21.2 0 38.5 17.3 38.5 38.5 0 21.3-17.2 38.5-38.5 38.5zm282.1 243h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9V416z"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tabs", "navigation.tabs.sticky", "navigation.tracking", "navigation.instant", "navigation.path", "navigation.top", "navigation.footer", "search.suggest", "search.highlight", "toc.follow", "content.tabs.link", "content.code.annotation", "content.code.copy"], "search": "../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.8fd75fb4.min.js"></script>
      
    
  </body>
</html>