
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../model/">
      
      
        <link rel="next" href="../utils/">
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.5.9">
    
    
      
        <title>sequence - symdesign Documentation</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.f2e4d321.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../../css/code_select.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="deep-purple" data-md-color-accent="green">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#structure.sequence" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="symdesign Documentation" class="md-header__button md-logo" aria-label="symdesign Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            symdesign Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              sequence
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="deep-purple" data-md-color-accent="green"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="deep-purple" data-md-color-accent="green"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m14.3 16-.7-2h-3.2l-.7 2H7.8L11 7h2l3.2 9h-1.9M20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69m-9.15 3.96h2.3L12 9l-1.15 3.65Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="deep-purple" data-md-color-accent="green"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/kylemeador/symdesign" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../.." class="md-tabs__link">
        
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../../user_guide/" class="md-tabs__link">
        
  
    
  
  User Guide

      </a>
    </li>
  

      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../../" class="md-tabs__link">
          
  
  API reference

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="symdesign Documentation" class="md-nav__button md-logo" aria-label="symdesign Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    symdesign Documentation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/kylemeador/symdesign" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../user_guide/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    User Guide
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="">
            
  
  <span class="md-ellipsis">
    API reference
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            API reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    symdesign
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../flags/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    flags
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_3" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../metrics/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    metrics
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3_3" id="__nav_3_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_3">
            <span class="md-nav__icon md-icon"></span>
            metrics
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../metrics/pose/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    pose
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_4" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../protocols/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    protocols
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3_4" id="__nav_3_4_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_4">
            <span class="md-nav__icon md-icon"></span>
            protocols
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../protocols/align/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    align
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../protocols/cluster/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    cluster
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../protocols/config/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    config
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../protocols/fragdock/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    fragdock
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../protocols/pose/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    pose
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../protocols/select/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    select
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../protocols/utils/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    utils
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_5" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../resources/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    resources
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3_5" id="__nav_3_5_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_5">
            <span class="md-nav__icon md-icon"></span>
            resources
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../resources/config/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    config
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../resources/database/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    database
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../resources/distribute/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    distribute
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../resources/job/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    job
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../resources/ml/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ml
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../resources/monomer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    monomer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../resources/multimer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    multimer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_5_8" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../resources/query/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    query
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3_5_8" id="__nav_3_5_8_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_3_5_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_5_8">
            <span class="md-nav__icon md-icon"></span>
            query
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../resources/query/pdb/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    pdb
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../resources/query/uniprot/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    uniprot
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../resources/sql/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    sql
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../resources/structure_db/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    structure_db
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../resources/wrapapi/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    wrapapi
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_6" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../sequence/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    sequence
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3_6" id="__nav_3_6_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_6">
            <span class="md-nav__icon md-icon"></span>
            sequence
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../sequence/constants/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    constants
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../sequence/expression/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    expression
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_7" checked>
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    structure
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3_7" id="__nav_3_7_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_7_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3_7">
            <span class="md-nav__icon md-icon"></span>
            structure
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../base/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    base
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../coordinates/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    coordinates
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_7_3" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../fragment/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    fragment
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3_7_3" id="__nav_3_7_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_3_7_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_7_3">
            <span class="md-nav__icon md-icon"></span>
            fragment
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../fragment/db/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    db
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../fragment/info/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    info
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../fragment/metrics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    metrics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../fragment/visuals/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    visuals
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../model/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    model
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    sequence
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    sequence
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#structure.sequence" class="md-nav__link">
    <span class="md-ellipsis">
      sequence
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.blank_profile_entry" class="md-nav__link">
    <span class="md-ellipsis">
      blank_profile_entry
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.aa_nan_counts_alph3" class="md-nav__link">
    <span class="md-ellipsis">
      aa_nan_counts_alph3
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.nan_profile_entry" class="md-nav__link">
    <span class="md-ellipsis">
      nan_profile_entry
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.Profile" class="md-nav__link">
    <span class="md-ellipsis">
      Profile
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Profile">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#structure.sequence.Profile.data" class="md-nav__link">
    <span class="md-ellipsis">
      data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.Profile.available_keys" class="md-nav__link">
    <span class="md-ellipsis">
      available_keys
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.Profile.lods" class="md-nav__link">
    <span class="md-ellipsis">
      lods
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.Profile.types" class="md-nav__link">
    <span class="md-ellipsis">
      types
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.Profile.weights" class="md-nav__link">
    <span class="md-ellipsis">
      weights
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.Profile.info" class="md-nav__link">
    <span class="md-ellipsis">
      info
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.Profile.as_array" class="md-nav__link">
    <span class="md-ellipsis">
      as_array
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.Profile.write" class="md-nav__link">
    <span class="md-ellipsis">
      write
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity" class="md-nav__link">
    <span class="md-ellipsis">
      GeneEntity
    </span>
  </a>
  
    <nav class="md-nav" aria-label="GeneEntity">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.log" class="md-nav__link">
    <span class="md-ellipsis">
      log
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.sequence" class="md-nav__link">
    <span class="md-ellipsis">
      sequence
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.reference_sequence" class="md-nav__link">
    <span class="md-ellipsis">
      reference_sequence
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.number_of_residues" class="md-nav__link">
    <span class="md-ellipsis">
      number_of_residues
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.evolutionary_profile" class="md-nav__link">
    <span class="md-ellipsis">
      evolutionary_profile
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.msa" class="md-nav__link">
    <span class="md-ellipsis">
      msa
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.sequence_numeric" class="md-nav__link">
    <span class="md-ellipsis">
      sequence_numeric
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.hydrophobic_collapse" class="md-nav__link">
    <span class="md-ellipsis">
      hydrophobic_collapse
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.add_evolutionary_profile" class="md-nav__link">
    <span class="md-ellipsis">
      add_evolutionary_profile
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.create_null_profile" class="md-nav__link">
    <span class="md-ellipsis">
      create_null_profile
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.create_null_entries" class="md-nav__link">
    <span class="md-ellipsis">
      create_null_entries
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.hhblits" class="md-nav__link">
    <span class="md-ellipsis">
      hhblits
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.add_msa_from_file" class="md-nav__link">
    <span class="md-ellipsis">
      add_msa_from_file
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.collapse_profile" class="md-nav__link">
    <span class="md-ellipsis">
      collapse_profile
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.direct_coupling_analysis" class="md-nav__link">
    <span class="md-ellipsis">
      direct_coupling_analysis
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.write_sequence_to_fasta" class="md-nav__link">
    <span class="md-ellipsis">
      write_sequence_to_fasta
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.sequence_to_one_hot" class="md-nav__link">
    <span class="md-ellipsis">
      sequence_to_one_hot
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.sequence_to_numeric" class="md-nav__link">
    <span class="md-ellipsis">
      sequence_to_numeric
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.sequences_to_numeric" class="md-nav__link">
    <span class="md-ellipsis">
      sequences_to_numeric
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.pssm_as_array" class="md-nav__link">
    <span class="md-ellipsis">
      pssm_as_array
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.concatenate_profile" class="md-nav__link">
    <span class="md-ellipsis">
      concatenate_profile
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.write_pssm_file" class="md-nav__link">
    <span class="md-ellipsis">
      write_pssm_file
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.format_frequencies" class="md-nav__link">
    <span class="md-ellipsis">
      format_frequencies
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.overlap_consensus" class="md-nav__link">
    <span class="md-ellipsis">
      overlap_consensus
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.get_cluster_dicts" class="md-nav__link">
    <span class="md-ellipsis">
      get_cluster_dicts
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.fragment_overlap" class="md-nav__link">
    <span class="md-ellipsis">
      fragment_overlap
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.offset_index" class="md-nav__link">
    <span class="md-ellipsis">
      offset_index
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.residue_object_to_number" class="md-nav__link">
    <span class="md-ellipsis">
      residue_object_to_number
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.convert_to_residue_cluster_map" class="md-nav__link">
    <span class="md-ellipsis">
      convert_to_residue_cluster_map
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.consensus_sequence" class="md-nav__link">
    <span class="md-ellipsis">
      consensus_sequence
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.sequence_difference" class="md-nav__link">
    <span class="md-ellipsis">
      sequence_difference
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.msa_from_dictionary" class="md-nav__link">
    <span class="md-ellipsis">
      msa_from_dictionary
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.msa_from_seq_records" class="md-nav__link">
    <span class="md-ellipsis">
      msa_from_seq_records
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.make_mutations_chain_agnostic" class="md-nav__link">
    <span class="md-ellipsis">
      make_mutations_chain_agnostic
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.simplify_mutation_dict" class="md-nav__link">
    <span class="md-ellipsis">
      simplify_mutation_dict
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.weave_mutation_dict" class="md-nav__link">
    <span class="md-ellipsis">
      weave_mutation_dict
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.clean_gaped_columns" class="md-nav__link">
    <span class="md-ellipsis">
      clean_gaped_columns
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.msa_to_prob_distribution" class="md-nav__link">
    <span class="md-ellipsis">
      msa_to_prob_distribution
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.window_score" class="md-nav__link">
    <span class="md-ellipsis">
      window_score
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.rank_possibilities" class="md-nav__link">
    <span class="md-ellipsis">
      rank_possibilities
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.multi_chain_alignment" class="md-nav__link">
    <span class="md-ellipsis">
      multi_chain_alignment
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../utils/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    utils
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_8" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../tools/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    tools
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3_8" id="__nav_3_8_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_8">
            <span class="md-nav__icon md-icon"></span>
            tools
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tools/BenchmarkCollapseHydrophobicityScale/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    BenchmarkCollapseHydrophobicityScale
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tools/ConcatenatePDBFiles/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ConcatenatePDBFiles
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tools/ProfileScripts/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ProfileScripts
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tools/SlurmControl/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    SlurmControl
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tools/distribute/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    distribute
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tools/fetch_commands_from_slurm_output/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    fetch_commands_from_slurm_output
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tools/format_commands/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    format_commands
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tools/format_pose_ids/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    format_pose_ids
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tools/generate_reference_docs/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    generate_reference_docs
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tools/get_array_numbers_from_identifiers/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    get_array_numbers_from_identifiers
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tools/list_files_in_directory/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    list_files_in_directory
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tools/list_overlap/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    list_overlap
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tools/models_to_multimodel/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    models_to_multimodel
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tools/profile_gpu_nodes/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    profile_gpu_nodes
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tools/retrieve_oligomers/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    retrieve_oligomers
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tools/retrieve_pdbs_by_advanced_query/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    retrieve_pdbs_by_advanced_query
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tools/write_indices_for_C1_docking/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    write_indices_for_C1_docking
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_9" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../utils/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    utils
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3_9" id="__nav_3_9_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_9">
            <span class="md-nav__icon md-icon"></span>
            utils
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../utils/SymEntry/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    SymEntry
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../utils/cluster/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    cluster
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../utils/guide/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    guide
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../utils/nanohedra_parsing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    nanohedra_parsing
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../utils/path/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    path
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../utils/query/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    query
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../utils/rosetta/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    rosetta
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../utils/symmetry/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    symmetry
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../utils/types/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    types
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_10" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../visualization/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    visualization
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3_10" id="__nav_3_10_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_10_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_10">
            <span class="md-nav__icon md-icon"></span>
            visualization
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../visualization/SymmetryMates/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    SymmetryMates
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../visualization/VisualizeUtils/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    VisualizeUtils
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../visualization/shapes/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    shapes
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#structure.sequence" class="md-nav__link">
    <span class="md-ellipsis">
      sequence
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.blank_profile_entry" class="md-nav__link">
    <span class="md-ellipsis">
      blank_profile_entry
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.aa_nan_counts_alph3" class="md-nav__link">
    <span class="md-ellipsis">
      aa_nan_counts_alph3
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.nan_profile_entry" class="md-nav__link">
    <span class="md-ellipsis">
      nan_profile_entry
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.Profile" class="md-nav__link">
    <span class="md-ellipsis">
      Profile
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Profile">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#structure.sequence.Profile.data" class="md-nav__link">
    <span class="md-ellipsis">
      data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.Profile.available_keys" class="md-nav__link">
    <span class="md-ellipsis">
      available_keys
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.Profile.lods" class="md-nav__link">
    <span class="md-ellipsis">
      lods
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.Profile.types" class="md-nav__link">
    <span class="md-ellipsis">
      types
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.Profile.weights" class="md-nav__link">
    <span class="md-ellipsis">
      weights
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.Profile.info" class="md-nav__link">
    <span class="md-ellipsis">
      info
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.Profile.as_array" class="md-nav__link">
    <span class="md-ellipsis">
      as_array
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.Profile.write" class="md-nav__link">
    <span class="md-ellipsis">
      write
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity" class="md-nav__link">
    <span class="md-ellipsis">
      GeneEntity
    </span>
  </a>
  
    <nav class="md-nav" aria-label="GeneEntity">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.log" class="md-nav__link">
    <span class="md-ellipsis">
      log
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.sequence" class="md-nav__link">
    <span class="md-ellipsis">
      sequence
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.reference_sequence" class="md-nav__link">
    <span class="md-ellipsis">
      reference_sequence
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.number_of_residues" class="md-nav__link">
    <span class="md-ellipsis">
      number_of_residues
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.evolutionary_profile" class="md-nav__link">
    <span class="md-ellipsis">
      evolutionary_profile
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.msa" class="md-nav__link">
    <span class="md-ellipsis">
      msa
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.sequence_numeric" class="md-nav__link">
    <span class="md-ellipsis">
      sequence_numeric
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.hydrophobic_collapse" class="md-nav__link">
    <span class="md-ellipsis">
      hydrophobic_collapse
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.add_evolutionary_profile" class="md-nav__link">
    <span class="md-ellipsis">
      add_evolutionary_profile
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.create_null_profile" class="md-nav__link">
    <span class="md-ellipsis">
      create_null_profile
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.create_null_entries" class="md-nav__link">
    <span class="md-ellipsis">
      create_null_entries
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.hhblits" class="md-nav__link">
    <span class="md-ellipsis">
      hhblits
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.add_msa_from_file" class="md-nav__link">
    <span class="md-ellipsis">
      add_msa_from_file
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.collapse_profile" class="md-nav__link">
    <span class="md-ellipsis">
      collapse_profile
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.direct_coupling_analysis" class="md-nav__link">
    <span class="md-ellipsis">
      direct_coupling_analysis
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#structure.sequence.GeneEntity.write_sequence_to_fasta" class="md-nav__link">
    <span class="md-ellipsis">
      write_sequence_to_fasta
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.sequence_to_one_hot" class="md-nav__link">
    <span class="md-ellipsis">
      sequence_to_one_hot
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.sequence_to_numeric" class="md-nav__link">
    <span class="md-ellipsis">
      sequence_to_numeric
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.sequences_to_numeric" class="md-nav__link">
    <span class="md-ellipsis">
      sequences_to_numeric
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.pssm_as_array" class="md-nav__link">
    <span class="md-ellipsis">
      pssm_as_array
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.concatenate_profile" class="md-nav__link">
    <span class="md-ellipsis">
      concatenate_profile
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.write_pssm_file" class="md-nav__link">
    <span class="md-ellipsis">
      write_pssm_file
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.format_frequencies" class="md-nav__link">
    <span class="md-ellipsis">
      format_frequencies
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.overlap_consensus" class="md-nav__link">
    <span class="md-ellipsis">
      overlap_consensus
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.get_cluster_dicts" class="md-nav__link">
    <span class="md-ellipsis">
      get_cluster_dicts
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.fragment_overlap" class="md-nav__link">
    <span class="md-ellipsis">
      fragment_overlap
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.offset_index" class="md-nav__link">
    <span class="md-ellipsis">
      offset_index
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.residue_object_to_number" class="md-nav__link">
    <span class="md-ellipsis">
      residue_object_to_number
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.convert_to_residue_cluster_map" class="md-nav__link">
    <span class="md-ellipsis">
      convert_to_residue_cluster_map
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.consensus_sequence" class="md-nav__link">
    <span class="md-ellipsis">
      consensus_sequence
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.sequence_difference" class="md-nav__link">
    <span class="md-ellipsis">
      sequence_difference
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.msa_from_dictionary" class="md-nav__link">
    <span class="md-ellipsis">
      msa_from_dictionary
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.msa_from_seq_records" class="md-nav__link">
    <span class="md-ellipsis">
      msa_from_seq_records
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.make_mutations_chain_agnostic" class="md-nav__link">
    <span class="md-ellipsis">
      make_mutations_chain_agnostic
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.simplify_mutation_dict" class="md-nav__link">
    <span class="md-ellipsis">
      simplify_mutation_dict
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.weave_mutation_dict" class="md-nav__link">
    <span class="md-ellipsis">
      weave_mutation_dict
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.clean_gaped_columns" class="md-nav__link">
    <span class="md-ellipsis">
      clean_gaped_columns
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.msa_to_prob_distribution" class="md-nav__link">
    <span class="md-ellipsis">
      msa_to_prob_distribution
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.window_score" class="md-nav__link">
    <span class="md-ellipsis">
      window_score
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.rank_possibilities" class="md-nav__link">
    <span class="md-ellipsis">
      rank_possibilities
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structure.sequence.multi_chain_alignment" class="md-nav__link">
    <span class="md-ellipsis">
      multi_chain_alignment
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


  <h1>sequence</h1>

<div class="doc doc-object doc-module">



<a id="structure.sequence"></a>
  <div class="doc doc-contents first">

  

  <div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h2 id="structure.sequence.blank_profile_entry" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">blank_profile_entry</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">blank_profile_entry</span> <span class="o">=</span> <span class="n"><span title="structure.sequence.aa_counts_alph3.copy">copy</span></span><span class="p">()</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>{utils.profile_keys, repeat(0))}</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h2 id="structure.sequence.aa_nan_counts_alph3" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">aa_nan_counts_alph3</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">aa_nan_counts_alph3</span> <span class="o">=</span> <span class="n">dict</span><span class="p">(</span><span class="n">zip</span><span class="p">(</span><span class="n"><span title="symdesign.sequence.protein_letters_alph3">protein_letters_alph3</span></span><span class="p">,</span> <span class="n"><span title="itertools.repeat">repeat</span></span><span class="p">(</span><span class="n"><span title="numpy.nan">nan</span></span><span class="p">)))</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>{protein_letters_alph3, repeat(numpy.nan))}</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h2 id="structure.sequence.nan_profile_entry" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">nan_profile_entry</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">nan_profile_entry</span> <span class="o">=</span> <span class="n"><span title="structure.sequence.aa_nan_counts_alph3.copy">copy</span></span><span class="p">()</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>{utils.profile_keys, repeat(numpy.nan))}</p>
  </div>

</div>


<div class="doc doc-object doc-class">



<h2 id="structure.sequence.Profile" class="doc doc-heading">
          <span class="doc doc-object-name doc-class-name">Profile</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">Profile</span><span class="p">(</span><span class="n">entries</span><span class="p">:</span> <span class="n"><span title="collections.abc.Iterable">Iterable</span></span><span class="p">[</span><span class="n"><span title="symdesign.sequence.ProfileEntry">ProfileEntry</span></span><span class="p">],</span> <span class="n">dtype</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="s1">&#39;profile&#39;</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</code></pre></div>

  <div class="doc doc-contents ">
          <p class="doc doc-class-bases">
            Bases: <code><span title="collections.UserList">UserList</span>[<span title="symdesign.sequence.ProfileEntry">ProfileEntry</span>]</code></p>

  



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>entries</code></b>
            (<code><span title="collections.abc.Iterable">Iterable</span>[<span title="symdesign.sequence.ProfileEntry">ProfileEntry</span>]</code>)
        –
        <div class="doc-md-description">
          <p>The per-residue entries to create the instance</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>dtype</code></b>
            (<code>str</code>, default:
                <code>&#39;profile&#39;</code>
)
        –
        <div class="doc-md-description">
          <p>The datatype of the profile.</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>**kwargs</code></b>
        –
        <div class="doc-md-description">
          
        </div>
      </li>
  </ul>

                <details class="quote">
                  <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
                  <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">entries</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="n">ProfileEntry</span><span class="p">],</span> <span class="n">dtype</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;profile&#39;</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Construct the instance</span>

<span class="sd">    Args:</span>
<span class="sd">        entries: The per-residue entries to create the instance</span>
<span class="sd">        dtype: The datatype of the profile.</span>
<span class="sd">        **kwargs:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">initlist</span><span class="o">=</span><span class="n">entries</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>  <span class="c1"># initlist sets UserList.data</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">dtype</span> <span class="o">=</span> <span class="n">dtype</span>
</code></pre></div></td></tr></table></div>
                </details>

  

  <div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h3 id="structure.sequence.Profile.data" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">data</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-instance-attribute"><code>instance-attribute</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">data</span><span class="p">:</span> <span class="n">list</span><span class="p">[</span><span class="n"><span title="symdesign.sequence.ProfileEntry">ProfileEntry</span></span><span class="p">]</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>[{'A': 0, 'R': 0, ..., 'lod': {'A': -5, 'R': -5, ...},
 'type': 'W', 'info': 3.20, 'weight': 0.73},
{}, ...]</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="structure.sequence.Profile.available_keys" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">available_keys</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">available_keys</span><span class="p">:</span> <span class="n">tuple</span><span class="p">[</span><span class="n"><span title="typing.Any">Any</span></span><span class="p">,</span> <span class="o">...</span><span class="p">]</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Returns the available ProfileEntry keys that are present in the Profile</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="structure.sequence.Profile.lods" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">lods</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">lods</span><span class="p">:</span> <span class="n">list</span><span class="p">[</span><span class="n"><span title="symdesign.sequence.AminoAcidDistribution">AminoAcidDistribution</span></span><span class="p">]</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>The log of odds values, given for each amino acid type, for each entry in the Profile</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="structure.sequence.Profile.types" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">types</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">types</span><span class="p">:</span> <span class="n">list</span><span class="p">[</span><span class="n"><span title="symdesign.sequence.protein_letters_literal">protein_letters_literal</span></span><span class="p">]</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>The amino acid type, for each entry in the Profile</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="structure.sequence.Profile.weights" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">weights</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">weights</span><span class="p">:</span> <span class="n">list</span><span class="p">[</span><span class="n">float</span><span class="p">]</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>The weight assigned to each entry in the Profile</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="structure.sequence.Profile.info" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">info</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">info</span><span class="p">:</span> <span class="n">list</span><span class="p">[</span><span class="n">float</span><span class="p">]</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>The information present for each entry in the Profile</p>
  </div>

</div>




<div class="doc doc-object doc-function">



<h3 id="structure.sequence.Profile.as_array" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">as_array</span>


</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">as_array</span><span class="p">(</span><span class="n">alphabet</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="n">protein_letters_alph1</span><span class="p">,</span> <span class="n">lod</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n"><span title="numpy.ndarray">ndarray</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Convert the Profile into a numeric array</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>alphabet</code></b>
            (<code>str</code>, default:
                <code><span title="symdesign.sequence.protein_letters_alph1">protein_letters_alph1</span></code>
)
        –
        <div class="doc-md-description">
          <p>The amino acid alphabet to use. Array values will be returned in this order</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>lod</code></b>
            (<code>bool</code>, default:
                <code>False</code>
)
        –
        <div class="doc-md-description">
          <p>Whether to return the array for the log of odds values</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
            <code><span title="numpy.ndarray">ndarray</span></code>
        –
        <div class="doc-md-description">
          <p>The numerically encoded pssm where each entry along axis 0 is the position, and the entries on axis 1 are
the frequency data at every indexed amino acid. Indices are according to the specified amino acid
alphabet, i.e. array([[0.1, 0.01, 0.12, ...], ...])</p>
      </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">as_array</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">alphabet</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">protein_letters_alph1</span><span class="p">,</span> <span class="n">lod</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Convert the Profile into a numeric array</span>

<span class="sd">    Args:</span>
<span class="sd">        alphabet: The amino acid alphabet to use. Array values will be returned in this order</span>
<span class="sd">        lod: Whether to return the array for the log of odds values</span>

<span class="sd">    Returns:</span>
<span class="sd">        The numerically encoded pssm where each entry along axis 0 is the position, and the entries on axis 1 are</span>
<span class="sd">            the frequency data at every indexed amino acid. Indices are according to the specified amino acid</span>
<span class="sd">            alphabet, i.e. array([[0.1, 0.01, 0.12, ...], ...])</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">lod</span><span class="p">:</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">lods</span><span class="p">:</span>
            <span class="n">data_source</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">lods</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;There aren&#39;t any values available for </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">.lods&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">data_source</span> <span class="o">=</span> <span class="bp">self</span>

    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="n">position_info</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">alphabet</span><span class="p">]</span>
                     <span class="k">for</span> <span class="n">position_info</span> <span class="ow">in</span> <span class="n">data_source</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="structure.sequence.Profile.write" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">write</span>


</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">write</span><span class="p">(</span><span class="n">file_name</span><span class="p">:</span> <span class="n"><span title="typing.AnyStr">AnyStr</span></span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">out_dir</span><span class="p">:</span> <span class="n"><span title="typing.AnyStr">AnyStr</span></span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">())</span> <span class="o">-&gt;</span> <span class="n"><span title="typing.AnyStr">AnyStr</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Create a PSI-BLAST format PSSM file from a PSSM dictionary. Assumes residue numbering is 1 to last entry</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>file_name</code></b>
            (<code><span title="typing.AnyStr">AnyStr</span></code>, default:
                <code>None</code>
)
        –
        <div class="doc-md-description">
          <p>The explicit name of the file</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>name</code></b>
            (<code>str</code>, default:
                <code>None</code>
)
        –
        <div class="doc-md-description">
          <p>The name of the file. Will be used as the default file_name base name if file_name not provided</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>out_dir</code></b>
            (<code><span title="typing.AnyStr">AnyStr</span></code>, default:
                <code><span title="os.getcwd">getcwd</span>()</code>
)
        –
        <div class="doc-md-description">
          <p>The location on disk to output the file. Only used if file_name not explicitly provided</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
            <code><span title="typing.AnyStr">AnyStr</span></code>
        –
        <div class="doc-md-description">
          <p>Disk location of newly created .pssm file</p>
      </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">write</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file_name</span><span class="p">:</span> <span class="n">AnyStr</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">out_dir</span><span class="p">:</span> <span class="n">AnyStr</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">())</span> <span class="o">-&gt;</span> <span class="n">AnyStr</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create a PSI-BLAST format PSSM file from a PSSM dictionary. Assumes residue numbering is 1 to last entry</span>

<span class="sd">    Args:</span>
<span class="sd">        file_name: The explicit name of the file</span>
<span class="sd">        name: The name of the file. Will be used as the default file_name base name if file_name not provided</span>
<span class="sd">        out_dir: The location on disk to output the file. Only used if file_name not explicitly provided</span>

<span class="sd">    Returns:</span>
<span class="sd">        Disk location of newly created .pssm file</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Format the Profile according to the write_pssm_file mechanism</span>
    <span class="c1"># This takes the shape of 1 to the last entry</span>
    <span class="n">data</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">list</span><span class="p">[</span><span class="nb">float</span><span class="p">]]</span> <span class="o">=</span> <span class="p">[[</span><span class="n">position_info</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">protein_letters_alph3</span><span class="p">]</span> <span class="k">for</span> <span class="n">position_info</span> <span class="ow">in</span> <span class="bp">self</span><span class="p">]</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">dtype</span> <span class="o">==</span> <span class="s1">&#39;fragment&#39;</span><span class="p">:</span>
        <span class="c1"># Need to convert np.nan to zeros</span>
        <span class="n">data</span> <span class="o">=</span> <span class="p">[[</span><span class="mi">0</span> <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span> <span class="k">else</span> <span class="n">num</span> <span class="k">for</span> <span class="n">num</span> <span class="ow">in</span> <span class="n">entry</span><span class="p">]</span> <span class="k">for</span> <span class="n">entry</span> <span class="ow">in</span> <span class="n">data</span><span class="p">]</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Converting </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">dtype</span><span class="si">}</span><span class="s1"> type Profile np.nan values to 0.0&#39;</span><span class="p">)</span>

    <span class="c1"># Find out if the pssm has values expressed as frequencies (percentages) or as counts and modify accordingly</span>
    <span class="n">lods</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">lods</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">lods</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;A&#39;</span><span class="p">],</span> <span class="nb">float</span><span class="p">):</span>
        <span class="n">separation1</span> <span class="o">=</span> <span class="s2">&quot; &quot;</span> <span class="o">*</span> <span class="mi">4</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">separation1</span> <span class="o">=</span> <span class="s2">&quot; &quot;</span> <span class="o">*</span> <span class="mi">3</span>
        <span class="c1"># lod_freq = True</span>
    <span class="c1"># if type(pssm[first_key][&#39;A&#39;]) == float:</span>
    <span class="c1">#     counts_freq = True</span>

    <span class="k">if</span> <span class="n">file_name</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">name</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Must provide argument &#39;file_name&#39; or &#39;name&#39; as a str to </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">write</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">file_name</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">out_dir</span><span class="p">,</span> <span class="n">name</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">file_name</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;&#39;</span><span class="p">:</span>  <span class="c1"># No extension</span>
        <span class="n">file_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">file_name</span><span class="si">}</span><span class="s1">.pssm&#39;</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="se">\n\n</span><span class="si">{</span><span class="s2">&quot; &quot;</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="mi">12</span><span class="si">}{</span><span class="n">separation1</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">protein_letters_alph3</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span>
                <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">separation1</span><span class="si">}{</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="mi">3</span><span class="p">)</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">protein_letters_alph3</span><span class="p">)</span><span class="si">}</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">residue_number</span><span class="p">,</span> <span class="p">(</span><span class="n">entry</span><span class="p">,</span> <span class="n">lod</span><span class="p">,</span> <span class="n">_type</span><span class="p">,</span> <span class="n">info</span><span class="p">,</span> <span class="n">weight</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span>
                <span class="nb">zip</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">lods</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">types</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">info</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">weights</span><span class="p">),</span> <span class="mi">1</span><span class="p">):</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">lod</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">],</span> <span class="nb">float</span><span class="p">):</span>  <span class="c1"># relevant for favor_fragment</span>
                <span class="n">lod_string</span> <span class="o">=</span> \
                    <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">lod</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span><span class="si">:</span><span class="s1">&gt;4.2f</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">protein_letters_alph3</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">lod_string</span> <span class="o">=</span> \
                    <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">lod</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span><span class="si">:</span><span class="s1">&gt;3d</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">protein_letters_alph3</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span>

            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">entry</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">float</span><span class="p">):</span>  <span class="c1"># relevant for freq calculations</span>
                <span class="n">counts_string</span> <span class="o">=</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">floor</span><span class="p">(</span><span class="n">value</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="mi">100</span><span class="p">)</span><span class="si">:</span><span class="s1">&gt;3.0f</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">entry</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">counts_string</span> <span class="o">=</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">value</span><span class="si">:</span><span class="s1">&gt;3d</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">entry</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span>

            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">residue_number</span><span class="si">:</span><span class="s1">&gt;5d</span><span class="si">}</span><span class="s1"> </span><span class="si">{</span><span class="n">_type</span><span class="si">:</span><span class="s1">1s</span><span class="si">}</span><span class="s1">   </span><span class="si">{</span><span class="n">lod_string</span><span class="si">:</span><span class="s1">80s</span><span class="si">}</span><span class="s1"> </span><span class="si">{</span><span class="n">counts_string</span><span class="si">:</span><span class="s1">80s</span><span class="si">}</span><span class="s1"> &#39;</span>
                    <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="nb">round</span><span class="p">(</span><span class="n">info</span><span class="p">,</span><span class="w"> </span><span class="mi">4</span><span class="p">)</span><span class="si">:</span><span class="s1">4.2f</span><span class="si">}</span><span class="s1"> </span><span class="si">{</span><span class="nb">round</span><span class="p">(</span><span class="n">weight</span><span class="p">,</span><span class="w"> </span><span class="mi">4</span><span class="p">)</span><span class="si">:</span><span class="s1">4.2f</span><span class="si">}</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">file_name</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>


</div>

<div class="doc doc-object doc-class">



<h2 id="structure.sequence.GeneEntity" class="doc doc-heading">
          <span class="doc doc-object-name doc-class-name">GeneEntity</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">GeneEntity</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</code></pre></div>

  <div class="doc doc-contents ">
          <p class="doc doc-class-bases">
            Bases: <code><span title="abc.ABC">ABC</span></code></p>

  
      <p>Contains the sequence information for a ContainsResidues.</p>
  



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>**kwargs</code></b>
        –
        <div class="doc-md-description">
          
        </div>
      </li>
  </ul>

                <details class="quote">
                  <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
                  <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Construct the instance</span>

<span class="sd">    Args:</span>
<span class="sd">        **kwargs:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>  <span class="c1"># GeneEntity</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_evolutionary_profile</span> <span class="o">=</span> <span class="p">{}</span>  <span class="c1"># position specific scoring matrix</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">h_fields</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">j_couplings</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">profile</span> <span class="o">=</span> <span class="p">{}</span>  <span class="c1"># design/structure specific scoring matrix</span>
</code></pre></div></td></tr></table></div>
                </details>

  

  <div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h3 id="structure.sequence.GeneEntity.log" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">log</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-abstractmethod"><code>abstractmethod</code></small>
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">log</span><span class="p">:</span> <span class="n"><span title="logging.Logger">Logger</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="structure.sequence.GeneEntity.sequence" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">sequence</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-abstractmethod"><code>abstractmethod</code></small>
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">sequence</span><span class="p">:</span> <span class="n">str</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="structure.sequence.GeneEntity.reference_sequence" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">reference_sequence</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-abstractmethod"><code>abstractmethod</code></small>
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">reference_sequence</span><span class="p">:</span> <span class="n">str</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="structure.sequence.GeneEntity.number_of_residues" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">number_of_residues</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">number_of_residues</span><span class="p">:</span> <span class="n">int</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="structure.sequence.GeneEntity.evolutionary_profile" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">evolutionary_profile</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
      <small class="doc doc-label doc-label-writable"><code>writable</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">evolutionary_profile</span><span class="p">:</span> <span class="n">dict</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Access the evolutionary_profile</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="structure.sequence.GeneEntity.msa" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">msa</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
      <small class="doc doc-label doc-label-writable"><code>writable</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">msa</span><span class="p">:</span> <span class="n"><span title="symdesign.sequence.MultipleSequenceAlignment">MultipleSequenceAlignment</span></span> <span class="o">|</span> <span class="kc">None</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>The MultipleSequenceAlignment object for the instance</p>
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="structure.sequence.GeneEntity.sequence_numeric" class="doc doc-heading">
          <span class="doc doc-object-name doc-attribute-name">sequence_numeric</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">sequence_numeric</span><span class="p">:</span> <span class="n"><span title="numpy.ndarray">ndarray</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Return the sequence as an integer array (number_of_residuces, alphabet_length) of the amino acid characters</p>
<p>Maps "ACDEFGHIKLMNPQRSTVWY-" to the resulting index</p>
  </div>

</div>




<div class="doc doc-object doc-function">



<h3 id="structure.sequence.GeneEntity.hydrophobic_collapse" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">hydrophobic_collapse</span>


</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">hydrophobic_collapse</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n"><span title="numpy.array">array</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Return the hydrophobic collapse for the Sequence</p>



  <p>Other Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>hydrophobicity</code></b>
        –
        <div class="doc-md-description">
          <p>int = 'standard' – The hydrophobicity scale to consider. Either 'standard' (FILV),
'expanded' (FMILYVW), or provide one with 'custom' keyword argument</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>custom</code></b>
        –
        <div class="doc-md-description">
          <p>mapping[str, float | int] = None – A user defined mapping of amino acid type, hydrophobicity value pairs</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>alphabet_type</code></b>
        –
        <div class="doc-md-description">
          <p>alphabet_types = None – The amino acid alphabet if the sequence consists of integer characters</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>lower_window</code></b>
        –
        <div class="doc-md-description">
          <p>int = 3 – The smallest window used to measure</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>upper_window</code></b>
        –
        <div class="doc-md-description">
          <p>int = 9 – The largest window used to measure</p>
        </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">hydrophobic_collapse</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Return the hydrophobic collapse for the Sequence</span>

<span class="sd">    Keyword Args:</span>
<span class="sd">        hydrophobicity: int = &#39;standard&#39; – The hydrophobicity scale to consider. Either &#39;standard&#39; (FILV),</span>
<span class="sd">            &#39;expanded&#39; (FMILYVW), or provide one with &#39;custom&#39; keyword argument</span>
<span class="sd">        custom: mapping[str, float | int] = None – A user defined mapping of amino acid type, hydrophobicity value pairs</span>
<span class="sd">        alphabet_type: alphabet_types = None – The amino acid alphabet if the sequence consists of integer characters</span>
<span class="sd">        lower_window: int = 3 – The smallest window used to measure</span>
<span class="sd">        upper_window: int = 9 – The largest window used to measure</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_hydrophobic_collapse</span>
    <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_hydrophobic_collapse</span> <span class="o">=</span> <span class="n">metrics</span><span class="o">.</span><span class="n">hydrophobic_collapse_index</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_hydrophobic_collapse</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="structure.sequence.GeneEntity.add_evolutionary_profile" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">add_evolutionary_profile</span>


</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">add_evolutionary_profile</span><span class="p">(</span><span class="n">file</span><span class="p">:</span> <span class="n"><span title="typing.AnyStr">AnyStr</span></span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">out_dir</span><span class="p">:</span> <span class="n"><span title="typing.AnyStr">AnyStr</span></span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">(),</span> <span class="n">profile_source</span><span class="p">:</span> <span class="n"><span title="symdesign.sequence.alignment_programs_literal">alignment_programs_literal</span></span> <span class="o">=</span> <span class="n">putils</span><span class="o">.</span><span class="n">hhblits</span><span class="p">,</span> <span class="n">force</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Add the evolutionary profile to the GeneEntity. If the profile isn't provided, it is generated through search
of homologous protein sequences using the profile_source argument</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>file</code></b>
            (<code><span title="typing.AnyStr">AnyStr</span></code>, default:
                <code>None</code>
)
        –
        <div class="doc-md-description">
          <p>Location where profile file should be loaded from</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>out_dir</code></b>
            (<code><span title="typing.AnyStr">AnyStr</span></code>, default:
                <code><span title="os.getcwd">getcwd</span>()</code>
)
        –
        <div class="doc-md-description">
          <p>Location where sequence files should be written</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>profile_source</code></b>
            (<code><span title="symdesign.sequence.alignment_programs_literal">alignment_programs_literal</span></code>, default:
                <code><span title="structure.sequence.putils.hhblits">hhblits</span></code>
)
        –
        <div class="doc-md-description">
          <p>One of 'hhblits' or 'psiblast'</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>force</code></b>
            (<code>bool</code>, default:
                <code>False</code>
)
        –
        <div class="doc-md-description">
          <p>Whether to force generation of a new profile</p>
        </div>
      </li>
  </ul>

<details class="sets" open>
  <summary>Sets</summary>
  <p>self.evolutionary_profile (ProfileDict)</p>
</details>
          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_evolutionary_profile</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file</span><span class="p">:</span> <span class="n">AnyStr</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">out_dir</span><span class="p">:</span> <span class="n">AnyStr</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">(),</span>
                             <span class="n">profile_source</span><span class="p">:</span> <span class="n">alignment_programs_literal</span> <span class="o">=</span> <span class="n">putils</span><span class="o">.</span><span class="n">hhblits</span><span class="p">,</span> <span class="n">force</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
                             <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Add the evolutionary profile to the GeneEntity. If the profile isn&#39;t provided, it is generated through search</span>
<span class="sd">    of homologous protein sequences using the profile_source argument</span>

<span class="sd">    Args:</span>
<span class="sd">        file: Location where profile file should be loaded from</span>
<span class="sd">        out_dir: Location where sequence files should be written</span>
<span class="sd">        profile_source: One of &#39;hhblits&#39; or &#39;psiblast&#39;</span>
<span class="sd">        force: Whether to force generation of a new profile</span>

<span class="sd">    Sets:</span>
<span class="sd">        self.evolutionary_profile (ProfileDict)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">profile_source</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">alignment_programs</span><span class="p">:</span>  <span class="c1"># [putils.hhblits, &#39;psiblast&#39;]:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">add_evolutionary_profile</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s1">: Profile generation only possible from &#39;</span>
            <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">alignment_programs</span><span class="p">)</span><span class="si">}</span><span class="s1">, not </span><span class="si">{</span><span class="n">profile_source</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">pssm_file</span> <span class="o">=</span> <span class="n">file</span>
    <span class="k">else</span><span class="p">:</span>  <span class="c1"># Check to see if the files of interest already exist</span>
        <span class="c1"># Extract/Format Sequence Information. SEQRES is prioritized if available</span>
        <span class="n">name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span>
        <span class="n">sequence_file</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">write_sequence_to_fasta</span><span class="p">(</span><span class="n">out_dir</span><span class="o">=</span><span class="n">out_dir</span><span class="p">)</span>
        <span class="n">temp_file</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">out_dir</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s1">.hold&#39;</span><span class="p">)</span>
        <span class="n">pssm_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">out_dir</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s1">.hmm&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">pssm_file</span><span class="p">)</span> <span class="ow">or</span> <span class="n">force</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">temp_file</span><span class="p">):</span>  <span class="c1"># No work on this pssm file has been initiated</span>
                <span class="c1"># Create blocking file to prevent excess work</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">temp_file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Fetching &#39;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&#39; sequence data&quot;</span><span class="p">)</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Generating Evolutionary Profile for </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
                <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">profile_source</span><span class="p">)(</span><span class="n">sequence_file</span><span class="o">=</span><span class="n">sequence_file</span><span class="p">,</span> <span class="n">out_dir</span><span class="o">=</span><span class="n">out_dir</span><span class="p">)</span>
                <span class="n">temp_file</span><span class="o">.</span><span class="n">unlink</span><span class="p">(</span><span class="n">missing_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                <span class="c1"># if os.path.exists(temp_file):</span>
                <span class="c1">#     os.remove(temp_file)</span>
            <span class="k">else</span><span class="p">:</span>  <span class="c1"># Block is in place, another process is working</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Waiting for &#39;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&#39; profile generation...&quot;</span><span class="p">)</span>
                <span class="k">while</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">pssm_file</span><span class="p">):</span>
                    <span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">())</span> <span class="o">-</span> <span class="nb">int</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">getmtime</span><span class="p">(</span><span class="n">temp_file</span><span class="p">))</span> <span class="o">&gt;</span> <span class="mi">5400</span><span class="p">:</span>  <span class="c1"># &gt; 1 hr 30 minutes have passed</span>
                        <span class="c1"># os.remove(temp_file)</span>
                        <span class="n">temp_file</span><span class="o">.</span><span class="n">unlink</span><span class="p">(</span><span class="n">missing_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                        <span class="k">raise</span> <span class="n">StructureException</span><span class="p">(</span>
                            <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">add_evolutionary_profile</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s1">: Generation of the profile for </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s1"> &#39;</span>
                            <span class="s1">&#39;took longer than the time limit</span><span class="se">\n</span><span class="s1">Killed&#39;</span><span class="p">)</span>
                    <span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="mi">20</span><span class="p">)</span>

    <span class="c1"># Set self.evolutionary_profile</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">evolutionary_profile</span> <span class="o">=</span> <span class="n">parse_hhblits_pssm</span><span class="p">(</span><span class="n">pssm_file</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="structure.sequence.GeneEntity.create_null_profile" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">create_null_profile</span>


</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">create_null_profile</span><span class="p">(</span><span class="n">nan</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">zero_index</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n"><span title="symdesign.sequence.ProfileDict">ProfileDict</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Make a blank profile</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>nan</code></b>
            (<code>bool</code>, default:
                <code>False</code>
)
        –
        <div class="doc-md-description">
          <p>Whether to fill the null profile with np.nan</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>zero_index</code></b>
            (<code>bool</code>, default:
                <code>False</code>
)
        –
        <div class="doc-md-description">
          <p>bool = False - If True, return the dictionary with zero indexing</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
            <code><span title="symdesign.sequence.ProfileDict">ProfileDict</span></code>
        –
        <div class="doc-md-description">
          <p>Dictionary containing profile information with keys as the index (zero or one-indexed), values as PSSM</p>
      </div>
      </li>
      <li class="field-body">
<b><code>Ex</code></b> (            <code><span title="symdesign.sequence.ProfileDict">ProfileDict</span></code>
)        –
        <div class="doc-md-description">
          <p>{1: {'A': 0, 'R': 0, ..., 'lod': {'A': -5, 'R': -5, ...}, 'type': 'W', 'info': 3.20, 'weight': 0.73},
 2: {}, ...}</p>
      </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">create_null_profile</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">nan</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">zero_index</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">ProfileDict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Make a blank profile</span>

<span class="sd">    Args:</span>
<span class="sd">        nan: Whether to fill the null profile with np.nan</span>
<span class="sd">        zero_index: bool = False - If True, return the dictionary with zero indexing</span>

<span class="sd">    Returns:</span>
<span class="sd">        Dictionary containing profile information with keys as the index (zero or one-indexed), values as PSSM</span>
<span class="sd">        Ex: {1: {&#39;A&#39;: 0, &#39;R&#39;: 0, ..., &#39;lod&#39;: {&#39;A&#39;: -5, &#39;R&#39;: -5, ...}, &#39;type&#39;: &#39;W&#39;, &#39;info&#39;: 3.20, &#39;weight&#39;: 0.73},</span>
<span class="sd">             2: {}, ...}</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">offset</span> <span class="o">=</span> <span class="mi">0</span> <span class="k">if</span> <span class="n">zero_index</span> <span class="k">else</span> <span class="n">ZERO_OFFSET</span>

    <span class="k">if</span> <span class="n">nan</span><span class="p">:</span>
        <span class="n">_profile_entry</span> <span class="o">=</span> <span class="n">nan_profile_entry</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">_profile_entry</span> <span class="o">=</span> <span class="n">blank_profile_entry</span>

    <span class="n">profile</span> <span class="o">=</span> <span class="p">{</span><span class="n">residue</span><span class="p">:</span> <span class="n">_profile_entry</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
               <span class="k">for</span> <span class="n">residue</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">offset</span><span class="p">,</span> <span class="n">offset</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">number_of_residues</span><span class="p">)}</span>

    <span class="k">for</span> <span class="n">residue_data</span><span class="p">,</span> <span class="n">residue_type</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">profile</span><span class="o">.</span><span class="n">values</span><span class="p">(),</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">):</span>
        <span class="n">residue_data</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">residue_type</span>

    <span class="k">return</span> <span class="n">profile</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="structure.sequence.GeneEntity.create_null_entries" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">create_null_entries</span>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">create_null_entries</span><span class="p">(</span><span class="n">entry_numbers</span><span class="p">:</span> <span class="n"><span title="collections.abc.Iterable">Iterable</span></span><span class="p">[</span><span class="n">int</span><span class="p">],</span> <span class="n">nan</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n"><span title="symdesign.sequence.ProfileDict">ProfileDict</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Make a blank profile</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>entry_numbers</code></b>
            (<code><span title="collections.abc.Iterable">Iterable</span>[int]</code>)
        –
        <div class="doc-md-description">
          <p>The numbers to generate null entries for</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>nan</code></b>
            (<code>bool</code>, default:
                <code>False</code>
)
        –
        <div class="doc-md-description">
          <p>Whether to fill the null profile with np.nan</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
            <code><span title="symdesign.sequence.ProfileDict">ProfileDict</span></code>
        –
        <div class="doc-md-description">
          <p>Dictionary containing profile information with the specified entries as the index, values as PSSM</p>
      </div>
      </li>
      <li class="field-body">
<b><code>Ex</code></b> (            <code><span title="symdesign.sequence.ProfileDict">ProfileDict</span></code>
)        –
        <div class="doc-md-description">
          <p>{1: {'A': 0, 'R': 0, ..., 'lod': {'A': -5, 'R': -5, ...}, 'info': 3.20, 'weight': 0.73},
 2: {}, ...}</p>
      </div>
      </li>
      <li class="field-body">
            <code><span title="symdesign.sequence.ProfileDict">ProfileDict</span></code>
        –
        <div class="doc-md-description">
          <p>Importantly, there is no 'type' key. This must be added</p>
      </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@staticmethod</span>
<span class="k">def</span> <span class="nf">create_null_entries</span><span class="p">(</span><span class="n">entry_numbers</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">int</span><span class="p">],</span> <span class="n">nan</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">ProfileDict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Make a blank profile</span>

<span class="sd">    Args:</span>
<span class="sd">        entry_numbers: The numbers to generate null entries for</span>
<span class="sd">        nan: Whether to fill the null profile with np.nan</span>

<span class="sd">    Returns:</span>
<span class="sd">        Dictionary containing profile information with the specified entries as the index, values as PSSM</span>
<span class="sd">        Ex: {1: {&#39;A&#39;: 0, &#39;R&#39;: 0, ..., &#39;lod&#39;: {&#39;A&#39;: -5, &#39;R&#39;: -5, ...}, &#39;info&#39;: 3.20, &#39;weight&#39;: 0.73},</span>
<span class="sd">             2: {}, ...}</span>
<span class="sd">        Importantly, there is no &#39;type&#39; key. This must be added</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># offset = 0 if zero_index else ZERO_OFFSET</span>

    <span class="k">if</span> <span class="n">nan</span><span class="p">:</span>
        <span class="n">_profile_entry</span> <span class="o">=</span> <span class="n">nan_profile_entry</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">_profile_entry</span> <span class="o">=</span> <span class="n">blank_profile_entry</span>

    <span class="k">return</span> <span class="p">{</span><span class="n">entry</span><span class="p">:</span> <span class="n">_profile_entry</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span> <span class="k">for</span> <span class="n">entry</span> <span class="ow">in</span> <span class="n">entry_numbers</span><span class="p">}</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="structure.sequence.GeneEntity.hhblits" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">hhblits</span>


</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">hhblits</span><span class="p">(</span><span class="n">out_dir</span><span class="p">:</span> <span class="n"><span title="typing.AnyStr">AnyStr</span></span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">(),</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">list</span><span class="p">[</span><span class="n">str</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Generate a position specific scoring matrix from hhblits using Hidden Markov Models</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>out_dir</code></b>
            (<code><span title="typing.AnyStr">AnyStr</span></code>, default:
                <code><span title="os.getcwd">getcwd</span>()</code>
)
        –
        <div class="doc-md-description">
          <p>Disk location where generated file should be written</p>
        </div>
      </li>
  </ul>



  <p>Other Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>sequence_file</code></b>
        –
        <div class="doc-md-description">
          <p>AnyStr = None - The file containing the sequence to use</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>threads</code></b>
        –
        <div class="doc-md-description">
          <p>Number of cpu's to use for the process</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>return_command</code></b>
        –
        <div class="doc-md-description">
          <p>Whether to simply return the hhblits command</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
            <code>list[str] | None</code>
        –
        <div class="doc-md-description">
          <p>The command if return_command is True, otherwise None</p>
      </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">887</span>
<span class="normal">888</span>
<span class="normal">889</span>
<span class="normal">890</span>
<span class="normal">891</span>
<span class="normal">892</span>
<span class="normal">893</span>
<span class="normal">894</span>
<span class="normal">895</span>
<span class="normal">896</span>
<span class="normal">897</span>
<span class="normal">898</span>
<span class="normal">899</span>
<span class="normal">900</span>
<span class="normal">901</span>
<span class="normal">902</span>
<span class="normal">903</span>
<span class="normal">904</span>
<span class="normal">905</span>
<span class="normal">906</span>
<span class="normal">907</span>
<span class="normal">908</span>
<span class="normal">909</span>
<span class="normal">910</span>
<span class="normal">911</span>
<span class="normal">912</span>
<span class="normal">913</span>
<span class="normal">914</span>
<span class="normal">915</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">hhblits</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">out_dir</span><span class="p">:</span> <span class="n">AnyStr</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">(),</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Generate a position specific scoring matrix from hhblits using Hidden Markov Models</span>

<span class="sd">    Args:</span>
<span class="sd">        out_dir: Disk location where generated file should be written</span>

<span class="sd">    Keyword Args:</span>
<span class="sd">        sequence_file: AnyStr = None - The file containing the sequence to use</span>
<span class="sd">        threads: Number of cpu&#39;s to use for the process</span>
<span class="sd">        return_command: Whether to simply return the hhblits command</span>

<span class="sd">    Returns:</span>
<span class="sd">        The command if return_command is True, otherwise None</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">hhblits</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">out_dir</span><span class="o">=</span><span class="n">out_dir</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">result</span><span class="p">:</span>  <span class="c1"># return_command is True</span>
        <span class="k">return</span> <span class="n">result</span>
    <span class="c1"># Otherwise, make alignment file(s)</span>
    <span class="n">name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span>
    <span class="c1"># Set file attributes according to logic of hhblits()</span>
    <span class="n">a3m_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">out_dir</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s1">.a3m&#39;</span><span class="p">)</span>
    <span class="n">msa_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">out_dir</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s1">.sto&#39;</span><span class="p">)</span>
    <span class="n">fasta_msa</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">out_dir</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s1">.fasta&#39;</span><span class="p">)</span>
    <span class="c1"># Preferred alignment type</span>
    <span class="n">p</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">([</span><span class="n">putils</span><span class="o">.</span><span class="n">reformat_msa_exe_path</span><span class="p">,</span> <span class="n">a3m_file</span><span class="p">,</span> <span class="n">msa_file</span><span class="p">,</span> <span class="s1">&#39;-num&#39;</span><span class="p">,</span> <span class="s1">&#39;-uc&#39;</span><span class="p">])</span>
    <span class="n">p</span><span class="o">.</span><span class="n">communicate</span><span class="p">()</span>
    <span class="n">p</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">([</span><span class="n">putils</span><span class="o">.</span><span class="n">reformat_msa_exe_path</span><span class="p">,</span> <span class="n">a3m_file</span><span class="p">,</span> <span class="n">fasta_msa</span><span class="p">,</span> <span class="s1">&#39;-M&#39;</span><span class="p">,</span> <span class="s1">&#39;first&#39;</span><span class="p">,</span> <span class="s1">&#39;-r&#39;</span><span class="p">])</span>
    <span class="n">p</span><span class="o">.</span><span class="n">communicate</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="structure.sequence.GeneEntity.add_msa_from_file" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">add_msa_from_file</span>


</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">add_msa_from_file</span><span class="p">(</span><span class="n">msa_file</span><span class="p">:</span> <span class="n"><span title="typing.AnyStr">AnyStr</span></span><span class="p">,</span> <span class="n">file_format</span><span class="p">:</span> <span class="n"><span title="structure.sequence.msa_supported_types_literal">msa_supported_types_literal</span></span> <span class="o">=</span> <span class="s1">&#39;stockholm&#39;</span><span class="p">)</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Add a multiple sequence alignment to the profile. Handles correct sizing of the MSA</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>msa_file</code></b>
            (<code><span title="typing.AnyStr">AnyStr</span></code>)
        –
        <div class="doc-md-description">
          <p>The multiple sequence alignment file to add to the Entity</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>file_format</code></b>
            (<code><span title="structure.sequence.msa_supported_types_literal">msa_supported_types_literal</span></code>, default:
                <code>&#39;stockholm&#39;</code>
)
        –
        <div class="doc-md-description">
          <p>The file type to read the multiple sequence alignment</p>
        </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">917</span>
<span class="normal">918</span>
<span class="normal">919</span>
<span class="normal">920</span>
<span class="normal">921</span>
<span class="normal">922</span>
<span class="normal">923</span>
<span class="normal">924</span>
<span class="normal">925</span>
<span class="normal">926</span>
<span class="normal">927</span>
<span class="normal">928</span>
<span class="normal">929</span>
<span class="normal">930</span>
<span class="normal">931</span>
<span class="normal">932</span>
<span class="normal">933</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_msa_from_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">msa_file</span><span class="p">:</span> <span class="n">AnyStr</span><span class="p">,</span> <span class="n">file_format</span><span class="p">:</span> <span class="n">msa_supported_types_literal</span> <span class="o">=</span> <span class="s1">&#39;stockholm&#39;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Add a multiple sequence alignment to the profile. Handles correct sizing of the MSA</span>

<span class="sd">    Args:</span>
<span class="sd">        msa_file: The multiple sequence alignment file to add to the Entity</span>
<span class="sd">        file_format: The file type to read the multiple sequence alignment</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">file_format</span> <span class="o">==</span> <span class="s1">&#39;stockholm&#39;</span><span class="p">:</span>
        <span class="n">constructor</span> <span class="o">=</span> <span class="n">MultipleSequenceAlignment</span><span class="o">.</span><span class="n">from_stockholm</span>
    <span class="k">elif</span> <span class="n">file_format</span> <span class="o">==</span> <span class="s1">&#39;fasta&#39;</span><span class="p">:</span>
        <span class="n">constructor</span> <span class="o">=</span> <span class="n">MultipleSequenceAlignment</span><span class="o">.</span><span class="n">from_fasta</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;The file format &#39;</span><span class="si">{</span><span class="n">file_format</span><span class="si">}</span><span class="s2">&#39; isn&#39;t an available format. Available formats include &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">msa_supported_types</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">msa</span> <span class="o">=</span> <span class="n">constructor</span><span class="p">(</span><span class="n">msa_file</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="structure.sequence.GeneEntity.collapse_profile" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">collapse_profile</span>


</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">collapse_profile</span><span class="p">(</span><span class="n">msa_file</span><span class="p">:</span> <span class="n"><span title="typing.AnyStr">AnyStr</span></span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n"><span title="numpy.ndarray">ndarray</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Make a profile out of the hydrophobic collapse index (HCI) for each sequence in a multiple sequence alignment</p>
<p>Takes ~5-10 seconds depending on the size of the msa</p>
<p>Calculate HCI for each sequence in the MSA (which are different lengths). This is the Hydro Collapse array. For
each sequence, make a Gap mask, with full shape (length, number_of_residues) to account for gaps in
each sequence. Apply the mask using a map between the Gap mask and the Hydro Collapse array. Finally, drop the
columns from the array that are gaps in the reference sequence.</p>
<p>iter array   -   Gap mask      -       Hydro Collapse array     -     Aligned HCI     - -     Final HCI</p>
<hr />
<p>iter - - - - - - 0 is gap    - -     compute for each     -     account for gaps   -  (drop idx 2)</p>
<p>it 1 2 3 4  - - 0 | 1 | 2 - - - - - - - 0 | 1 | 2 - - - - - - 0 | 1 | 2 - - - - - - - 0 | 1 | 3 | ... N</p>
<p>0 0 1 2 2  - - 1 | 1 | 0 - - -   - - - 0.5 0.2 0.5 - -  = - - 0.5 0.2 0.0 -  -&gt;   - - 0.5 0.2 0.4 ... 0.3</p>
<p>1 0 0 1 2  - - 0 | 1 | 1 - - -   - - - 0.4 0.7 0.4 - -  = - - 0.0 0.4 0.7 -  -&gt;   - - 0.0 0.4 0.4 ... 0.1</p>
<p>2 0 0 1 2  - - 0 | 1 | 1 - - -   - - - 0.3 0.6 0.3 - -  = - - 0.0 0.3 0.6 -  -&gt;   - - 0.0 0.3 0.4 ... 0.0</p>
<p>Where index 0 is the MSA query sequence</p>
<p>After iteration cumulative summation, the iterator is multiplied by the gap mask. Next the Hydro Collapse array
value is accessed by the gaped iterator. This places the Hydro Collapse array or np.nan (if there is a 0 index,
i.e. a gap). After calculation, the element at index 2 is dropped from the array when the aligned sequence gaps
are removed. Finally, only the indices of the query sequence are left in the profile, essentially giving the HCI
for each sequence in the native context, adjusted to the specific context of the protein sequence at hand</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>msa_file</code></b>
            (<code><span title="typing.AnyStr">AnyStr</span></code>, default:
                <code>None</code>
)
        –
        <div class="doc-md-description">
          <p>The multiple sequence alignment file to use for collapse. Will use .msa attribute if not provided</p>
        </div>
      </li>
  </ul>



  <p>Other Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>file_format</code></b>
        –
        <div class="doc-md-description">
          <p>msa_supported_types_literal = 'stockholm' - The file type to read the multiple sequence
alignment</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>hydrophobicity</code></b>
        –
        <div class="doc-md-description">
          <p>int = 'standard' – The hydrophobicity scale to consider. Either 'standard' (FILV),
'expanded' (FMILYVW), or provide one with 'custom' keyword argument</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>custom</code></b>
        –
        <div class="doc-md-description">
          <p>mapping[str, float | int] = None – A user defined mapping of amino acid type, hydrophobicity value
pairs</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>alphabet_type</code></b>
        –
        <div class="doc-md-description">
          <p>alphabet_types = None – The amino acid alphabet if the sequence consists of integer
characters</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>lower_window</code></b>
        –
        <div class="doc-md-description">
          <p>int = 3 – The smallest window used to measure</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>upper_window</code></b>
        –
        <div class="doc-md-description">
          <p>int = 9 – The largest window used to measure</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
            <code><span title="numpy.ndarray">ndarray</span></code>
        –
        <div class="doc-md-description">
          <p>Array with shape (length, number_of_residues) containing the hydrophobic collapse values for
per-residue, per-sequence in the profile. The "query" sequence from the MultipleSequenceAlignment.query
is located at index 0 on axis=0</p>
      </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 935</span>
<span class="normal"> 936</span>
<span class="normal"> 937</span>
<span class="normal"> 938</span>
<span class="normal"> 939</span>
<span class="normal"> 940</span>
<span class="normal"> 941</span>
<span class="normal"> 942</span>
<span class="normal"> 943</span>
<span class="normal"> 944</span>
<span class="normal"> 945</span>
<span class="normal"> 946</span>
<span class="normal"> 947</span>
<span class="normal"> 948</span>
<span class="normal"> 949</span>
<span class="normal"> 950</span>
<span class="normal"> 951</span>
<span class="normal"> 952</span>
<span class="normal"> 953</span>
<span class="normal"> 954</span>
<span class="normal"> 955</span>
<span class="normal"> 956</span>
<span class="normal"> 957</span>
<span class="normal"> 958</span>
<span class="normal"> 959</span>
<span class="normal"> 960</span>
<span class="normal"> 961</span>
<span class="normal"> 962</span>
<span class="normal"> 963</span>
<span class="normal"> 964</span>
<span class="normal"> 965</span>
<span class="normal"> 966</span>
<span class="normal"> 967</span>
<span class="normal"> 968</span>
<span class="normal"> 969</span>
<span class="normal"> 970</span>
<span class="normal"> 971</span>
<span class="normal"> 972</span>
<span class="normal"> 973</span>
<span class="normal"> 974</span>
<span class="normal"> 975</span>
<span class="normal"> 976</span>
<span class="normal"> 977</span>
<span class="normal"> 978</span>
<span class="normal"> 979</span>
<span class="normal"> 980</span>
<span class="normal"> 981</span>
<span class="normal"> 982</span>
<span class="normal"> 983</span>
<span class="normal"> 984</span>
<span class="normal"> 985</span>
<span class="normal"> 986</span>
<span class="normal"> 987</span>
<span class="normal"> 988</span>
<span class="normal"> 989</span>
<span class="normal"> 990</span>
<span class="normal"> 991</span>
<span class="normal"> 992</span>
<span class="normal"> 993</span>
<span class="normal"> 994</span>
<span class="normal"> 995</span>
<span class="normal"> 996</span>
<span class="normal"> 997</span>
<span class="normal"> 998</span>
<span class="normal"> 999</span>
<span class="normal">1000</span>
<span class="normal">1001</span>
<span class="normal">1002</span>
<span class="normal">1003</span>
<span class="normal">1004</span>
<span class="normal">1005</span>
<span class="normal">1006</span>
<span class="normal">1007</span>
<span class="normal">1008</span>
<span class="normal">1009</span>
<span class="normal">1010</span>
<span class="normal">1011</span>
<span class="normal">1012</span>
<span class="normal">1013</span>
<span class="normal">1014</span>
<span class="normal">1015</span>
<span class="normal">1016</span>
<span class="normal">1017</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">collapse_profile</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">msa_file</span><span class="p">:</span> <span class="n">AnyStr</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Make a profile out of the hydrophobic collapse index (HCI) for each sequence in a multiple sequence alignment</span>

<span class="sd">    Takes ~5-10 seconds depending on the size of the msa</span>

<span class="sd">    Calculate HCI for each sequence in the MSA (which are different lengths). This is the Hydro Collapse array. For</span>
<span class="sd">    each sequence, make a Gap mask, with full shape (length, number_of_residues) to account for gaps in</span>
<span class="sd">    each sequence. Apply the mask using a map between the Gap mask and the Hydro Collapse array. Finally, drop the</span>
<span class="sd">    columns from the array that are gaps in the reference sequence.</span>

<span class="sd">    iter array   -   Gap mask      -       Hydro Collapse array     -     Aligned HCI     - -     Final HCI</span>

<span class="sd">    ------------</span>

<span class="sd">    iter - - - - - - 0 is gap    - -     compute for each     -     account for gaps   -  (drop idx 2)</span>

<span class="sd">    it 1 2 3 4  - - 0 | 1 | 2 - - - - - - - 0 | 1 | 2 - - - - - - 0 | 1 | 2 - - - - - - - 0 | 1 | 3 | ... N</span>

<span class="sd">    0 0 1 2 2  - - 1 | 1 | 0 - - -   - - - 0.5 0.2 0.5 - -  = - - 0.5 0.2 0.0 -  -&gt;   - - 0.5 0.2 0.4 ... 0.3</span>

<span class="sd">    1 0 0 1 2  - - 0 | 1 | 1 - - -   - - - 0.4 0.7 0.4 - -  = - - 0.0 0.4 0.7 -  -&gt;   - - 0.0 0.4 0.4 ... 0.1</span>

<span class="sd">    2 0 0 1 2  - - 0 | 1 | 1 - - -   - - - 0.3 0.6 0.3 - -  = - - 0.0 0.3 0.6 -  -&gt;   - - 0.0 0.3 0.4 ... 0.0</span>

<span class="sd">    Where index 0 is the MSA query sequence</span>

<span class="sd">    After iteration cumulative summation, the iterator is multiplied by the gap mask. Next the Hydro Collapse array</span>
<span class="sd">    value is accessed by the gaped iterator. This places the Hydro Collapse array or np.nan (if there is a 0 index,</span>
<span class="sd">    i.e. a gap). After calculation, the element at index 2 is dropped from the array when the aligned sequence gaps</span>
<span class="sd">    are removed. Finally, only the indices of the query sequence are left in the profile, essentially giving the HCI</span>
<span class="sd">    for each sequence in the native context, adjusted to the specific context of the protein sequence at hand</span>

<span class="sd">    Args:</span>
<span class="sd">        msa_file: The multiple sequence alignment file to use for collapse. Will use .msa attribute if not provided</span>

<span class="sd">    Keyword Args:</span>
<span class="sd">        file_format: msa_supported_types_literal = &#39;stockholm&#39; - The file type to read the multiple sequence</span>
<span class="sd">            alignment</span>
<span class="sd">        hydrophobicity: int = &#39;standard&#39; – The hydrophobicity scale to consider. Either &#39;standard&#39; (FILV),</span>
<span class="sd">            &#39;expanded&#39; (FMILYVW), or provide one with &#39;custom&#39; keyword argument</span>
<span class="sd">        custom: mapping[str, float | int] = None – A user defined mapping of amino acid type, hydrophobicity value</span>
<span class="sd">            pairs</span>
<span class="sd">        alphabet_type: alphabet_types = None – The amino acid alphabet if the sequence consists of integer</span>
<span class="sd">            characters</span>
<span class="sd">        lower_window: int = 3 – The smallest window used to measure</span>
<span class="sd">        upper_window: int = 9 – The largest window used to measure</span>

<span class="sd">    Returns:</span>
<span class="sd">        Array with shape (length, number_of_residues) containing the hydrophobic collapse values for</span>
<span class="sd">            per-residue, per-sequence in the profile. The &quot;query&quot; sequence from the MultipleSequenceAlignment.query</span>
<span class="sd">            is located at index 0 on axis=0</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_collapse_profile</span>
    <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
        <span class="n">msa</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">msa</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">msa</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">add_msa_from_file</span><span class="p">(</span><span class="n">msa_file</span><span class="p">)</span>
            <span class="n">msa</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">msa</span>

        <span class="c1"># Make the output array. Use one additional length to add np.nan value at the 0 index for gaps</span>
        <span class="n">evolutionary_collapse_np</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">msa</span><span class="o">.</span><span class="n">length</span><span class="p">,</span> <span class="n">msa</span><span class="o">.</span><span class="n">number_of_positions</span> <span class="o">+</span> <span class="mi">1</span><span class="p">))</span>
        <span class="n">evolutionary_collapse_np</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>  <span class="c1"># np.nan for all missing indices</span>
        <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">msa</span><span class="o">.</span><span class="n">sequences</span><span class="p">):</span>
            <span class="n">non_gaped_sequence</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span>
            <span class="n">evolutionary_collapse_np</span><span class="p">[</span><span class="n">idx</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span><span class="nb">len</span><span class="p">(</span><span class="n">non_gaped_sequence</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> \
                <span class="n">metrics</span><span class="o">.</span><span class="n">hydrophobic_collapse_index</span><span class="p">(</span><span class="n">non_gaped_sequence</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="c1"># Todo this should be possible now metrics.hydrophobic_collapse_index(self.msa.array)</span>

        <span class="n">msa_sequence_indices</span> <span class="o">=</span> <span class="n">msa</span><span class="o">.</span><span class="n">sequence_indices</span>
        <span class="n">iterator_np</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">msa_sequence_indices</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="n">msa_sequence_indices</span>
        <span class="n">aligned_hci_np</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">take_along_axis</span><span class="p">(</span><span class="n">evolutionary_collapse_np</span><span class="p">,</span> <span class="n">iterator_np</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="c1"># Select only the query sequence indices</span>
        <span class="c1"># sequence_hci_np = aligned_hci_np[:, self.msa.query_indices]</span>
        <span class="c1"># print(&#39;aligned_hci_np&#39;, aligned_hci_np.shape, aligned_hci_np)</span>
        <span class="c1"># print(&#39;self.msa.query_indices&#39;, self.msa.query_indices.shape, self.msa.query_indices)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_collapse_profile</span> <span class="o">=</span> <span class="n">aligned_hci_np</span><span class="p">[:,</span> <span class="n">msa</span><span class="o">.</span><span class="n">query_indices</span><span class="p">]</span>
        <span class="c1"># self._collapse_profile = pd.DataFrame(aligned_hci_np[:, self.msa.query_indices],</span>
        <span class="c1">#                                       columns=list(range(1, self.msa.query_length + 1)))  # One-indexed</span>
        <span class="c1"># summary = pd.concat([sequence_hci_df, pd.concat([sequence_hci_df.mean(), sequence_hci_df.std()], axis=1,</span>
        <span class="c1">#                                                 keys=[&#39;mean&#39;, &#39;std&#39;]).T])</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_collapse_profile</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="structure.sequence.GeneEntity.direct_coupling_analysis" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">direct_coupling_analysis</span>


</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">direct_coupling_analysis</span><span class="p">(</span><span class="n">msa_file</span><span class="p">:</span> <span class="n"><span title="typing.AnyStr">AnyStr</span></span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n"><span title="numpy.ndarray">ndarray</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Using boltzmann machine direct coupling analysis (bmDCA), score each sequence in an alignment based on the
 statistical energy compared to the learned DCA model</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>msa_file</code></b>
            (<code><span title="typing.AnyStr">AnyStr</span></code>, default:
                <code>None</code>
)
        –
        <div class="doc-md-description">
          <p>The multiple sequence alignment file to use for collapse. Will use .msa attribute if not provided</p>
        </div>
      </li>
  </ul>



  <p>Other Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>file_format</code></b>
        –
        <div class="doc-md-description">
          <p>msa_supported_types_literal = 'stockholm' - The file type to read the multiple sequence
alignment</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
            <code><span title="numpy.ndarray">ndarray</span></code>
        –
        <div class="doc-md-description">
          <p>Array with shape (length, number_of_residues) where the values are the energy for each residue/sequence
based on direct coupling analysis parameters</p>
      </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1019</span>
<span class="normal">1020</span>
<span class="normal">1021</span>
<span class="normal">1022</span>
<span class="normal">1023</span>
<span class="normal">1024</span>
<span class="normal">1025</span>
<span class="normal">1026</span>
<span class="normal">1027</span>
<span class="normal">1028</span>
<span class="normal">1029</span>
<span class="normal">1030</span>
<span class="normal">1031</span>
<span class="normal">1032</span>
<span class="normal">1033</span>
<span class="normal">1034</span>
<span class="normal">1035</span>
<span class="normal">1036</span>
<span class="normal">1037</span>
<span class="normal">1038</span>
<span class="normal">1039</span>
<span class="normal">1040</span>
<span class="normal">1041</span>
<span class="normal">1042</span>
<span class="normal">1043</span>
<span class="normal">1044</span>
<span class="normal">1045</span>
<span class="normal">1046</span>
<span class="normal">1047</span>
<span class="normal">1048</span>
<span class="normal">1049</span>
<span class="normal">1050</span>
<span class="normal">1051</span>
<span class="normal">1052</span>
<span class="normal">1053</span>
<span class="normal">1054</span>
<span class="normal">1055</span>
<span class="normal">1056</span>
<span class="normal">1057</span>
<span class="normal">1058</span>
<span class="normal">1059</span>
<span class="normal">1060</span>
<span class="normal">1061</span>
<span class="normal">1062</span>
<span class="normal">1063</span>
<span class="normal">1064</span>
<span class="normal">1065</span>
<span class="normal">1066</span>
<span class="normal">1067</span>
<span class="normal">1068</span>
<span class="normal">1069</span>
<span class="normal">1070</span>
<span class="normal">1071</span>
<span class="normal">1072</span>
<span class="normal">1073</span>
<span class="normal">1074</span>
<span class="normal">1075</span>
<span class="normal">1076</span>
<span class="normal">1077</span>
<span class="normal">1078</span>
<span class="normal">1079</span>
<span class="normal">1080</span>
<span class="normal">1081</span>
<span class="normal">1082</span>
<span class="normal">1083</span>
<span class="normal">1084</span>
<span class="normal">1085</span>
<span class="normal">1086</span>
<span class="normal">1087</span>
<span class="normal">1088</span>
<span class="normal">1089</span>
<span class="normal">1090</span>
<span class="normal">1091</span>
<span class="normal">1092</span>
<span class="normal">1093</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">direct_coupling_analysis</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">msa_file</span><span class="p">:</span> <span class="n">AnyStr</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Using boltzmann machine direct coupling analysis (bmDCA), score each sequence in an alignment based on the</span>
<span class="sd">     statistical energy compared to the learned DCA model</span>

<span class="sd">    Args:</span>
<span class="sd">        msa_file: The multiple sequence alignment file to use for collapse. Will use .msa attribute if not provided</span>

<span class="sd">    Keyword Args:</span>
<span class="sd">        file_format: msa_supported_types_literal = &#39;stockholm&#39; - The file type to read the multiple sequence</span>
<span class="sd">            alignment</span>

<span class="sd">    Returns:</span>
<span class="sd">        Array with shape (length, number_of_residues) where the values are the energy for each residue/sequence</span>
<span class="sd">            based on direct coupling analysis parameters</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Check if required attributes are present</span>
    <span class="n">_raise</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="n">missing_attrs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">msa</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">msa</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">msa</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">msa_file</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">add_msa_from_file</span><span class="p">(</span><span class="n">msa_file</span><span class="p">)</span>
            <span class="n">msa</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">msa</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">missing_attrs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;.msa&#39;</span><span class="p">)</span>
    <span class="n">h_fields</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">h_fields</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">h_fields</span><span class="p">:</span>
        <span class="n">missing_attrs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;.h_fields&#39;</span><span class="p">)</span>
    <span class="n">j_couplings</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">j_couplings</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">j_couplings</span><span class="p">:</span>
        <span class="n">missing_attrs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;.j_couplings&#39;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">missing_attrs</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">AttributeError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;The required attribute(s) </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">missing_attrs</span><span class="p">)</span><span class="si">}</span><span class="s2"> aren&#39;t available. Add to the &quot;</span>
            <span class="sa">f</span><span class="s1">&#39;Entity before </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">direct_coupling_analysis</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="n">analysis_length</span> <span class="o">=</span> <span class="n">msa</span><span class="o">.</span><span class="n">query_length</span>
    <span class="n">idx_range</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">analysis_length</span><span class="p">)</span>
    <span class="c1"># h_fields = bmdca.load_fields(os.path.join(data_dir, &#39;%s_bmDCA&#39; % self.name, &#39;parameters_h_final.bin&#39;))</span>
    <span class="c1"># h_fields = h_fields.T  # this isn&#39;t required when coming in Fortran order, i.e. (21, analysis_length)</span>
    <span class="c1"># sum the h_fields values for each sequence position in every sequence</span>
    <span class="n">h_values</span> <span class="o">=</span> <span class="n">h_fields</span><span class="p">[</span><span class="n">msa</span><span class="o">.</span><span class="n">numerical_alignment</span><span class="p">,</span> <span class="n">idx_range</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="p">:]]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">h_sum</span> <span class="o">=</span> <span class="n">h_values</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># coming in as a 4 dimension (analysis_length, analysis_length, alphabet_number, alphabet_number) ndarray</span>
    <span class="c1"># j_couplings = bmdca.load_couplings(os.path.join(data_dir, &#39;%s_bmDCA&#39; % self.name, &#39;parameters_J_final.bin&#39;))</span>
    <span class="n">i_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">repeat</span><span class="p">(</span><span class="n">idx_range</span><span class="p">,</span> <span class="n">analysis_length</span><span class="p">)</span>
    <span class="n">j_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">tile</span><span class="p">(</span><span class="n">idx_range</span><span class="p">,</span> <span class="n">analysis_length</span><span class="p">)</span>
    <span class="n">i_aa</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">repeat</span><span class="p">(</span><span class="n">msa</span><span class="o">.</span><span class="n">numerical_alignment</span><span class="p">,</span> <span class="n">analysis_length</span><span class="p">)</span>
    <span class="n">j_aa</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">tile</span><span class="p">(</span><span class="n">msa</span><span class="o">.</span><span class="n">numerical_alignment</span><span class="p">,</span> <span class="n">msa</span><span class="o">.</span><span class="n">query_length</span><span class="p">)</span>
    <span class="n">j_values</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">msa</span><span class="o">.</span><span class="n">length</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">i_idx</span><span class="p">)))</span>
    <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">msa</span><span class="o">.</span><span class="n">length</span><span class="p">):</span>
        <span class="n">j_values</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">=</span> <span class="n">j_couplings</span><span class="p">[</span><span class="n">i_idx</span><span class="p">,</span> <span class="n">j_idx</span><span class="p">,</span> <span class="n">i_aa</span><span class="p">,</span> <span class="n">j_aa</span><span class="p">]</span>
    <span class="c1"># this mask is not necessary when the array comes in as a non-symmetry matrix. All i &gt; j result in 0 values...</span>
    <span class="c1"># mask = np.triu(np.ones((analysis_length, analysis_length)), k=1).flatten()</span>
    <span class="c1"># j_sum = j_values[:, mask].sum(axis=1)</span>
    <span class="c1"># sum the j_values for every design (axis 0) at every residue position (axis 1)</span>
    <span class="n">j_values</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">j_values</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>
    <span class="n">j_sum</span> <span class="o">=</span> <span class="n">j_values</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="c1"># couplings_idx = np.stack((i_idx, j_idx, i_aa, j_aa), axis=1)</span>
    <span class="c1"># this stacks all arrays like so</span>
    <span class="c1">#  [[[ i_idx1, i_idx2, ..., i_idxN],</span>
    <span class="c1">#    [ j_idx1, j_idx2, ..., j_idxN],  &lt;- this is for one sequence</span>
    <span class="c1">#    [ i_aa 1, i_aa 2, ..., i_aa N],</span>
    <span class="c1">#    [ j_aa 1, j_aa 2, ..., j_aa N]],</span>
    <span class="c1">#   [[NEXT SEQUENCE],</span>
    <span class="c1">#    [</span>
    <span class="c1"># this stacks all arrays the transpose, which would match the indexing style on j_couplings much better...</span>
    <span class="c1"># couplings_idx = np.stack((i_idx, j_idx, i_aa, j_aa), axis=2)</span>
    <span class="c1"># j_sum = np.zeros((self.msa.length, len(couplings_idx)))</span>
    <span class="c1"># for idx in range(self.msa.length):</span>
    <span class="c1">#     j_sum[idx] = j_couplings[couplings_idx[idx]]</span>
    <span class="c1"># return -h_sum - j_sum</span>
    <span class="k">return</span> <span class="o">-</span><span class="n">h_values</span> <span class="o">-</span> <span class="n">j_values</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="structure.sequence.GeneEntity.write_sequence_to_fasta" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">write_sequence_to_fasta</span>


</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">write_sequence_to_fasta</span><span class="p">(</span><span class="n">dtype</span><span class="p">:</span> <span class="n"><span title="structure.sequence.sequence_type_literal">sequence_type_literal</span></span> <span class="o">=</span> <span class="s1">&#39;reference_sequence&#39;</span><span class="p">,</span> <span class="n">file_name</span><span class="p">:</span> <span class="n"><span title="typing.AnyStr">AnyStr</span></span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">out_dir</span><span class="p">:</span> <span class="n"><span title="typing.AnyStr">AnyStr</span></span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">())</span> <span class="o">-&gt;</span> <span class="n"><span title="typing.AnyStr">AnyStr</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Write a sequence to a .fasta file with fasta format and return the file location.
'.fasta' is appended if not specified in the name argument</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>dtype</code></b>
            (<code><span title="structure.sequence.sequence_type_literal">sequence_type_literal</span></code>, default:
                <code>&#39;reference_sequence&#39;</code>
)
        –
        <div class="doc-md-description">
          <p>The type of sequence to write. Can be the the keywords 'reference_sequence' or 'sequence'</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>file_name</code></b>
            (<code><span title="typing.AnyStr">AnyStr</span></code>, default:
                <code>None</code>
)
        –
        <div class="doc-md-description">
          <p>The explicit name of the file</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>name</code></b>
            (<code>str</code>, default:
                <code>None</code>
)
        –
        <div class="doc-md-description">
          <p>The name of the sequence record. If not provided, the instance name will be used.
Will be used as the default file_name base name if file_name not provided</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>out_dir</code></b>
            (<code><span title="typing.AnyStr">AnyStr</span></code>, default:
                <code><span title="os.getcwd">getcwd</span>()</code>
)
        –
        <div class="doc-md-description">
          <p>The location on disk to output the file. Only used if file_name not explicitly provided</p>
        </div>
      </li>
  </ul>
      <p>Returns:
    The path to the output file</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1095</span>
<span class="normal">1096</span>
<span class="normal">1097</span>
<span class="normal">1098</span>
<span class="normal">1099</span>
<span class="normal">1100</span>
<span class="normal">1101</span>
<span class="normal">1102</span>
<span class="normal">1103</span>
<span class="normal">1104</span>
<span class="normal">1105</span>
<span class="normal">1106</span>
<span class="normal">1107</span>
<span class="normal">1108</span>
<span class="normal">1109</span>
<span class="normal">1110</span>
<span class="normal">1111</span>
<span class="normal">1112</span>
<span class="normal">1113</span>
<span class="normal">1114</span>
<span class="normal">1115</span>
<span class="normal">1116</span>
<span class="normal">1117</span>
<span class="normal">1118</span>
<span class="normal">1119</span>
<span class="normal">1120</span>
<span class="normal">1121</span>
<span class="normal">1122</span>
<span class="normal">1123</span>
<span class="normal">1124</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">write_sequence_to_fasta</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">dtype</span><span class="p">:</span> <span class="n">sequence_type_literal</span> <span class="o">=</span> <span class="s1">&#39;reference_sequence&#39;</span><span class="p">,</span> <span class="n">file_name</span><span class="p">:</span> <span class="n">AnyStr</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                            <span class="n">name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">out_dir</span><span class="p">:</span> <span class="n">AnyStr</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">())</span> <span class="o">-&gt;</span> <span class="n">AnyStr</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Write a sequence to a .fasta file with fasta format and return the file location.</span>
<span class="sd">    &#39;.fasta&#39; is appended if not specified in the name argument</span>

<span class="sd">    Args:</span>
<span class="sd">        dtype: The type of sequence to write. Can be the the keywords &#39;reference_sequence&#39; or &#39;sequence&#39;</span>
<span class="sd">        file_name: The explicit name of the file</span>
<span class="sd">        name: The name of the sequence record. If not provided, the instance name will be used.</span>
<span class="sd">            Will be used as the default file_name base name if file_name not provided</span>
<span class="sd">        out_dir: The location on disk to output the file. Only used if file_name not explicitly provided</span>
<span class="sd">    Returns:</span>
<span class="sd">        The path to the output file</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">dtype</span> <span class="ow">in</span> <span class="n">sequence_types</span><span class="p">:</span>
        <span class="c1"># Get the attribute from the instance</span>
        <span class="n">sequence</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">dtype</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Couldn&#39;t find a sequence matching the </span><span class="si">{</span><span class="n">dtype</span><span class="si">=}</span><span class="s2">&quot;</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">name</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span>
    <span class="k">if</span> <span class="n">file_name</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">file_name</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">out_dir</span><span class="p">,</span> <span class="n">name</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">file_name</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.fasta&#39;</span><span class="p">):</span>
            <span class="n">file_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">file_name</span><span class="si">}</span><span class="s1">.fasta&#39;</span>

    <span class="k">return</span> <span class="n">write_sequence_to_fasta</span><span class="p">(</span><span class="n">sequence</span><span class="o">=</span><span class="n">sequence</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="n">name</span><span class="p">,</span> <span class="n">file_name</span><span class="o">=</span><span class="n">file_name</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>


</div>



<div class="doc doc-object doc-function">



<h2 id="structure.sequence.sequence_to_one_hot" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">sequence_to_one_hot</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">sequence_to_one_hot</span><span class="p">(</span><span class="n">sequence</span><span class="p">:</span> <span class="n"><span title="collections.abc.Sequence">Sequence</span></span><span class="p">[</span><span class="n">str</span><span class="p">],</span> <span class="n">translation_table</span><span class="p">:</span> <span class="n">dict</span><span class="p">[</span><span class="n">str</span><span class="p">,</span> <span class="n">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">alphabet_order</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n"><span title="numpy.ndarray">ndarray</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Convert a sequence into a numeric array</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>sequence</code></b>
            (<code><span title="collections.abc.Sequence">Sequence</span>[str]</code>)
        –
        <div class="doc-md-description">
          <p>The sequence to encode</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>translation_table</code></b>
            (<code>dict[str, int]</code>, default:
                <code>None</code>
)
        –
        <div class="doc-md-description">
          <p>If a translation table (in bytes) is provided, it will be used. If not, use alphabet_order</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>alphabet_order</code></b>
            (<code>int</code>, default:
                <code>1</code>
)
        –
        <div class="doc-md-description">
          <p>The alphabetical order of the amino acid alphabet. Can be either 1 or 3</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
            <code><span title="numpy.ndarray">ndarray</span></code>
        –
        <div class="doc-md-description">
          <p>The one-hot encoded sequence with shape (sequence length, translation_table length)</p>
      </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">sequence_to_one_hot</span><span class="p">(</span><span class="n">sequence</span><span class="p">:</span> <span class="n">Sequence</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">translation_table</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                        <span class="n">alphabet_order</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Convert a sequence into a numeric array</span>

<span class="sd">    Args:</span>
<span class="sd">        sequence: The sequence to encode</span>
<span class="sd">        translation_table: If a translation table (in bytes) is provided, it will be used. If not, use alphabet_order</span>
<span class="sd">        alphabet_order: The alphabetical order of the amino acid alphabet. Can be either 1 or 3</span>

<span class="sd">    Returns:</span>
<span class="sd">        The one-hot encoded sequence with shape (sequence length, translation_table length)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">numeric_sequence</span> <span class="o">=</span> <span class="n">sequence_to_numeric</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">translation_table</span><span class="p">,</span> <span class="n">alphabet_order</span><span class="o">=</span><span class="n">alphabet_order</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">translation_table</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># Assumes that alphabet_order is used and there aren&#39;t missing letters...</span>
        <span class="n">num_entries</span> <span class="o">=</span> <span class="mi">20</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">num_entries</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">translation_table</span><span class="p">)</span>
    <span class="n">one_hot</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">),</span> <span class="n">num_entries</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">one_hot</span><span class="p">[:,</span> <span class="n">numeric_sequence</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="k">except</span> <span class="ne">IndexError</span><span class="p">:</span>  <span class="c1"># Our assumption above was wrong</span>
        <span class="kn">from</span> <span class="nn">symdesign</span> <span class="kn">import</span> <span class="n">sequence</span> <span class="k">as</span> <span class="n">_seq</span>
        <span class="n">embedding</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">_seq</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;numerical_translation_alph</span><span class="si">{</span><span class="n">alphabet_order</span><span class="si">}</span><span class="s1">_bytes&#39;</span><span class="p">)</span> \
            <span class="k">if</span> <span class="n">translation_table</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">translation_table</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Couldn&#39;t produce a proper one-hot encoding for the provided sequence embedding: </span><span class="si">{</span><span class="n">embedding</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">one_hot</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="structure.sequence.sequence_to_numeric" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">sequence_to_numeric</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">sequence_to_numeric</span><span class="p">(</span><span class="n">sequence</span><span class="p">:</span> <span class="n"><span title="collections.abc.Sequence">Sequence</span></span><span class="p">[</span><span class="n">str</span><span class="p">],</span> <span class="n">translation_table</span><span class="p">:</span> <span class="n">dict</span><span class="p">[</span><span class="n">str</span><span class="p">,</span> <span class="n">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">alphabet_order</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n"><span title="numpy.ndarray">ndarray</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Convert a sequence into a numeric array</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>sequence</code></b>
            (<code><span title="collections.abc.Sequence">Sequence</span>[str]</code>)
        –
        <div class="doc-md-description">
          <p>The sequence to encode</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>translation_table</code></b>
            (<code>dict[str, int]</code>, default:
                <code>None</code>
)
        –
        <div class="doc-md-description">
          <p>If a translation table (in bytes) is provided, it will be used. If not, use alphabet_order</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>alphabet_order</code></b>
            (<code>int</code>, default:
                <code>1</code>
)
        –
        <div class="doc-md-description">
          <p>The alphabetical order of the amino acid alphabet. Can be either 1 or 3</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
            <code><span title="numpy.ndarray">ndarray</span></code>
        –
        <div class="doc-md-description">
          <p>The numerically encoded sequence where each entry along axis=0 is the indexed amino acid. Indices are according
to the 1 letter alphabetical amino acid</p>
      </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">sequence_to_numeric</span><span class="p">(</span><span class="n">sequence</span><span class="p">:</span> <span class="n">Sequence</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">translation_table</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                        <span class="n">alphabet_order</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Convert a sequence into a numeric array</span>

<span class="sd">    Args:</span>
<span class="sd">        sequence: The sequence to encode</span>
<span class="sd">        translation_table: If a translation table (in bytes) is provided, it will be used. If not, use alphabet_order</span>
<span class="sd">        alphabet_order: The alphabetical order of the amino acid alphabet. Can be either 1 or 3</span>

<span class="sd">    Returns:</span>
<span class="sd">        The numerically encoded sequence where each entry along axis=0 is the indexed amino acid. Indices are according</span>
<span class="sd">            to the 1 letter alphabetical amino acid</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">_array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">sequence</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">string_</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">translation_table</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">vectorize</span><span class="p">(</span><span class="n">translation_table</span><span class="o">.</span><span class="fm">__getitem__</span><span class="p">)(</span><span class="n">_array</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">alphabet_order</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">vectorize</span><span class="p">(</span><span class="n">numerical_translation_alph1_bytes</span><span class="o">.</span><span class="fm">__getitem__</span><span class="p">)(</span><span class="n">_array</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">alphabet_order</span> <span class="o">==</span> <span class="mi">3</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s1">&#39;Need to make the &quot;numerical_translation_alph3_bytes&quot; table&#39;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">vectorize</span><span class="p">(</span><span class="n">numerical_translation_alph3_bytes</span><span class="o">.</span><span class="fm">__getitem__</span><span class="p">)(</span><span class="n">_array</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;The &#39;alphabet_order&#39; </span><span class="si">{</span><span class="n">alphabet_order</span><span class="si">}</span><span class="s2"> isn&#39;t valid. Choose from either 1 or 3&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="structure.sequence.sequences_to_numeric" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">sequences_to_numeric</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">sequences_to_numeric</span><span class="p">(</span><span class="n">sequences</span><span class="p">:</span> <span class="n"><span title="collections.abc.Iterable">Iterable</span></span><span class="p">[</span><span class="n"><span title="collections.abc.Sequence">Sequence</span></span><span class="p">[</span><span class="n">str</span><span class="p">]],</span> <span class="n">translation_table</span><span class="p">:</span> <span class="n">dict</span><span class="p">[</span><span class="n">str</span><span class="p">,</span> <span class="n">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">alphabet_order</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n"><span title="numpy.ndarray">ndarray</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Convert sequences into a numeric array</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>sequences</code></b>
            (<code><span title="collections.abc.Iterable">Iterable</span>[<span title="collections.abc.Sequence">Sequence</span>[str]]</code>)
        –
        <div class="doc-md-description">
          <p>The sequences to encode</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>translation_table</code></b>
            (<code>dict[str, int]</code>, default:
                <code>None</code>
)
        –
        <div class="doc-md-description">
          <p>If a translation table (in bytes) is provided, it will be used. If not, use alphabet_order</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>alphabet_order</code></b>
            (<code>int</code>, default:
                <code>1</code>
)
        –
        <div class="doc-md-description">
          <p>The alphabetical order of the amino acid alphabet. Can be either 1 or 3</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
            <code><span title="numpy.ndarray">ndarray</span></code>
        –
        <div class="doc-md-description">
          <p>The numerically encoded sequence where each entry along axis=0 is the indexed amino acid. Indices are according
to the 1 letter alphabetical amino acid</p>
      </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">sequences_to_numeric</span><span class="p">(</span><span class="n">sequences</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="n">Sequence</span><span class="p">[</span><span class="nb">str</span><span class="p">]],</span> <span class="n">translation_table</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                         <span class="n">alphabet_order</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Convert sequences into a numeric array</span>

<span class="sd">    Args:</span>
<span class="sd">        sequences: The sequences to encode</span>
<span class="sd">        translation_table: If a translation table (in bytes) is provided, it will be used. If not, use alphabet_order</span>
<span class="sd">        alphabet_order: The alphabetical order of the amino acid alphabet. Can be either 1 or 3</span>

<span class="sd">    Returns:</span>
<span class="sd">        The numerically encoded sequence where each entry along axis=0 is the indexed amino acid. Indices are according</span>
<span class="sd">            to the 1 letter alphabetical amino acid</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">_array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="nb">list</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span> <span class="k">for</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">sequences</span><span class="p">],</span> <span class="n">np</span><span class="o">.</span><span class="n">string_</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">translation_table</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">vectorize</span><span class="p">(</span><span class="n">translation_table</span><span class="o">.</span><span class="fm">__getitem__</span><span class="p">)(</span><span class="n">_array</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">alphabet_order</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">vectorize</span><span class="p">(</span><span class="n">numerical_translation_alph1_bytes</span><span class="o">.</span><span class="fm">__getitem__</span><span class="p">)(</span><span class="n">_array</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">alphabet_order</span> <span class="o">==</span> <span class="mi">3</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s1">&#39;Need to make the &quot;numerical_translation_alph3_bytes&quot; table&#39;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">vectorize</span><span class="p">(</span><span class="n">numerical_translation_alph3_bytes</span><span class="o">.</span><span class="fm">__getitem__</span><span class="p">)(</span><span class="n">_array</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;The &#39;alphabet_order&#39; </span><span class="si">{</span><span class="n">alphabet_order</span><span class="si">}</span><span class="s2"> isn&#39;t valid. Choose from either 1 or 3&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="structure.sequence.pssm_as_array" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">pssm_as_array</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">pssm_as_array</span><span class="p">(</span><span class="n">pssm</span><span class="p">:</span> <span class="n"><span title="symdesign.sequence.ProfileDict">ProfileDict</span></span><span class="p">,</span> <span class="n">alphabet</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="n">protein_letters_alph1</span><span class="p">,</span> <span class="n">lod</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n"><span title="numpy.ndarray">ndarray</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Convert a position specific profile matrix into a numeric array</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>pssm</code></b>
            (<code><span title="symdesign.sequence.ProfileDict">ProfileDict</span></code>)
        –
        <div class="doc-md-description">
          <p>{1: {'A': 0, 'R': 0, ..., 'lod': {'A': -5, 'R': -5, ...}, 'type': 'W', 'info': 3.20, 'weight': 0.73},
      2: {}, ...}</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>alphabet</code></b>
            (<code>str</code>, default:
                <code><span title="symdesign.sequence.protein_letters_alph1">protein_letters_alph1</span></code>
)
        –
        <div class="doc-md-description">
          <p>The amino acid alphabet to use. Array values will be returned in this order</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>lod</code></b>
            (<code>bool</code>, default:
                <code>False</code>
)
        –
        <div class="doc-md-description">
          <p>Whether to return the array for the log of odds values</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
            <code><span title="numpy.ndarray">ndarray</span></code>
        –
        <div class="doc-md-description">
          <p>The numerically encoded pssm where each entry along axis 0 is the position, and the entries on axis 1 are the
frequency data at every indexed amino acid. Indices are according to the specified amino acid alphabet,
i.e array([[0.1, 0.01, 0.12, ...], ...])</p>
      </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">pssm_as_array</span><span class="p">(</span><span class="n">pssm</span><span class="p">:</span> <span class="n">ProfileDict</span><span class="p">,</span> <span class="n">alphabet</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">protein_letters_alph1</span><span class="p">,</span> <span class="n">lod</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> \
        <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Convert a position specific profile matrix into a numeric array</span>

<span class="sd">    Args:</span>
<span class="sd">        pssm: {1: {&#39;A&#39;: 0, &#39;R&#39;: 0, ..., &#39;lod&#39;: {&#39;A&#39;: -5, &#39;R&#39;: -5, ...}, &#39;type&#39;: &#39;W&#39;, &#39;info&#39;: 3.20, &#39;weight&#39;: 0.73},</span>
<span class="sd">                  2: {}, ...}</span>
<span class="sd">        alphabet: The amino acid alphabet to use. Array values will be returned in this order</span>
<span class="sd">        lod: Whether to return the array for the log of odds values</span>

<span class="sd">    Returns:</span>
<span class="sd">        The numerically encoded pssm where each entry along axis 0 is the position, and the entries on axis 1 are the</span>
<span class="sd">            frequency data at every indexed amino acid. Indices are according to the specified amino acid alphabet,</span>
<span class="sd">            i.e array([[0.1, 0.01, 0.12, ...], ...])</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">lod</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="n">position_info</span><span class="p">[</span><span class="s1">&#39;lod&#39;</span><span class="p">][</span><span class="n">aa</span><span class="p">]</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">alphabet</span><span class="p">]</span>
                         <span class="k">for</span> <span class="n">position_info</span> <span class="ow">in</span> <span class="n">pssm</span><span class="o">.</span><span class="n">values</span><span class="p">()],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="n">position_info</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">alphabet</span><span class="p">]</span>
                         <span class="k">for</span> <span class="n">position_info</span> <span class="ow">in</span> <span class="n">pssm</span><span class="o">.</span><span class="n">values</span><span class="p">()],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="structure.sequence.concatenate_profile" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">concatenate_profile</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">concatenate_profile</span><span class="p">(</span><span class="n">profiles</span><span class="p">:</span> <span class="n"><span title="collections.abc.Iterable">Iterable</span></span><span class="p">[</span><span class="n"><span title="typing.Any">Any</span></span><span class="p">],</span> <span class="n">start_at</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">dict</span><span class="p">[</span><span class="n">int</span><span class="p">,</span> <span class="n"><span title="typing.Any">Any</span></span><span class="p">]</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Combine a list of profiles (parsed PSSMs) by incrementing the entry index for each additional profile</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>profiles</code></b>
            (<code><span title="collections.abc.Iterable">Iterable</span>[<span title="typing.Any">Any</span>]</code>)
        –
        <div class="doc-md-description">
          <p>The profiles to concatenate</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>start_at</code></b>
            (<code>int</code>, default:
                <code>1</code>
)
        –
        <div class="doc-md-description">
          <p>The integer to start the resulting dictionary at</p>
        </div>
      </li>
  </ul>
      <p>Returns
    The concatenated input profiles, make a concatenated PSSM
        {1: {'A': 0.04, 'C': 0.12, ..., 'lod': {'A': -5, 'C': -9, ...}, 'type': 'W', 'info': 0.00,
             'weight': 0.00}, ...}}</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">concatenate_profile</span><span class="p">(</span><span class="n">profiles</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="n">Any</span><span class="p">],</span> <span class="n">start_at</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">Any</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Combine a list of profiles (parsed PSSMs) by incrementing the entry index for each additional profile</span>

<span class="sd">    Args:</span>
<span class="sd">        profiles: The profiles to concatenate</span>
<span class="sd">        start_at: The integer to start the resulting dictionary at</span>

<span class="sd">    Returns</span>
<span class="sd">        The concatenated input profiles, make a concatenated PSSM</span>
<span class="sd">            {1: {&#39;A&#39;: 0.04, &#39;C&#39;: 0.12, ..., &#39;lod&#39;: {&#39;A&#39;: -5, &#39;C&#39;: -9, ...}, &#39;type&#39;: &#39;W&#39;, &#39;info&#39;: 0.00,</span>
<span class="sd">                 &#39;weight&#39;: 0.00}, ...}}</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">_count</span> <span class="o">=</span> <span class="n">count</span><span class="p">(</span><span class="n">start_at</span><span class="p">)</span>
    <span class="k">return</span> <span class="p">{</span><span class="nb">next</span><span class="p">(</span><span class="n">_count</span><span class="p">):</span> <span class="n">position_profile</span> <span class="k">for</span> <span class="n">profile</span> <span class="ow">in</span> <span class="n">profiles</span> <span class="k">for</span> <span class="n">position_profile</span> <span class="ow">in</span> <span class="n">profile</span><span class="o">.</span><span class="n">values</span><span class="p">()}</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="structure.sequence.write_pssm_file" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">write_pssm_file</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">write_pssm_file</span><span class="p">(</span><span class="n">pssm</span><span class="p">:</span> <span class="n"><span title="symdesign.sequence.ProfileDict">ProfileDict</span></span><span class="p">,</span> <span class="n">file_name</span><span class="p">:</span> <span class="n"><span title="typing.AnyStr">AnyStr</span></span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">out_dir</span><span class="p">:</span> <span class="n"><span title="typing.AnyStr">AnyStr</span></span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">())</span> <span class="o">-&gt;</span> <span class="n"><span title="typing.AnyStr">AnyStr</span></span> <span class="o">|</span> <span class="kc">None</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Create a PSI-BLAST format PSSM file from a PSSM dictionary. Assumes residue numbering is correct!</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>pssm</code></b>
            (<code><span title="symdesign.sequence.ProfileDict">ProfileDict</span></code>)
        –
        <div class="doc-md-description">
          <p>A dictionary which has the keys: 'A', 'C', ... (all aa's), 'lod', 'type', 'info', 'weight'</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>file_name</code></b>
            (<code><span title="typing.AnyStr">AnyStr</span></code>, default:
                <code>None</code>
)
        –
        <div class="doc-md-description">
          <p>The explicit name of the file</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>name</code></b>
            (<code>str</code>, default:
                <code>None</code>
)
        –
        <div class="doc-md-description">
          <p>The name of the file. Will be used as the default file_name base name if file_name not provided</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>out_dir</code></b>
            (<code><span title="typing.AnyStr">AnyStr</span></code>, default:
                <code><span title="os.getcwd">getcwd</span>()</code>
)
        –
        <div class="doc-md-description">
          <p>The location on disk to output the file. Only used if file_name not explicitly provided</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
            <code><span title="typing.AnyStr">AnyStr</span> | None</code>
        –
        <div class="doc-md-description">
          <p>Disk location of newly created .pssm file</p>
      </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">write_pssm_file</span><span class="p">(</span><span class="n">pssm</span><span class="p">:</span> <span class="n">ProfileDict</span><span class="p">,</span> <span class="n">file_name</span><span class="p">:</span> <span class="n">AnyStr</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                    <span class="n">out_dir</span><span class="p">:</span> <span class="n">AnyStr</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">())</span> <span class="o">-&gt;</span> <span class="n">AnyStr</span> <span class="o">|</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create a PSI-BLAST format PSSM file from a PSSM dictionary. Assumes residue numbering is correct!</span>

<span class="sd">    Args:</span>
<span class="sd">        pssm: A dictionary which has the keys: &#39;A&#39;, &#39;C&#39;, ... (all aa&#39;s), &#39;lod&#39;, &#39;type&#39;, &#39;info&#39;, &#39;weight&#39;</span>
<span class="sd">        file_name: The explicit name of the file</span>
<span class="sd">        name: The name of the file. Will be used as the default file_name base name if file_name not provided</span>
<span class="sd">        out_dir: The location on disk to output the file. Only used if file_name not explicitly provided</span>

<span class="sd">    Returns:</span>
<span class="sd">        Disk location of newly created .pssm file</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">pssm</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="c1"># Find out if the pssm has values expressed as frequencies (percentages) or as counts and modify accordingly</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">pssm</span><span class="o">.</span><span class="n">values</span><span class="p">())[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;lod&#39;</span><span class="p">][</span><span class="s1">&#39;A&#39;</span><span class="p">],</span> <span class="nb">float</span><span class="p">):</span>
        <span class="n">separation1</span> <span class="o">=</span> <span class="s2">&quot; &quot;</span> <span class="o">*</span> <span class="mi">4</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">separation1</span> <span class="o">=</span> <span class="s2">&quot; &quot;</span> <span class="o">*</span> <span class="mi">3</span>

    <span class="k">if</span> <span class="n">file_name</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">name</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Must provide argument &quot;file_name&quot; or &quot;name&quot; as a str to </span><span class="si">{</span><span class="n">write_sequences</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">file_name</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">out_dir</span><span class="p">,</span> <span class="n">name</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">file_name</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;&#39;</span><span class="p">:</span>  <span class="c1"># No extension</span>
        <span class="n">file_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">file_name</span><span class="si">}</span><span class="s1">.pssm&#39;</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="se">\n\n</span><span class="si">{</span><span class="s2">&quot; &quot;</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="mi">12</span><span class="si">}{</span><span class="n">separation1</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">protein_letters_alph3</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span>
                <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">separation1</span><span class="si">}{</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="mi">3</span><span class="p">)</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">protein_letters_alph3</span><span class="p">)</span><span class="si">}</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">residue_number</span><span class="p">,</span> <span class="n">profile</span> <span class="ow">in</span> <span class="n">pssm</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">profile</span><span class="p">[</span><span class="s1">&#39;lod&#39;</span><span class="p">][</span><span class="s1">&#39;A&#39;</span><span class="p">],</span> <span class="nb">float</span><span class="p">):</span>  <span class="c1"># lod_freq:  # relevant for favor_fragment</span>
                <span class="n">lod_string</span> <span class="o">=</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">profile</span><span class="p">[</span><span class="s2">&quot;lod&quot;</span><span class="p">][</span><span class="n">aa</span><span class="p">]</span><span class="si">:</span><span class="s1">&gt;4.2f</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">protein_letters_alph3</span><span class="p">)</span> \
                    <span class="o">+</span> <span class="s1">&#39; &#39;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">lod_string</span> <span class="o">=</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">profile</span><span class="p">[</span><span class="s2">&quot;lod&quot;</span><span class="p">][</span><span class="n">aa</span><span class="p">]</span><span class="si">:</span><span class="s1">&gt;3d</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">protein_letters_alph3</span><span class="p">)</span> \
                    <span class="o">+</span> <span class="s1">&#39; &#39;</span>

            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">profile</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">],</span> <span class="nb">float</span><span class="p">):</span>  <span class="c1"># counts_freq: # relevant for freq calculations</span>
                <span class="n">counts_string</span> <span class="o">=</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">floor</span><span class="p">(</span><span class="n">profile</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="mi">100</span><span class="p">)</span><span class="si">:</span><span class="s1">&gt;3.0f</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span>
                                         <span class="n">protein_letters_alph3</span><span class="p">)</span> \
                    <span class="o">+</span> <span class="s1">&#39; &#39;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">counts_string</span> <span class="o">=</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">profile</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span><span class="si">:</span><span class="s1">&gt;3d</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">protein_letters_alph3</span><span class="p">)</span> \
                    <span class="o">+</span> <span class="s1">&#39; &#39;</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">residue_number</span><span class="si">:</span><span class="s1">&gt;5d</span><span class="si">}</span><span class="s1"> </span><span class="si">{</span><span class="n">profile</span><span class="p">[</span><span class="s2">&quot;type&quot;</span><span class="p">]</span><span class="si">:</span><span class="s1">1s</span><span class="si">}</span><span class="s1">   </span><span class="si">{</span><span class="n">lod_string</span><span class="si">:</span><span class="s1">80s</span><span class="si">}</span><span class="s1"> </span><span class="si">{</span><span class="n">counts_string</span><span class="si">:</span><span class="s1">80s</span><span class="si">}</span><span class="s1"> &#39;</span>
                    <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="nb">round</span><span class="p">(</span><span class="n">profile</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;info&quot;</span><span class="p">,</span><span class="w"> </span><span class="mf">0.</span><span class="p">),</span><span class="w"> </span><span class="mi">4</span><span class="p">)</span><span class="si">:</span><span class="s1">4.2f</span><span class="si">}</span><span class="s1"> </span><span class="si">{</span><span class="nb">round</span><span class="p">(</span><span class="n">profile</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;weight&quot;</span><span class="p">,</span><span class="w"> </span><span class="mf">0.</span><span class="p">),</span><span class="w"> </span><span class="mi">4</span><span class="p">)</span><span class="si">:</span><span class="s1">4.2f</span><span class="si">}</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">file_name</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="structure.sequence.format_frequencies" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">format_frequencies</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">format_frequencies</span><span class="p">(</span><span class="n">frequency_list</span><span class="p">:</span> <span class="n">list</span><span class="p">,</span> <span class="n">flip</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">dict</span><span class="p">[</span><span class="n">str</span><span class="p">,</span> <span class="n">dict</span><span class="p">[</span><span class="n">str</span><span class="p">,</span> <span class="n">float</span><span class="p">]]</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Format list of paired frequency data into parsable paired format</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>frequency_list</code></b>
            (<code>list</code>)
        –
        <div class="doc-md-description">
          <p>[(('D', 'A'), 0.0822), (('D', 'V'), 0.0685), ...]</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>flip</code></b>
            (<code>bool</code>, default:
                <code>False</code>
)
        –
        <div class="doc-md-description">
          <p>Whether to invert the mapping of internal tuple</p>
        </div>
      </li>
  </ul>
      <p>Returns:
    {'A': {'S': 0.02, 'T': 0.12}, ...}</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1130</span>
<span class="normal">1131</span>
<span class="normal">1132</span>
<span class="normal">1133</span>
<span class="normal">1134</span>
<span class="normal">1135</span>
<span class="normal">1136</span>
<span class="normal">1137</span>
<span class="normal">1138</span>
<span class="normal">1139</span>
<span class="normal">1140</span>
<span class="normal">1141</span>
<span class="normal">1142</span>
<span class="normal">1143</span>
<span class="normal">1144</span>
<span class="normal">1145</span>
<span class="normal">1146</span>
<span class="normal">1147</span>
<span class="normal">1148</span>
<span class="normal">1149</span>
<span class="normal">1150</span>
<span class="normal">1151</span>
<span class="normal">1152</span>
<span class="normal">1153</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">format_frequencies</span><span class="p">(</span><span class="n">frequency_list</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">flip</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">float</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Format list of paired frequency data into parsable paired format</span>

<span class="sd">    Args:</span>
<span class="sd">        frequency_list: [((&#39;D&#39;, &#39;A&#39;), 0.0822), ((&#39;D&#39;, &#39;V&#39;), 0.0685), ...]</span>
<span class="sd">        flip: Whether to invert the mapping of internal tuple</span>
<span class="sd">    Returns:</span>
<span class="sd">        {&#39;A&#39;: {&#39;S&#39;: 0.02, &#39;T&#39;: 0.12}, ...}</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">flip</span><span class="p">:</span>
        <span class="n">i</span><span class="p">,</span> <span class="n">j</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">i</span><span class="p">,</span> <span class="n">j</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span>
    <span class="n">freq_d</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">tup</span> <span class="ow">in</span> <span class="n">frequency_list</span><span class="p">:</span>
        <span class="n">aa_mapped</span> <span class="o">=</span> <span class="n">tup</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="n">i</span><span class="p">]</span>  <span class="c1"># 0</span>
        <span class="n">aa_paired</span> <span class="o">=</span> <span class="n">tup</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="n">j</span><span class="p">]</span>  <span class="c1"># 1</span>
        <span class="n">freq</span> <span class="o">=</span> <span class="n">tup</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">aa_mapped</span> <span class="ow">in</span> <span class="n">freq_d</span><span class="p">:</span>
            <span class="n">freq_d</span><span class="p">[</span><span class="n">aa_mapped</span><span class="p">][</span><span class="n">aa_paired</span><span class="p">]</span> <span class="o">=</span> <span class="n">freq</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">freq_d</span><span class="p">[</span><span class="n">aa_mapped</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="n">aa_paired</span><span class="p">:</span> <span class="n">freq</span><span class="p">}</span>

    <span class="k">return</span> <span class="n">freq_d</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="structure.sequence.overlap_consensus" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">overlap_consensus</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">overlap_consensus</span><span class="p">(</span><span class="n">issm</span><span class="p">,</span> <span class="n">aa_set</span><span class="p">)</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Find the overlap constrained consensus sequence</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>issm</code></b>
            (<code>dict</code>)
        –
        <div class="doc-md-description">
          <p>{1: {'A': 0.1, 'C': 0.0, ...}, 14: {...}, ...}</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>aa_set</code></b>
            (<code>dict</code>)
        –
        <div class="doc-md-description">
          <p>{residue: {'A', 'I', 'M', 'V'}, ...}</p>
        </div>
      </li>
  </ul>
      <p>Returns:
    (dict): {23: 'T', 29: 'A', ...}</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1156</span>
<span class="normal">1157</span>
<span class="normal">1158</span>
<span class="normal">1159</span>
<span class="normal">1160</span>
<span class="normal">1161</span>
<span class="normal">1162</span>
<span class="normal">1163</span>
<span class="normal">1164</span>
<span class="normal">1165</span>
<span class="normal">1166</span>
<span class="normal">1167</span>
<span class="normal">1168</span>
<span class="normal">1169</span>
<span class="normal">1170</span>
<span class="normal">1171</span>
<span class="normal">1172</span>
<span class="normal">1173</span>
<span class="normal">1174</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">overlap_consensus</span><span class="p">(</span><span class="n">issm</span><span class="p">,</span> <span class="n">aa_set</span><span class="p">):</span>  <span class="c1"># UNUSED</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Find the overlap constrained consensus sequence</span>

<span class="sd">    Args:</span>
<span class="sd">        issm (dict): {1: {&#39;A&#39;: 0.1, &#39;C&#39;: 0.0, ...}, 14: {...}, ...}</span>
<span class="sd">        aa_set (dict): {residue: {&#39;A&#39;, &#39;I&#39;, &#39;M&#39;, &#39;V&#39;}, ...}</span>
<span class="sd">    Returns:</span>
<span class="sd">        (dict): {23: &#39;T&#39;, 29: &#39;A&#39;, ...}</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">consensus</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">aa_set</span><span class="p">:</span>
        <span class="n">max_freq</span> <span class="o">=</span> <span class="mf">0.0</span>
        <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">aa_set</span><span class="p">[</span><span class="n">res</span><span class="p">]:</span>
            <span class="c1"># if max_freq &lt; issm[(res, partner)][]:</span>
            <span class="k">if</span> <span class="n">issm</span><span class="p">[</span><span class="n">res</span><span class="p">][</span><span class="n">aa</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">max_freq</span><span class="p">:</span>
                <span class="n">max_freq</span> <span class="o">=</span> <span class="n">issm</span><span class="p">[</span><span class="n">res</span><span class="p">][</span><span class="n">aa</span><span class="p">]</span>
                <span class="n">consensus</span><span class="p">[</span><span class="n">res</span><span class="p">]</span> <span class="o">=</span> <span class="n">aa</span>

    <span class="k">return</span> <span class="n">consensus</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="structure.sequence.get_cluster_dicts" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">get_cluster_dicts</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">get_cluster_dicts</span><span class="p">(</span><span class="n">db</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="n">putils</span><span class="o">.</span><span class="n">biological_interfaces</span><span class="p">,</span> <span class="n">id_list</span><span class="p">:</span> <span class="n">list</span><span class="p">[</span><span class="n">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">dict</span><span class="p">[</span><span class="n">str</span><span class="p">,</span> <span class="n">dict</span><span class="p">]</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Generate an interface specific scoring matrix from the fragment library</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>db</code></b>
            (<code>str</code>, default:
                <code><span title="structure.sequence.putils.biological_interfaces">biological_interfaces</span></code>
)
        –
        <div class="doc-md-description">
          <p>The source of the fragment information</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>id_list</code></b>
            (<code>list[str]</code>, default:
                <code>None</code>
)
        –
        <div class="doc-md-description">
          <p>[1_2_24, ...]</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
<b><code>cluster_dict</code></b> (            <code>dict[str, dict]</code>
)        –
        <div class="doc-md-description">
          <p>{'1_2_45': {'size': ..., 'rmsd': ..., 'rep': ..., 'mapped': ..., 'paired': ...}, ...}</p>
      </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1177</span>
<span class="normal">1178</span>
<span class="normal">1179</span>
<span class="normal">1180</span>
<span class="normal">1181</span>
<span class="normal">1182</span>
<span class="normal">1183</span>
<span class="normal">1184</span>
<span class="normal">1185</span>
<span class="normal">1186</span>
<span class="normal">1187</span>
<span class="normal">1188</span>
<span class="normal">1189</span>
<span class="normal">1190</span>
<span class="normal">1191</span>
<span class="normal">1192</span>
<span class="normal">1193</span>
<span class="normal">1194</span>
<span class="normal">1195</span>
<span class="normal">1196</span>
<span class="normal">1197</span>
<span class="normal">1198</span>
<span class="normal">1199</span>
<span class="normal">1200</span>
<span class="normal">1201</span>
<span class="normal">1202</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_cluster_dicts</span><span class="p">(</span><span class="n">db</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">putils</span><span class="o">.</span><span class="n">biological_interfaces</span><span class="p">,</span> <span class="n">id_list</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">dict</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Generate an interface specific scoring matrix from the fragment library</span>

<span class="sd">    Args:</span>
<span class="sd">        db: The source of the fragment information</span>
<span class="sd">        id_list: [1_2_24, ...]</span>

<span class="sd">    Returns:</span>
<span class="sd">         cluster_dict: {&#39;1_2_45&#39;: {&#39;size&#39;: ..., &#39;rmsd&#39;: ..., &#39;rep&#39;: ..., &#39;mapped&#39;: ..., &#39;paired&#39;: ...}, ...}</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">info_db</span> <span class="o">=</span> <span class="n">putils</span><span class="o">.</span><span class="n">frag_directory</span><span class="p">[</span><span class="n">db</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">id_list</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">directory_list</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">get_base_root_paths_recursively</span><span class="p">(</span><span class="n">info_db</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">directory_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">_id</span> <span class="ow">in</span> <span class="n">id_list</span><span class="p">:</span>
            <span class="n">c_id</span> <span class="o">=</span> <span class="n">_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)</span>
            <span class="n">_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">info_db</span><span class="p">,</span> <span class="n">c_id</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">c_id</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">c_id</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">c_id</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">c_id</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">c_id</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
            <span class="n">directory_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">_dir</span><span class="p">)</span>

    <span class="n">cluster_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="n">directory_list</span><span class="p">:</span>
        <span class="n">filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cluster</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">cluster</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;.pkl&#39;</span><span class="p">)</span>
        <span class="n">cluster_dict</span><span class="p">[</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">cluster</span><span class="p">)]</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">unpickle</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">cluster_dict</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="structure.sequence.fragment_overlap" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">fragment_overlap</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">fragment_overlap</span><span class="p">(</span><span class="n">residues</span><span class="p">,</span> <span class="n">interaction_graph</span><span class="p">,</span> <span class="n">freq_map</span><span class="p">)</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Take fragment contact list to find the possible AA types allowed in fragment pairs from the contact list</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>residues</code></b>
            (<code>iter</code>)
        –
        <div class="doc-md-description">
          <p>Iterable of residue numbers</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>interaction_graph</code></b>
            (<code>dict</code>)
        –
        <div class="doc-md-description">
          <p>{52: [54, 56, 72, 206], ...}</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>freq_map</code></b>
            (<code>dict</code>)
        –
        <div class="doc-md-description">
          <p>{(78, 87, ...): {'A': {'S': 0.02, 'T': 0.12}, ...}, ...}</p>
        </div>
      </li>
  </ul>
      <p>Returns:
    overlap (dict): {residue: {'A', 'I', 'M', 'V'}, ...}</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1225</span>
<span class="normal">1226</span>
<span class="normal">1227</span>
<span class="normal">1228</span>
<span class="normal">1229</span>
<span class="normal">1230</span>
<span class="normal">1231</span>
<span class="normal">1232</span>
<span class="normal">1233</span>
<span class="normal">1234</span>
<span class="normal">1235</span>
<span class="normal">1236</span>
<span class="normal">1237</span>
<span class="normal">1238</span>
<span class="normal">1239</span>
<span class="normal">1240</span>
<span class="normal">1241</span>
<span class="normal">1242</span>
<span class="normal">1243</span>
<span class="normal">1244</span>
<span class="normal">1245</span>
<span class="normal">1246</span>
<span class="normal">1247</span>
<span class="normal">1248</span>
<span class="normal">1249</span>
<span class="normal">1250</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">fragment_overlap</span><span class="p">(</span><span class="n">residues</span><span class="p">,</span> <span class="n">interaction_graph</span><span class="p">,</span> <span class="n">freq_map</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Take fragment contact list to find the possible AA types allowed in fragment pairs from the contact list</span>

<span class="sd">    Args:</span>
<span class="sd">        residues (iter): Iterable of residue numbers</span>
<span class="sd">        interaction_graph (dict): {52: [54, 56, 72, 206], ...}</span>
<span class="sd">        freq_map (dict): {(78, 87, ...): {&#39;A&#39;: {&#39;S&#39;: 0.02, &#39;T&#39;: 0.12}, ...}, ...}</span>
<span class="sd">    Returns:</span>
<span class="sd">        overlap (dict): {residue: {&#39;A&#39;, &#39;I&#39;, &#39;M&#39;, &#39;V&#39;}, ...}</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">overlap</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">residues</span><span class="p">:</span>
        <span class="n">overlap</span><span class="p">[</span><span class="n">res</span><span class="p">]</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">interaction_graph</span><span class="p">:</span>  <span class="c1"># check for existence as some fragment info is not in the interface set</span>
            <span class="c1"># overlap[res] = set()</span>
            <span class="k">for</span> <span class="n">partner</span> <span class="ow">in</span> <span class="n">interaction_graph</span><span class="p">[</span><span class="n">res</span><span class="p">]:</span>
                <span class="k">if</span> <span class="p">(</span><span class="n">res</span><span class="p">,</span> <span class="n">partner</span><span class="p">)</span> <span class="ow">in</span> <span class="n">freq_map</span><span class="p">:</span>
                    <span class="n">overlap</span><span class="p">[</span><span class="n">res</span><span class="p">]</span> <span class="o">|=</span> <span class="nb">set</span><span class="p">(</span><span class="n">freq_map</span><span class="p">[(</span><span class="n">res</span><span class="p">,</span> <span class="n">partner</span><span class="p">)]</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

    <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">residues</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">interaction_graph</span><span class="p">:</span>  <span class="c1"># check for existence as some fragment info is not in the interface set</span>
            <span class="k">for</span> <span class="n">partner</span> <span class="ow">in</span> <span class="n">interaction_graph</span><span class="p">[</span><span class="n">res</span><span class="p">]:</span>
                <span class="k">if</span> <span class="p">(</span><span class="n">res</span><span class="p">,</span> <span class="n">partner</span><span class="p">)</span> <span class="ow">in</span> <span class="n">freq_map</span><span class="p">:</span>
                    <span class="n">overlap</span><span class="p">[</span><span class="n">res</span><span class="p">]</span> <span class="o">&amp;=</span> <span class="nb">set</span><span class="p">(</span><span class="n">freq_map</span><span class="p">[(</span><span class="n">res</span><span class="p">,</span> <span class="n">partner</span><span class="p">)]</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

    <span class="k">return</span> <span class="n">overlap</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="structure.sequence.offset_index" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">offset_index</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">offset_index</span><span class="p">(</span><span class="n">dictionary</span><span class="p">:</span> <span class="n">dict</span><span class="p">[</span><span class="n">int</span><span class="p">,</span> <span class="n"><span title="typing.Any">Any</span></span><span class="p">],</span> <span class="n">to_zero</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">dict</span><span class="p">[</span><span class="n">int</span><span class="p">,</span> <span class="n">dict</span><span class="p">]</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Modify the index of a sequence dictionary. Default is to one-indexed. to_zero=True gives zero-indexed</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1253</span>
<span class="normal">1254</span>
<span class="normal">1255</span>
<span class="normal">1256</span>
<span class="normal">1257</span>
<span class="normal">1258</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">offset_index</span><span class="p">(</span><span class="n">dictionary</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span> <span class="n">to_zero</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">dict</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Modify the index of a sequence dictionary. Default is to one-indexed. to_zero=True gives zero-indexed&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">to_zero</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">{</span><span class="n">residue</span> <span class="o">-</span> <span class="n">ZERO_OFFSET</span><span class="p">:</span> <span class="n">dictionary</span><span class="p">[</span><span class="n">residue</span><span class="p">]</span> <span class="k">for</span> <span class="n">residue</span> <span class="ow">in</span> <span class="n">dictionary</span><span class="p">}</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">{</span><span class="n">residue</span> <span class="o">+</span> <span class="n">ZERO_OFFSET</span><span class="p">:</span> <span class="n">dictionary</span><span class="p">[</span><span class="n">residue</span><span class="p">]</span> <span class="k">for</span> <span class="n">residue</span> <span class="ow">in</span> <span class="n">dictionary</span><span class="p">}</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="structure.sequence.residue_object_to_number" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">residue_object_to_number</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">residue_object_to_number</span><span class="p">(</span><span class="n">residue_dict</span><span class="p">:</span> <span class="n">dict</span><span class="p">[</span><span class="n">str</span><span class="p">,</span> <span class="n"><span title="collections.abc.Iterable">Iterable</span></span><span class="p">[</span><span class="s1">&#39;structure.base.Residue&#39;</span><span class="p">]])</span> <span class="o">-&gt;</span> <span class="n">dict</span><span class="p">[</span><span class="n">str</span><span class="p">,</span> <span class="n">list</span><span class="p">[</span><span class="n">tuple</span><span class="p">[</span><span class="n">int</span><span class="p">,</span> <span class="o">...</span><span class="p">]]]</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Convert sets of PDB.Residue objects to residue numbers</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>residue_dict</code></b>
            (<code>dict</code>)
        –
        <div class="doc-md-description">
          <p>{'key1': [(residue1_ca_atom, residue2_ca_atom, ...), ...] ...}</p>
        </div>
      </li>
  </ul>
      <p>Returns:
    residue_dict (dict): {'key1': [(78, 87, ...),], ...} - Entry mapped to residue sets</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1261</span>
<span class="normal">1262</span>
<span class="normal">1263</span>
<span class="normal">1264</span>
<span class="normal">1265</span>
<span class="normal">1266</span>
<span class="normal">1267</span>
<span class="normal">1268</span>
<span class="normal">1269</span>
<span class="normal">1270</span>
<span class="normal">1271</span>
<span class="normal">1272</span>
<span class="normal">1273</span>
<span class="normal">1274</span>
<span class="normal">1275</span>
<span class="normal">1276</span>
<span class="normal">1277</span>
<span class="normal">1278</span>
<span class="normal">1279</span>
<span class="normal">1280</span>
<span class="normal">1281</span>
<span class="normal">1282</span>
<span class="normal">1283</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">residue_object_to_number</span><span class="p">(</span>
    <span class="n">residue_dict</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">[</span><span class="s1">&#39;structure.base.Residue&#39;</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">list</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="o">...</span><span class="p">]]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Convert sets of PDB.Residue objects to residue numbers</span>

<span class="sd">    Args:</span>
<span class="sd">        residue_dict (dict): {&#39;key1&#39;: [(residue1_ca_atom, residue2_ca_atom, ...), ...] ...}</span>
<span class="sd">    Returns:</span>
<span class="sd">        residue_dict (dict): {&#39;key1&#39;: [(78, 87, ...),], ...} - Entry mapped to residue sets</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">entry</span> <span class="ow">in</span> <span class="n">residue_dict</span><span class="p">:</span>
        <span class="n">pairs</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="c1"># for _set in range(len(residue_dict[entry])):</span>
        <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">_set</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">residue_dict</span><span class="p">[</span><span class="n">entry</span><span class="p">]):</span>
            <span class="n">residue_num_set</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="c1"># for i, residue in enumerate(residue_dict[entry][_set]):</span>
            <span class="k">for</span> <span class="n">residue</span> <span class="ow">in</span> <span class="n">_set</span><span class="p">:</span>
                <span class="n">resi_number</span> <span class="o">=</span> <span class="n">residue</span><span class="o">.</span><span class="n">residue_number</span>
                <span class="n">residue_num_set</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">resi_number</span><span class="p">)</span>
            <span class="n">pairs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">tuple</span><span class="p">(</span><span class="n">residue_num_set</span><span class="p">))</span>
        <span class="n">residue_dict</span><span class="p">[</span><span class="n">entry</span><span class="p">]</span> <span class="o">=</span> <span class="n">pairs</span>

    <span class="k">return</span> <span class="n">residue_dict</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="structure.sequence.convert_to_residue_cluster_map" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">convert_to_residue_cluster_map</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">convert_to_residue_cluster_map</span><span class="p">(</span><span class="n">residue_cluster_dict</span><span class="p">,</span> <span class="n">frag_range</span><span class="p">)</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Make a residue and cluster/fragment index map</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>residue_cluster_dict</code></b>
            (<code>dict</code>)
        –
        <div class="doc-md-description">
          <p>{'1_2_45': [(residue1_ca_atom, residue2_ca_atom), ...] ...}</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>frag_range</code></b>
            (<code>dict</code>)
        –
        <div class="doc-md-description">
          <p>A range of the fragment size to search over. Ex: (-2, 3) for fragments of length 5</p>
        </div>
      </li>
  </ul>
      <p>Returns:
    cluster_map (dict): {48: {'source': 'mapped', 'cluster': [(-2, 1_1_54), ...]}, ...}
        Where the key is the 0 indexed residue id</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1286</span>
<span class="normal">1287</span>
<span class="normal">1288</span>
<span class="normal">1289</span>
<span class="normal">1290</span>
<span class="normal">1291</span>
<span class="normal">1292</span>
<span class="normal">1293</span>
<span class="normal">1294</span>
<span class="normal">1295</span>
<span class="normal">1296</span>
<span class="normal">1297</span>
<span class="normal">1298</span>
<span class="normal">1299</span>
<span class="normal">1300</span>
<span class="normal">1301</span>
<span class="normal">1302</span>
<span class="normal">1303</span>
<span class="normal">1304</span>
<span class="normal">1305</span>
<span class="normal">1306</span>
<span class="normal">1307</span>
<span class="normal">1308</span>
<span class="normal">1309</span>
<span class="normal">1310</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">convert_to_residue_cluster_map</span><span class="p">(</span><span class="n">residue_cluster_dict</span><span class="p">,</span> <span class="n">frag_range</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Make a residue and cluster/fragment index map</span>

<span class="sd">    Args:</span>
<span class="sd">        residue_cluster_dict (dict): {&#39;1_2_45&#39;: [(residue1_ca_atom, residue2_ca_atom), ...] ...}</span>
<span class="sd">        frag_range (dict): A range of the fragment size to search over. Ex: (-2, 3) for fragments of length 5</span>
<span class="sd">    Returns:</span>
<span class="sd">        cluster_map (dict): {48: {&#39;source&#39;: &#39;mapped&#39;, &#39;cluster&#39;: [(-2, 1_1_54), ...]}, ...}</span>
<span class="sd">            Where the key is the 0 indexed residue id</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">cluster_map</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="n">residue_cluster_dict</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">pair</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">residue_cluster_dict</span><span class="p">[</span><span class="n">cluster</span><span class="p">])):</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">residue_atom</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">residue_cluster_dict</span><span class="p">[</span><span class="n">cluster</span><span class="p">][</span><span class="n">pair</span><span class="p">]):</span>
                <span class="c1"># for each residue in map add the same cluster to the range of fragment residue numbers</span>
                <span class="n">residue_num</span> <span class="o">=</span> <span class="n">residue_atom</span><span class="o">.</span><span class="n">residue_number</span> <span class="o">-</span> <span class="n">ZERO_OFFSET</span>  <span class="c1"># zero index</span>
                <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="o">*</span><span class="n">frag_range</span><span class="p">):</span>
                    <span class="k">if</span> <span class="n">residue_num</span> <span class="o">+</span> <span class="n">j</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">cluster_map</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                            <span class="n">cluster_map</span><span class="p">[</span><span class="n">residue_num</span> <span class="o">+</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;source&#39;</span><span class="p">:</span> <span class="s1">&#39;mapped&#39;</span><span class="p">,</span> <span class="s1">&#39;cluster&#39;</span><span class="p">:</span> <span class="p">[]}</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">cluster_map</span><span class="p">[</span><span class="n">residue_num</span> <span class="o">+</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;source&#39;</span><span class="p">:</span> <span class="s1">&#39;paired&#39;</span><span class="p">,</span> <span class="s1">&#39;cluster&#39;</span><span class="p">:</span> <span class="p">[]}</span>
                    <span class="n">cluster_map</span><span class="p">[</span><span class="n">residue_num</span> <span class="o">+</span> <span class="n">j</span><span class="p">][</span><span class="s1">&#39;cluster&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">j</span><span class="p">,</span> <span class="n">cluster</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">cluster_map</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="structure.sequence.consensus_sequence" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">consensus_sequence</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">consensus_sequence</span><span class="p">(</span><span class="n">pssm</span><span class="p">)</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Return the consensus sequence from a PSSM</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>pssm</code></b>
            (<code>dict</code>)
        –
        <div class="doc-md-description">
          <p>pssm dictionary</p>
        </div>
      </li>
  </ul>
      <p>Return:
    consensus_identities (dict): {1: 'M', 2: 'H', ...} One-indexed</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1436</span>
<span class="normal">1437</span>
<span class="normal">1438</span>
<span class="normal">1439</span>
<span class="normal">1440</span>
<span class="normal">1441</span>
<span class="normal">1442</span>
<span class="normal">1443</span>
<span class="normal">1444</span>
<span class="normal">1445</span>
<span class="normal">1446</span>
<span class="normal">1447</span>
<span class="normal">1448</span>
<span class="normal">1449</span>
<span class="normal">1450</span>
<span class="normal">1451</span>
<span class="normal">1452</span>
<span class="normal">1453</span>
<span class="normal">1454</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">consensus_sequence</span><span class="p">(</span><span class="n">pssm</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Return the consensus sequence from a PSSM</span>

<span class="sd">    Args:</span>
<span class="sd">        pssm (dict): pssm dictionary</span>
<span class="sd">    Return:</span>
<span class="sd">        consensus_identities (dict): {1: &#39;M&#39;, 2: &#39;H&#39;, ...} One-indexed</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">consensus_identities</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">residue</span> <span class="ow">in</span> <span class="n">pssm</span><span class="p">:</span>
        <span class="n">max_lod</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">max_res</span> <span class="o">=</span> <span class="n">pssm</span><span class="p">[</span><span class="n">residue</span><span class="p">][</span><span class="s1">&#39;type&#39;</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">protein_letters_alph3</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">pssm</span><span class="p">[</span><span class="n">residue</span><span class="p">][</span><span class="s1">&#39;lod&#39;</span><span class="p">][</span><span class="n">aa</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">max_lod</span><span class="p">:</span>
                <span class="n">max_lod</span> <span class="o">=</span> <span class="n">pssm</span><span class="p">[</span><span class="n">residue</span><span class="p">][</span><span class="s1">&#39;lod&#39;</span><span class="p">][</span><span class="n">aa</span><span class="p">]</span>
                <span class="n">max_res</span> <span class="o">=</span> <span class="n">aa</span>
        <span class="n">consensus_identities</span><span class="p">[</span><span class="n">residue</span> <span class="o">+</span> <span class="n">ZERO_OFFSET</span><span class="p">]</span> <span class="o">=</span> <span class="n">max_res</span>

    <span class="k">return</span> <span class="n">consensus_identities</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="structure.sequence.sequence_difference" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">sequence_difference</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">sequence_difference</span><span class="p">(</span><span class="n">seq1</span><span class="p">:</span> <span class="n"><span title="collections.abc.Sequence">Sequence</span></span><span class="p">,</span> <span class="n">seq2</span><span class="p">:</span> <span class="n"><span title="collections.abc.Sequence">Sequence</span></span><span class="p">,</span> <span class="n">d</span><span class="p">:</span> <span class="n">dict</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">matrix</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="s1">&#39;BLOSUM62&#39;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">float</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Returns the sequence difference between two sequence iterators</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>seq1</code></b>
            (<code><span title="collections.abc.Sequence">Sequence</span></code>)
        –
        <div class="doc-md-description">
          <p>Either an iterable with residue type as array, or key, with residue type as d[seq1][residue]['type']</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>seq2</code></b>
            (<code><span title="collections.abc.Sequence">Sequence</span></code>)
        –
        <div class="doc-md-description">
          <p>Either an iterable with residue type as array, or key, with residue type as d[seq2][residue]['type']</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>d</code></b>
            (<code>dict</code>, default:
                <code>None</code>
)
        –
        <div class="doc-md-description">
          <p>The dictionary to look up seq1 and seq2 if they are keys in d</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>matrix</code></b>
            (<code>str</code>, default:
                <code>&#39;BLOSUM62&#39;</code>
)
        –
        <div class="doc-md-description">
          <p>The type of matrix to score the sequence differences on</p>
        </div>
      </li>
  </ul>
      <p>Returns:
    The computed sequence difference between seq1 and seq2</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1457</span>
<span class="normal">1458</span>
<span class="normal">1459</span>
<span class="normal">1460</span>
<span class="normal">1461</span>
<span class="normal">1462</span>
<span class="normal">1463</span>
<span class="normal">1464</span>
<span class="normal">1465</span>
<span class="normal">1466</span>
<span class="normal">1467</span>
<span class="normal">1468</span>
<span class="normal">1469</span>
<span class="normal">1470</span>
<span class="normal">1471</span>
<span class="normal">1472</span>
<span class="normal">1473</span>
<span class="normal">1474</span>
<span class="normal">1475</span>
<span class="normal">1476</span>
<span class="normal">1477</span>
<span class="normal">1478</span>
<span class="normal">1479</span>
<span class="normal">1480</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">sequence_difference</span><span class="p">(</span><span class="n">seq1</span><span class="p">:</span> <span class="n">Sequence</span><span class="p">,</span> <span class="n">seq2</span><span class="p">:</span> <span class="n">Sequence</span><span class="p">,</span> <span class="n">d</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">matrix</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;BLOSUM62&#39;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>  <span class="c1"># TODO AMS</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the sequence difference between two sequence iterators</span>

<span class="sd">    Args:</span>
<span class="sd">        seq1: Either an iterable with residue type as array, or key, with residue type as d[seq1][residue][&#39;type&#39;]</span>
<span class="sd">        seq2: Either an iterable with residue type as array, or key, with residue type as d[seq2][residue][&#39;type&#39;]</span>
<span class="sd">        d: The dictionary to look up seq1 and seq2 if they are keys in d</span>
<span class="sd">        matrix: The type of matrix to score the sequence differences on</span>
<span class="sd">    Returns:</span>
<span class="sd">        The computed sequence difference between seq1 and seq2</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">d</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># seq1 = d[seq1]</span>
        <span class="c1"># seq2 = d[seq2]</span>
        <span class="c1"># for residue in d[seq1]:</span>
        <span class="c1">#     s.append((d[seq1][residue][&#39;type&#39;], d[seq2][residue][&#39;type&#39;]))</span>
        <span class="n">pairs</span> <span class="o">=</span> <span class="p">[(</span><span class="n">d</span><span class="p">[</span><span class="n">seq1</span><span class="p">][</span><span class="n">residue</span><span class="p">][</span><span class="s1">&#39;type&#39;</span><span class="p">],</span> <span class="n">d</span><span class="p">[</span><span class="n">seq2</span><span class="p">][</span><span class="n">residue</span><span class="p">][</span><span class="s1">&#39;type&#39;</span><span class="p">])</span> <span class="k">for</span> <span class="n">residue</span> <span class="ow">in</span> <span class="n">d</span><span class="p">[</span><span class="n">seq1</span><span class="p">]]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">pairs</span> <span class="o">=</span> <span class="nb">zip</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">)</span>

    <span class="n">matrix_</span> <span class="o">=</span> <span class="n">substitution_matrices</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">matrix</span><span class="p">)</span>
    <span class="n">scores</span> <span class="o">=</span> <span class="p">[</span><span class="n">matrix_</span><span class="o">.</span><span class="n">get</span><span class="p">((</span><span class="n">letter1</span><span class="p">,</span> <span class="n">letter2</span><span class="p">),</span> <span class="p">(</span><span class="n">letter2</span><span class="p">,</span> <span class="n">letter1</span><span class="p">))</span> <span class="k">for</span> <span class="n">letter1</span><span class="p">,</span> <span class="n">letter2</span> <span class="ow">in</span> <span class="n">pairs</span><span class="p">]</span>

    <span class="k">return</span> <span class="nb">sum</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="structure.sequence.msa_from_dictionary" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">msa_from_dictionary</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">msa_from_dictionary</span><span class="p">(</span><span class="n">named_sequences</span><span class="p">:</span> <span class="n">dict</span><span class="p">[</span><span class="n">str</span><span class="p">,</span> <span class="n">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n"><span title="Bio.Align.MultipleSeqAlignment">MultipleSeqAlignment</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Create a MultipleSequenceAlignment from a dictionary of named sequences</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>named_sequences</code></b>
            (<code>dict[str, str]</code>)
        –
        <div class="doc-md-description">
          <p>{name: sequence, ...} ex: {'clean_asu': 'MNTEELQVAAFEI...', ...}</p>
        </div>
      </li>
  </ul>
      <p>Returns:
    The MultipleSequenceAlignment object for the provided sequences</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1510</span>
<span class="normal">1511</span>
<span class="normal">1512</span>
<span class="normal">1513</span>
<span class="normal">1514</span>
<span class="normal">1515</span>
<span class="normal">1516</span>
<span class="normal">1517</span>
<span class="normal">1518</span>
<span class="normal">1519</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">msa_from_dictionary</span><span class="p">(</span><span class="n">named_sequences</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">MultipleSeqAlignment</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create a MultipleSequenceAlignment from a dictionary of named sequences</span>

<span class="sd">    Args:</span>
<span class="sd">        named_sequences: {name: sequence, ...} ex: {&#39;clean_asu&#39;: &#39;MNTEELQVAAFEI...&#39;, ...}</span>
<span class="sd">    Returns:</span>
<span class="sd">        The MultipleSequenceAlignment object for the provided sequences</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">MultipleSeqAlignment</span><span class="p">([</span><span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="n">sequence</span><span class="p">),</span> <span class="n">annotations</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;molecule_type&#39;</span><span class="p">:</span> <span class="s1">&#39;Protein&#39;</span><span class="p">},</span> <span class="nb">id</span><span class="o">=</span><span class="n">name</span><span class="p">)</span>
                                 <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">named_sequences</span><span class="o">.</span><span class="n">items</span><span class="p">()])</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="structure.sequence.msa_from_seq_records" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">msa_from_seq_records</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">msa_from_seq_records</span><span class="p">(</span><span class="n">seq_records</span><span class="p">:</span> <span class="n"><span title="collections.abc.Iterable">Iterable</span></span><span class="p">[</span><span class="n"><span title="Bio.SeqRecord.SeqRecord">SeqRecord</span></span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n"><span title="Bio.Align.MultipleSeqAlignment">MultipleSeqAlignment</span></span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Create a BioPython Multiple Sequence Alignment from a SeqRecord Iterable</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>seq_records</code></b>
            (<code><span title="collections.abc.Iterable">Iterable</span>[<span title="Bio.SeqRecord.SeqRecord">SeqRecord</span>]</code>)
        –
        <div class="doc-md-description">
          <p>{name: sequence, ...} ex: {'clean_asu': 'MNTEELQVAAFEI...', ...}</p>
        </div>
      </li>
  </ul>
      <p>Returns:
    [SeqRecord(Seq('MNTEELQVAAFEI...', ...), id="Alpha"),
     SeqRecord(Seq('MNTEEL-VAAFEI...', ...), id="Beta"), ...]</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1522</span>
<span class="normal">1523</span>
<span class="normal">1524</span>
<span class="normal">1525</span>
<span class="normal">1526</span>
<span class="normal">1527</span>
<span class="normal">1528</span>
<span class="normal">1529</span>
<span class="normal">1530</span>
<span class="normal">1531</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">msa_from_seq_records</span><span class="p">(</span><span class="n">seq_records</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="n">SeqRecord</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">MultipleSeqAlignment</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create a BioPython Multiple Sequence Alignment from a SeqRecord Iterable</span>

<span class="sd">    Args:</span>
<span class="sd">        seq_records: {name: sequence, ...} ex: {&#39;clean_asu&#39;: &#39;MNTEELQVAAFEI...&#39;, ...}</span>
<span class="sd">    Returns:</span>
<span class="sd">        [SeqRecord(Seq(&#39;MNTEELQVAAFEI...&#39;, ...), id=&quot;Alpha&quot;),</span>
<span class="sd">         SeqRecord(Seq(&#39;MNTEEL-VAAFEI...&#39;, ...), id=&quot;Beta&quot;), ...]</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">MultipleSeqAlignment</span><span class="p">(</span><span class="n">seq_records</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="structure.sequence.make_mutations_chain_agnostic" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">make_mutations_chain_agnostic</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">make_mutations_chain_agnostic</span><span class="p">(</span><span class="n">mutations</span><span class="p">)</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Remove chain identifier from mutation dictionary</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>mutations</code></b>
            (<code>dict</code>)
        –
        <div class="doc-md-description">
          <p>{design: {chain_id: {mutation_index: {'from': 'A', 'to': 'K'}, ...}, ...}, ...}</p>
        </div>
      </li>
  </ul>
      <p>Returns:
    (dict): {design: {mutation_index: {'from': 'A', 'to': 'K'}, ...}, ...}</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1538</span>
<span class="normal">1539</span>
<span class="normal">1540</span>
<span class="normal">1541</span>
<span class="normal">1542</span>
<span class="normal">1543</span>
<span class="normal">1544</span>
<span class="normal">1545</span>
<span class="normal">1546</span>
<span class="normal">1547</span>
<span class="normal">1548</span>
<span class="normal">1549</span>
<span class="normal">1550</span>
<span class="normal">1551</span>
<span class="normal">1552</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">make_mutations_chain_agnostic</span><span class="p">(</span><span class="n">mutations</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Remove chain identifier from mutation dictionary</span>

<span class="sd">    Args:</span>
<span class="sd">        mutations (dict): {design: {chain_id: {mutation_index: {&#39;from&#39;: &#39;A&#39;, &#39;to&#39;: &#39;K&#39;}, ...}, ...}, ...}</span>
<span class="sd">    Returns:</span>
<span class="sd">        (dict): {design: {mutation_index: {&#39;from&#39;: &#39;A&#39;, &#39;to&#39;: &#39;K&#39;}, ...}, ...}</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">flattened_mutations</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">design</span><span class="p">,</span> <span class="n">chain_mutations</span> <span class="ow">in</span> <span class="n">mutations</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">flattened_mutations</span><span class="p">[</span><span class="n">design</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">chain</span><span class="p">,</span> <span class="n">mutations</span> <span class="ow">in</span> <span class="n">chain_mutations</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">flattened_mutations</span><span class="p">[</span><span class="n">design</span><span class="p">]</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">mutations</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">flattened_mutations</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="structure.sequence.simplify_mutation_dict" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">simplify_mutation_dict</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">simplify_mutation_dict</span><span class="p">(</span><span class="n">mutations</span><span class="p">:</span> <span class="n">dict</span><span class="p">[</span><span class="n">str</span><span class="p">,</span> <span class="n"><span title="symdesign.sequence.mutation_dictionary">mutation_dictionary</span></span><span class="p">],</span> <span class="n">to</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">dict</span><span class="p">[</span><span class="n">str</span><span class="p">,</span> <span class="n"><span title="symdesign.sequence.mutation_dictionary">mutation_dictionary</span></span><span class="p">]</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Simplify mutation dictionary to 'to'/'from' AA key</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>mutations</code></b>
            (<code>dict[str, <span title="symdesign.sequence.mutation_dictionary">mutation_dictionary</span>]</code>)
        –
        <div class="doc-md-description">
          <p>Ex: {alias: {mutation_index: {'from': 'A', 'to': 'K'}, ...}, ...}, ...}</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>to</code></b>
            (<code>bool</code>, default:
                <code>True</code>
)
        –
        <div class="doc-md-description">
          <p>Whether to simplify with the 'to' AA key (True) or the 'from' AA key (False)</p>
        </div>
      </li>
  </ul>
      <p>Returns:
    The simplified mutation dictionary. Ex: {alias: {mutation_index: 'K', ...}, ...}</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1555</span>
<span class="normal">1556</span>
<span class="normal">1557</span>
<span class="normal">1558</span>
<span class="normal">1559</span>
<span class="normal">1560</span>
<span class="normal">1561</span>
<span class="normal">1562</span>
<span class="normal">1563</span>
<span class="normal">1564</span>
<span class="normal">1565</span>
<span class="normal">1566</span>
<span class="normal">1567</span>
<span class="normal">1568</span>
<span class="normal">1569</span>
<span class="normal">1570</span>
<span class="normal">1571</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">simplify_mutation_dict</span><span class="p">(</span><span class="n">mutations</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">mutation_dictionary</span><span class="p">],</span> <span class="n">to</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span> \
        <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">mutation_dictionary</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Simplify mutation dictionary to &#39;to&#39;/&#39;from&#39; AA key</span>

<span class="sd">    Args:</span>
<span class="sd">        mutations: Ex: {alias: {mutation_index: {&#39;from&#39;: &#39;A&#39;, &#39;to&#39;: &#39;K&#39;}, ...}, ...}, ...}</span>
<span class="sd">        to: Whether to simplify with the &#39;to&#39; AA key (True) or the &#39;from&#39; AA key (False)</span>
<span class="sd">    Returns:</span>
<span class="sd">        The simplified mutation dictionary. Ex: {alias: {mutation_index: &#39;K&#39;, ...}, ...}</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">simplification</span> <span class="o">=</span> <span class="s1">&#39;to&#39;</span> <span class="k">if</span> <span class="n">to</span> <span class="k">else</span> <span class="s1">&#39;from&#39;</span>

    <span class="k">for</span> <span class="n">alias</span><span class="p">,</span> <span class="n">indexed_mutations</span> <span class="ow">in</span> <span class="n">mutations</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">indexed_mutations</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">mutations</span><span class="p">[</span><span class="n">alias</span><span class="p">][</span><span class="n">index</span><span class="p">]</span> <span class="o">=</span> <span class="n">mutation</span><span class="p">[</span><span class="n">simplification</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">mutations</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="structure.sequence.weave_mutation_dict" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">weave_mutation_dict</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">weave_mutation_dict</span><span class="p">(</span><span class="n">sorted_freq</span><span class="p">,</span> <span class="n">mut_prob</span><span class="p">,</span> <span class="n">resi_divergence</span><span class="p">,</span> <span class="n">int_divergence</span><span class="p">,</span> <span class="n">des_divergence</span><span class="p">)</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Make final dictionary, index to sequence</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>sorted_freq</code></b>
            (<code>dict</code>)
        –
        <div class="doc-md-description">
          <p>{15: ['S', 'A', 'T'], ... }</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>mut_prob</code></b>
            (<code>dict</code>)
        –
        <div class="doc-md-description">
          <p>{15: {'A': 0.05, 'C': 0.001, 'D': 0.1, ...}, 16: {}, ...}</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>resi_divergence</code></b>
            (<code>dict</code>)
        –
        <div class="doc-md-description">
          <p>{15: 0.732, 16: 0.552, ...}</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>int_divergence</code></b>
            (<code>dict</code>)
        –
        <div class="doc-md-description">
          <p>{15: 0.732, 16: 0.552, ...}</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>des_divergence</code></b>
            (<code>dict</code>)
        –
        <div class="doc-md-description">
          <p>{15: 0.732, 16: 0.552, ...}</p>
        </div>
      </li>
  </ul>
      <p>Returns:
    weaved_dict (dict): {16: {'S': 0.134, 'A': 0.050, ..., 'jsd': 0.732, 'int_jsd': 0.412}, ...}</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1574</span>
<span class="normal">1575</span>
<span class="normal">1576</span>
<span class="normal">1577</span>
<span class="normal">1578</span>
<span class="normal">1579</span>
<span class="normal">1580</span>
<span class="normal">1581</span>
<span class="normal">1582</span>
<span class="normal">1583</span>
<span class="normal">1584</span>
<span class="normal">1585</span>
<span class="normal">1586</span>
<span class="normal">1587</span>
<span class="normal">1588</span>
<span class="normal">1589</span>
<span class="normal">1590</span>
<span class="normal">1591</span>
<span class="normal">1592</span>
<span class="normal">1593</span>
<span class="normal">1594</span>
<span class="normal">1595</span>
<span class="normal">1596</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">weave_mutation_dict</span><span class="p">(</span><span class="n">sorted_freq</span><span class="p">,</span> <span class="n">mut_prob</span><span class="p">,</span> <span class="n">resi_divergence</span><span class="p">,</span> <span class="n">int_divergence</span><span class="p">,</span> <span class="n">des_divergence</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Make final dictionary, index to sequence</span>

<span class="sd">    Args:</span>
<span class="sd">        sorted_freq (dict): {15: [&#39;S&#39;, &#39;A&#39;, &#39;T&#39;], ... }</span>
<span class="sd">        mut_prob (dict): {15: {&#39;A&#39;: 0.05, &#39;C&#39;: 0.001, &#39;D&#39;: 0.1, ...}, 16: {}, ...}</span>
<span class="sd">        resi_divergence (dict): {15: 0.732, 16: 0.552, ...}</span>
<span class="sd">        int_divergence (dict): {15: 0.732, 16: 0.552, ...}</span>
<span class="sd">        des_divergence (dict): {15: 0.732, 16: 0.552, ...}</span>
<span class="sd">    Returns:</span>
<span class="sd">        weaved_dict (dict): {16: {&#39;S&#39;: 0.134, &#39;A&#39;: 0.050, ..., &#39;jsd&#39;: 0.732, &#39;int_jsd&#39;: 0.412}, ...}</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">weaved_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">residue</span> <span class="ow">in</span> <span class="n">sorted_freq</span><span class="p">:</span>
        <span class="n">final_resi</span> <span class="o">=</span> <span class="n">residue</span> <span class="o">+</span> <span class="n">ZERO_OFFSET</span>
        <span class="n">weaved_dict</span><span class="p">[</span><span class="n">final_resi</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">sorted_freq</span><span class="p">[</span><span class="n">residue</span><span class="p">]:</span>
            <span class="n">weaved_dict</span><span class="p">[</span><span class="n">final_resi</span><span class="p">][</span><span class="n">aa</span><span class="p">]</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="n">mut_prob</span><span class="p">[</span><span class="n">residue</span><span class="p">][</span><span class="n">aa</span><span class="p">],</span> <span class="mi">3</span><span class="p">)</span>
        <span class="n">weaved_dict</span><span class="p">[</span><span class="n">final_resi</span><span class="p">][</span><span class="s1">&#39;jsd&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">resi_divergence</span><span class="p">[</span><span class="n">residue</span><span class="p">]</span>
        <span class="n">weaved_dict</span><span class="p">[</span><span class="n">final_resi</span><span class="p">][</span><span class="s1">&#39;int_jsd&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">int_divergence</span><span class="p">[</span><span class="n">residue</span><span class="p">]</span>
        <span class="n">weaved_dict</span><span class="p">[</span><span class="n">final_resi</span><span class="p">][</span><span class="s1">&#39;des_jsd&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">des_divergence</span><span class="p">[</span><span class="n">residue</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">weaved_dict</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="structure.sequence.clean_gaped_columns" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">clean_gaped_columns</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">clean_gaped_columns</span><span class="p">(</span><span class="n">alignment_dict</span><span class="p">,</span> <span class="n">correct_index</span><span class="p">)</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Cleans an alignment dictionary by revising key list with correctly indexed positions. 0 indexed</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1599</span>
<span class="normal">1600</span>
<span class="normal">1601</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">clean_gaped_columns</span><span class="p">(</span><span class="n">alignment_dict</span><span class="p">,</span> <span class="n">correct_index</span><span class="p">):</span>  <span class="c1"># UNUSED</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Cleans an alignment dictionary by revising key list with correctly indexed positions. 0 indexed&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">{</span><span class="n">i</span><span class="p">:</span> <span class="n">alignment_dict</span><span class="p">[</span><span class="n">index</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">index</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">correct_index</span><span class="p">)}</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="structure.sequence.msa_to_prob_distribution" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">msa_to_prob_distribution</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">msa_to_prob_distribution</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Turn Alignment dictionary into a probability distribution</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>alignment</code></b>
            (<code>dict</code>)
        –
        <div class="doc-md-description">
          <p>{'meta': {'num_sequences': 214, 'query': 'MGSTHLVLK...', 'query_with_gaps': 'MGS--THLVLK...'},
               'msa': (Bio.Align.MultipleSeqAlignment)
               'counts': {1: {'A': 13, 'C': 1, 'D': 23, ...}, 2: {}, ...},
               'frequencies': {1: {'A': 0.05, 'C': 0.001, 'D': 0.1, ...}, 2: {}, ...},
               'rep': {1: 210, 2:211, ...}}
The msa formatted with counts and indexed by residue</p>
        </div>
      </li>
  </ul>
      <p>Returns:
    (dict): {'meta': {'num_sequences': 214, 'query': 'MGSTHLVLK...', 'query_with_gaps': 'MGS--THLVLK...'},
             'msa': (Bio.Align.MultipleSeqAlignment)
             'counts': {1: {'A': 13, 'C': 1, 'D': 23, ...}, 2: {}, ...},
             'frequencies': {1: {'A': 0.05, 'C': 0.001, 'D': 0.1, ...}, 2: {}, ...},
             'rep': {1: 210, 2:211, ...}}
        The msa formatted with counts and indexed by residue</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1609</span>
<span class="normal">1610</span>
<span class="normal">1611</span>
<span class="normal">1612</span>
<span class="normal">1613</span>
<span class="normal">1614</span>
<span class="normal">1615</span>
<span class="normal">1616</span>
<span class="normal">1617</span>
<span class="normal">1618</span>
<span class="normal">1619</span>
<span class="normal">1620</span>
<span class="normal">1621</span>
<span class="normal">1622</span>
<span class="normal">1623</span>
<span class="normal">1624</span>
<span class="normal">1625</span>
<span class="normal">1626</span>
<span class="normal">1627</span>
<span class="normal">1628</span>
<span class="normal">1629</span>
<span class="normal">1630</span>
<span class="normal">1631</span>
<span class="normal">1632</span>
<span class="normal">1633</span>
<span class="normal">1634</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">msa_to_prob_distribution</span><span class="p">(</span><span class="n">alignment</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Turn Alignment dictionary into a probability distribution</span>

<span class="sd">    Args:</span>
<span class="sd">        alignment (dict): {&#39;meta&#39;: {&#39;num_sequences&#39;: 214, &#39;query&#39;: &#39;MGSTHLVLK...&#39;, &#39;query_with_gaps&#39;: &#39;MGS--THLVLK...&#39;},</span>
<span class="sd">                           &#39;msa&#39;: (Bio.Align.MultipleSeqAlignment)</span>
<span class="sd">                           &#39;counts&#39;: {1: {&#39;A&#39;: 13, &#39;C&#39;: 1, &#39;D&#39;: 23, ...}, 2: {}, ...},</span>
<span class="sd">                           &#39;frequencies&#39;: {1: {&#39;A&#39;: 0.05, &#39;C&#39;: 0.001, &#39;D&#39;: 0.1, ...}, 2: {}, ...},</span>
<span class="sd">                           &#39;rep&#39;: {1: 210, 2:211, ...}}</span>
<span class="sd">            The msa formatted with counts and indexed by residue</span>
<span class="sd">    Returns:</span>
<span class="sd">        (dict): {&#39;meta&#39;: {&#39;num_sequences&#39;: 214, &#39;query&#39;: &#39;MGSTHLVLK...&#39;, &#39;query_with_gaps&#39;: &#39;MGS--THLVLK...&#39;},</span>
<span class="sd">                 &#39;msa&#39;: (Bio.Align.MultipleSeqAlignment)</span>
<span class="sd">                 &#39;counts&#39;: {1: {&#39;A&#39;: 13, &#39;C&#39;: 1, &#39;D&#39;: 23, ...}, 2: {}, ...},</span>
<span class="sd">                 &#39;frequencies&#39;: {1: {&#39;A&#39;: 0.05, &#39;C&#39;: 0.001, &#39;D&#39;: 0.1, ...}, 2: {}, ...},</span>
<span class="sd">                 &#39;rep&#39;: {1: 210, 2:211, ...}}</span>
<span class="sd">            The msa formatted with counts and indexed by residue</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">alignment</span><span class="p">[</span><span class="s1">&#39;frequencies&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">residue</span><span class="p">,</span> <span class="n">amino_acid_counts</span> <span class="ow">in</span> <span class="n">alignment</span><span class="p">[</span><span class="s1">&#39;counts&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">total_column_weight</span> <span class="o">=</span> <span class="n">alignment</span><span class="p">[</span><span class="s1">&#39;rep&#39;</span><span class="p">][</span><span class="n">residue</span><span class="p">]</span>
        <span class="k">assert</span> <span class="n">total_column_weight</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;</span><span class="si">%s</span><span class="s1">: Processing error... Downstream cannot divide by 0. Position = </span><span class="si">%s</span><span class="s1">&#39;</span> \
                                         <span class="o">%</span> <span class="p">(</span><span class="n">msa_to_prob_distribution</span><span class="o">.</span><span class="vm">__name__</span><span class="p">,</span> <span class="n">residue</span><span class="p">)</span>
        <span class="n">alignment</span><span class="p">[</span><span class="s1">&#39;frequencies&#39;</span><span class="p">][</span><span class="n">residue</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="n">aa</span><span class="p">:</span> <span class="n">count</span> <span class="o">/</span> <span class="n">total_column_weight</span> <span class="k">for</span> <span class="n">aa</span><span class="p">,</span> <span class="n">count</span> <span class="ow">in</span> <span class="n">amino_acid_counts</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>

    <span class="k">return</span> <span class="n">alignment</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="structure.sequence.window_score" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">window_score</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">window_score</span><span class="p">(</span><span class="n">scores</span><span class="p">:</span> <span class="n">dict</span><span class="p">[</span><span class="n">int</span><span class="p">,</span> <span class="n">float</span><span class="p">]</span> <span class="o">|</span> <span class="n"><span title="collections.abc.Sequence">Sequence</span></span><span class="p">[</span><span class="n">float</span><span class="p">],</span> <span class="n">window_len</span><span class="p">:</span> <span class="n">int</span><span class="p">,</span> <span class="n">lambda_</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">dict</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Takes an MSA score dict and transforms so that each position is a weighted average of the surrounding positions.
Positions with scores less than zero are not changed and are ignored calculation</p>
<p>Modified from Capra and Singh 2007 code</p>

<details class="notes" open>
  <summary>Notes</summary>
  <p>The input should be a one-indexed dictionary (if a dictionary)</p>
</details>


  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>scores</code></b>
            (<code>dict[int, float] | <span title="collections.abc.Sequence">Sequence</span>[float]</code>)
        –
        <div class="doc-md-description">
          <p>A dictionary with scores.</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>window_len</code></b>
            (<code>int</code>)
        –
        <div class="doc-md-description">
          <p>Number of residues on either side of the current residue</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>lambda_</code></b>
            (<code>float</code>, default:
                <code>0.5</code>
)
        –
        <div class="doc-md-description">
          <p>Float between 0 and 1 which parameterizes the amount of the score to pull from</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
            <code>dict</code>
        –
        <div class="doc-md-description">
          <p>A dictionary with the modified scores for the specified window, one-indexed</p>
      </div>
      </li>
  </ul>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1644</span>
<span class="normal">1645</span>
<span class="normal">1646</span>
<span class="normal">1647</span>
<span class="normal">1648</span>
<span class="normal">1649</span>
<span class="normal">1650</span>
<span class="normal">1651</span>
<span class="normal">1652</span>
<span class="normal">1653</span>
<span class="normal">1654</span>
<span class="normal">1655</span>
<span class="normal">1656</span>
<span class="normal">1657</span>
<span class="normal">1658</span>
<span class="normal">1659</span>
<span class="normal">1660</span>
<span class="normal">1661</span>
<span class="normal">1662</span>
<span class="normal">1663</span>
<span class="normal">1664</span>
<span class="normal">1665</span>
<span class="normal">1666</span>
<span class="normal">1667</span>
<span class="normal">1668</span>
<span class="normal">1669</span>
<span class="normal">1670</span>
<span class="normal">1671</span>
<span class="normal">1672</span>
<span class="normal">1673</span>
<span class="normal">1674</span>
<span class="normal">1675</span>
<span class="normal">1676</span>
<span class="normal">1677</span>
<span class="normal">1678</span>
<span class="normal">1679</span>
<span class="normal">1680</span>
<span class="normal">1681</span>
<span class="normal">1682</span>
<span class="normal">1683</span>
<span class="normal">1684</span>
<span class="normal">1685</span>
<span class="normal">1686</span>
<span class="normal">1687</span>
<span class="normal">1688</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">window_score</span><span class="p">(</span><span class="n">scores</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">float</span><span class="p">]</span> <span class="o">|</span> <span class="n">Sequence</span><span class="p">[</span><span class="nb">float</span><span class="p">],</span> <span class="n">window_len</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">lambda_</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>  <span class="c1"># UNUSED  incorporate into MultipleSequenceAlignment</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Takes an MSA score dict and transforms so that each position is a weighted average of the surrounding positions.</span>
<span class="sd">    Positions with scores less than zero are not changed and are ignored calculation</span>

<span class="sd">    Modified from Capra and Singh 2007 code</span>

<span class="sd">    Notes:</span>
<span class="sd">        The input should be a one-indexed dictionary (if a dictionary)</span>

<span class="sd">    Args:</span>
<span class="sd">        scores: A dictionary with scores.</span>
<span class="sd">        window_len: Number of residues on either side of the current residue</span>
<span class="sd">        lambda_: Float between 0 and 1 which parameterizes the amount of the score to pull from</span>

<span class="sd">    Returns:</span>
<span class="sd">        A dictionary with the modified scores for the specified window, one-indexed</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">window_len</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">return</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">enumerate</span><span class="p">(</span><span class="n">scores</span><span class="o">.</span><span class="n">values</span><span class="p">(),</span> <span class="n">ZERO_OFFSET</span><span class="p">))</span>
        <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>  <span class="c1"># Not a dict</span>
            <span class="k">return</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">enumerate</span><span class="p">(</span><span class="n">scores</span><span class="p">,</span> <span class="n">ZERO_OFFSET</span><span class="p">))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">window_scores</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">number_scores</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">number_scores</span> <span class="o">+</span> <span class="n">ZERO_OFFSET</span><span class="p">):</span>
            <span class="n">s</span> <span class="o">=</span> <span class="n">number_terms</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">if</span> <span class="n">i</span> <span class="o">&lt;=</span> <span class="n">window_len</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">ZERO_OFFSET</span><span class="p">,</span> <span class="n">i</span> <span class="o">+</span> <span class="n">window_len</span> <span class="o">+</span> <span class="n">ZERO_OFFSET</span><span class="p">):</span>
                    <span class="k">if</span> <span class="n">i</span> <span class="o">!=</span> <span class="n">j</span><span class="p">:</span>
                        <span class="n">number_terms</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="n">s</span> <span class="o">+=</span> <span class="n">scores</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
            <span class="k">elif</span> <span class="n">i</span> <span class="o">+</span> <span class="n">window_len</span> <span class="o">&gt;</span> <span class="n">number_scores</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">i</span> <span class="o">-</span> <span class="n">window_len</span><span class="p">,</span> <span class="n">number_scores</span> <span class="o">+</span> <span class="n">ZERO_OFFSET</span><span class="p">):</span>
                    <span class="k">if</span> <span class="n">i</span> <span class="o">!=</span> <span class="n">j</span><span class="p">:</span>
                        <span class="n">number_terms</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="n">s</span> <span class="o">+=</span> <span class="n">scores</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">i</span> <span class="o">-</span> <span class="n">window_len</span><span class="p">,</span> <span class="n">i</span> <span class="o">+</span> <span class="n">window_len</span> <span class="o">+</span> <span class="n">ZERO_OFFSET</span><span class="p">):</span>
                    <span class="k">if</span> <span class="n">i</span> <span class="o">!=</span> <span class="n">j</span><span class="p">:</span>
                        <span class="n">number_terms</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="n">s</span> <span class="o">+=</span> <span class="n">scores</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
            <span class="n">window_scores</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">lambda_</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">s</span> <span class="o">/</span> <span class="n">number_terms</span><span class="p">)</span> <span class="o">+</span> <span class="n">lambda_</span> <span class="o">*</span> <span class="n">scores</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>

        <span class="k">return</span> <span class="n">window_scores</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="structure.sequence.rank_possibilities" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">rank_possibilities</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">rank_possibilities</span><span class="p">(</span><span class="n">probability_dict</span><span class="p">)</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Gather alternative residues and sort them by probability.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>probability_dict</code></b>
            (<code>dict</code>)
        –
        <div class="doc-md-description">
          <p>{15: {'A': 0.05, 'C': 0.001, 'D': 0.1, ...}, 16: {}, ...}</p>
        </div>
      </li>
  </ul>
      <p>Returns:
     sorted_alternates_dict (dict): {15: ['S', 'A', 'T'], ... }</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1691</span>
<span class="normal">1692</span>
<span class="normal">1693</span>
<span class="normal">1694</span>
<span class="normal">1695</span>
<span class="normal">1696</span>
<span class="normal">1697</span>
<span class="normal">1698</span>
<span class="normal">1699</span>
<span class="normal">1700</span>
<span class="normal">1701</span>
<span class="normal">1702</span>
<span class="normal">1703</span>
<span class="normal">1704</span>
<span class="normal">1705</span>
<span class="normal">1706</span>
<span class="normal">1707</span>
<span class="normal">1708</span>
<span class="normal">1709</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">rank_possibilities</span><span class="p">(</span><span class="n">probability_dict</span><span class="p">):</span>  <span class="c1"># UNUSED  incorporate into MultipleSequenceAlignment</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Gather alternative residues and sort them by probability.</span>

<span class="sd">    Args:</span>
<span class="sd">        probability_dict (dict): {15: {&#39;A&#39;: 0.05, &#39;C&#39;: 0.001, &#39;D&#39;: 0.1, ...}, 16: {}, ...}</span>
<span class="sd">    Returns:</span>
<span class="sd">         sorted_alternates_dict (dict): {15: [&#39;S&#39;, &#39;A&#39;, &#39;T&#39;], ... }</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sorted_alternates_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">residue</span> <span class="ow">in</span> <span class="n">probability_dict</span><span class="p">:</span>
        <span class="n">residue_probability_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">probability_dict</span><span class="p">[</span><span class="n">residue</span><span class="p">]:</span>
            <span class="k">if</span> <span class="n">probability_dict</span><span class="p">[</span><span class="n">residue</span><span class="p">][</span><span class="n">aa</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">residue_probability_list</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">aa</span><span class="p">,</span> <span class="nb">round</span><span class="p">(</span><span class="n">probability_dict</span><span class="p">[</span><span class="n">residue</span><span class="p">][</span><span class="n">aa</span><span class="p">],</span> <span class="mi">5</span><span class="p">)))</span>  <span class="c1"># tuple instead of list</span>
        <span class="n">residue_probability_list</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">tup</span><span class="p">:</span> <span class="n">tup</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="c1"># [(&#39;S&#39;, 0.13190), (&#39;A&#39;, 0.0500), ...]</span>
        <span class="n">sorted_alternates_dict</span><span class="p">[</span><span class="n">residue</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">aa</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">residue_probability_list</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">sorted_alternates_dict</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="structure.sequence.multi_chain_alignment" class="doc doc-heading">
          <span class="doc doc-object-name doc-function-name">multi_chain_alignment</span>


</h2>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">multi_chain_alignment</span><span class="p">(</span><span class="n">mutated_sequences</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</code></pre></div>

  <div class="doc doc-contents ">
  
      <p>Combines different chain's Multiple Sequence Alignments into a single MSA. One-indexed</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>mutated_sequences</code></b>
            (<code>dict</code>)
        –
        <div class="doc-md-description">
          <p>{chain: {name: sequence, ...}</p>
        </div>
      </li>
  </ul>
      <p>Returns:
    (MultipleSequenceAlignment): The MSA object with counts, frequencies, sequences, and indexed by residue</p>

          <details class="quote">
            <summary>Source code in <code>symdesign/structure/sequence.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1712</span>
<span class="normal">1713</span>
<span class="normal">1714</span>
<span class="normal">1715</span>
<span class="normal">1716</span>
<span class="normal">1717</span>
<span class="normal">1718</span>
<span class="normal">1719</span>
<span class="normal">1720</span>
<span class="normal">1721</span>
<span class="normal">1722</span>
<span class="normal">1723</span>
<span class="normal">1724</span>
<span class="normal">1725</span>
<span class="normal">1726</span>
<span class="normal">1727</span>
<span class="normal">1728</span>
<span class="normal">1729</span>
<span class="normal">1730</span>
<span class="normal">1731</span>
<span class="normal">1732</span>
<span class="normal">1733</span>
<span class="normal">1734</span>
<span class="normal">1735</span>
<span class="normal">1736</span>
<span class="normal">1737</span>
<span class="normal">1738</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">multi_chain_alignment</span><span class="p">(</span><span class="n">mutated_sequences</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Combines different chain&#39;s Multiple Sequence Alignments into a single MSA. One-indexed</span>

<span class="sd">    Args:</span>
<span class="sd">        mutated_sequences (dict): {chain: {name: sequence, ...}</span>
<span class="sd">    Returns:</span>
<span class="sd">        (MultipleSequenceAlignment): The MSA object with counts, frequencies, sequences, and indexed by residue</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#         (dict): {&#39;meta&#39;: {&#39;num_sequences&#39;: 214, &#39;query&#39;: &#39;MGSTHLVLK...&#39;, &#39;query_with_gaps&#39;: &#39;MGS--THLVLK...&#39;},</span>
    <span class="c1">#                  &#39;msa&#39;: (Bio.Align.MultipleSeqAlignment)</span>
    <span class="c1">#                  &#39;counts&#39;: {1: {&#39;A&#39;: 13, &#39;C&#39;: 1, &#39;D&#39;: 23, ...}, 2: {}, ...},</span>
    <span class="c1">#                  &#39;frequencies&#39;: {1: {&#39;A&#39;: 0.05, &#39;C&#39;: 0.001, &#39;D&#39;: 0.1, ...}, 2: {}, ...},</span>
    <span class="c1">#                  &#39;rep&#39;: {1: 210, 2:211, ...}}</span>
    <span class="c1">#             The msa formatted with counts and indexed by residue</span>

    <span class="c1"># Combine alignments for all chains from design file Ex: A: 1-102, B: 1-130. Alignment: 1-232</span>
    <span class="n">total_alignment</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">named_sequences</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">mutated_sequences</span><span class="o">.</span><span class="n">values</span><span class="p">()):</span>
        <span class="k">if</span> <span class="n">idx</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">total_alignment</span> <span class="o">=</span> <span class="n">msa_from_dictionary</span><span class="p">(</span><span class="n">named_sequences</span><span class="p">)[:,</span> <span class="p">:]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">total_alignment</span> <span class="o">+=</span> <span class="n">msa_from_dictionary</span><span class="p">(</span><span class="n">named_sequences</span><span class="p">)[:,</span> <span class="p">:]</span>

    <span class="k">if</span> <span class="n">total_alignment</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">MultipleSequenceAlignment</span><span class="p">(</span><span class="n">alignment</span><span class="o">=</span><span class="n">total_alignment</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">multi_chain_alignment</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s1"> - No sequences were found!&#39;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>

</div>












                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var tab,labels=set.querySelector(".tabbed-labels");for(tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../model/" class="md-footer__link md-footer__link--prev" aria-label="Previous: model">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                model
              </div>
            </div>
          </a>
        
        
          
          <a href="../utils/" class="md-footer__link md-footer__link--next" aria-label="Next: utils">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                utils
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; 2024 Kyle Meador
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        <div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://www.linkedin.com/in/kylemeador" target="_blank" rel="noopener" title="www.linkedin.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3zM135.4 416H69V202.2h66.5V416zm-33.2-243c-21.3 0-38.5-17.3-38.5-38.5S80.9 96 102.2 96c21.2 0 38.5 17.3 38.5 38.5 0 21.3-17.2 38.5-38.5 38.5zm282.1 243h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9V416z"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.tabs", "navigation.tabs.sticky", "navigation.tracking", "navigation.instant", "navigation.path", "navigation.top", "navigation.footer", "search.suggest", "search.highlight", "toc.follow", "content.tabs.link", "content.code.annotation", "content.code.copy"], "search": "../../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../../assets/javascripts/bundle.8fd75fb4.min.js"></script>
      
    
  </body>
</html>